{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NetColoc analysis of rare variants in Autism spectrum disorder (ASD). \n",
    "\n",
    "Example of NetColoc workflow on genes associated with rare exome variants in childhjod demntia genes only, to demonstrate workflow from single gene set\n",
    "\n",
    "run this in the base conda env. \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Obtain input gene sets and interactome_\n",
    "\n",
    "## 1. Load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:21:41.078493Z",
     "start_time": "2022-06-24T21:21:35.224229Z"
    }
   },
   "outputs": [],
   "source": [
    "# load required packages\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import random\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "import getpass\n",
    "import ndex2\n",
    "\n",
    "import json\n",
    "import cdapsutil\n",
    "\n",
    "from gprofiler import GProfiler\n",
    "gp = GProfiler(\"MyToolName/0.1\")\n",
    "\n",
    "from scipy.stats import hypergeom\n",
    "from scipy.stats import norm\n",
    "\n",
    "# latex rendering of text in graphs\n",
    "import matplotlib as mpl\n",
    "mpl.rc('text', usetex = False)\n",
    "mpl.rc('font', family = 'serif')\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'sans-serif'\n",
    "rcParams['font.sans-serif'] = ['Arial']\n",
    "\n",
    "sns.set(font_scale=1.4)\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.set_style(\"ticks\", {\"xtick.major.size\": 15, \"ytick.major.size\": 15})\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "from datetime import datetime\n",
    "import sys\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:21:41.754661Z",
     "start_time": "2022-06-24T21:21:41.080114Z"
    }
   },
   "outputs": [],
   "source": [
    "# verify DDOT was installed\n",
    "import ddot\n",
    "\n",
    "from netcoloc import netprop_zscore, netprop, network_colocalization, validation\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Select one gene set of interest. Load gene set from text files into python.\n",
    "\n",
    "\n",
    "Identify one gene sets of interest. Gene sets should come from experimental data (not manual curation) to avoid bias. \n",
    "\n",
    "**Usage Note**: gene sets should be < 500 genes (propagation algorithm breaks down if seeded with larger sets). If your gene set is larger, only use the top 500 as seeds to the network propagation.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:21:41.757847Z",
     "start_time": "2022-06-24T21:21:41.755706Z"
    }
   },
   "outputs": [],
   "source": [
    "# set names of geneset 1\n",
    "# ------ customize this section based on your gene sets and how they should be labeled -------\n",
    "d1_name='ChildhoodDementia'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:21:41.828901Z",
     "start_time": "2022-06-24T21:21:41.759608Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of ChildhoodDementia genes: 235\n",
      "First 5 genes: ['SGSH', 'NAGLU', 'HGSNAT', 'GNS', 'MTTP']\n"
     ]
    }
   ],
   "source": [
    "# ------ customize this section based your input genesets -------\n",
    "\n",
    "# load rare variants Satterstrom et al 2020; PMID: 31981491\n",
    "D1_df = pd.read_csv('genesets/childhooddementiagenes.csv')\n",
    "D1_df.index = D1_df['GENE']\n",
    "print('Number of '+d1_name+' genes:', len(D1_df))\n",
    "D1_genes = D1_df.index.tolist() # define rare variant genes to seed network propagation\n",
    "print(\"First 5 genes:\", D1_genes[0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Select gene interaction network to use for the analysis.\n",
    "\n",
    "Identify network UUID on NDEx (ndexbio.org) and use this to import to this Jupyter notebook. We recommend using PCNet as a starting point, but a user may want to switch to “STRING high confidence” if using a machine with low memory (< 8GB RAM).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:21:57.611362Z",
     "start_time": "2022-06-24T21:21:41.868056Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes: 18820\n",
      "\n",
      "Number of edges: 2693109\n"
     ]
    }
   ],
   "source": [
    "interactome_uuid='4de852d9-9908-11e9-bcaf-0ac135e8bacf' # for PCNet\n",
    "# interactome_uuid='275bd84e-3d18-11e8-a935-0ac135e8bacf' # for STRING high confidence\n",
    "ndex_server='public.ndexbio.org'\n",
    "ndex_user=None\n",
    "ndex_password=None\n",
    "G_int = ndex2.create_nice_cx_from_server(\n",
    "            ndex_server, \n",
    "            username=ndex_user, \n",
    "            password=ndex_password, \n",
    "            uuid=interactome_uuid\n",
    "        ).to_networkx()\n",
    "nodes = list(G_int.nodes)\n",
    "\n",
    "# remove self edges from network\n",
    "G_int.remove_edges_from(nx.selfloop_edges(G_int))\n",
    "\n",
    "# print out the numbers of nodes and edges in the interatome for diagnostic purposes:\n",
    "print('Number of nodes:', len(G_int.nodes))\n",
    "print('\\nNumber of edges:', len(G_int.edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:21:57.616677Z",
     "start_time": "2022-06-24T21:21:57.612959Z"
    }
   },
   "outputs": [],
   "source": [
    "int_nodes = list(G_int.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Identify network colocalized gene network_\n",
    "\n",
    "## 4. Precalculate matrices needed for propagation. This step should take a few minutes (more for larger/denser networks) \n",
    "\n",
    "A benchmarking analysis demonstrates that the runtime required scales with the number of edges (w’) and the number of nodes (w’’). NetColoc includes functionality for saving and loading these matrices, by setting the XXX parameter, which can be useful if running multiple analyses. The diffusion parameter, which controls the rate of propagation through the network, may be set in this step. In practice, we have found that results are not dependent on the choice of this parameter, and recommend using the default value of 0.5.\n",
    "\n",
    "Background on network propagation: https://www.nature.com/articles/nrg.2017.38.pdf?origin=ppub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:25:22.674597Z",
     "start_time": "2022-06-24T21:21:57.617751Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "calculating w_prime\n",
      "\n",
      "calculating w_double_prime\n"
     ]
    }
   ],
   "source": [
    "# pre-calculate matrices used for network propagation. this step takes a few minutes, more for denser interactomes\n",
    "print('\\ncalculating w_prime')\n",
    "w_prime = netprop.get_normalized_adjacency_matrix(G_int, conserve_heat=True)\n",
    "\n",
    "print('\\ncalculating w_double_prime')\n",
    "w_double_prime = netprop.get_individual_heats_matrix(w_prime, .5)\n",
    "\n",
    "# w_double_prime can be saved to the file netcoloc_w_double_prime.npy\n",
    "# in current working directory with the following call:\n",
    "#\n",
    "# np.save('netcoloc_w_double_prime.npy', w_double_prime)\n",
    "\n",
    "# and reloaded later with:\n",
    "# w_double_prime = np.load('netcoloc_w_double_prime.npy')\n",
    "#\n",
    "# NOTE: Saving w_double_prime results in a several gigabyte file and\n",
    "#       takes a minute or more to save and load\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Subset seed genes to those found in the selected network. \n",
    "\n",
    "Only genes contained in the interaction network will be retained for downstream analysis. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:25:22.699275Z",
     "start_time": "2022-06-24T21:25:22.678138Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of D1 genes: 196\n",
      "Number of D1 genes in interactome: 196\n"
     ]
    }
   ],
   "source": [
    "# subset seed genes to those found in interactome\n",
    "print(\"Number of D1 genes:\", len(D1_genes))\n",
    "D1_genes = list(np.intersect1d(D1_genes,int_nodes))\n",
    "print(\"Number of D1 genes in interactome:\", len(D1_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Compute network proximity scores from seed gene set.\n",
    "\n",
    "The network proximity scores include a correction for the degree distribution of the input gene sets. The runtime required for computing the network proximity scores increases linearly with the number of nodes in the underlying interaction network and with the size of the input gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:25:55.101716Z",
     "start_time": "2022-06-24T21:25:22.700315Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating D1 z-scores: \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "476df06509774824aac80bcbd8eb93d1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1000 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DLAT</th>\n",
       "      <td>23.819753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF2B1</th>\n",
       "      <td>23.796715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRPPRC</th>\n",
       "      <td>23.454761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDHB</th>\n",
       "      <td>23.177739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLD</th>\n",
       "      <td>22.877058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                z\n",
       "DLAT    23.819753\n",
       "EIF2B1  23.796715\n",
       "LRPPRC  23.454761\n",
       "PDHB    23.177739\n",
       "DLD     22.877058"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# D1 network propagation\n",
    "print('\\nCalculating D1 z-scores: ')\n",
    "z_D1, Fnew_D1, Fnew_rand_D1 = netprop_zscore.calculate_heat_zscores(w_double_prime, int_nodes, \n",
    "                                                                    dict(G_int.degree), \n",
    "                                                                    D1_genes, num_reps=1000,\n",
    "                                                                    minimum_bin_size=100)\n",
    "\n",
    "z_D1 = pd.DataFrame({'z':z_D1})\n",
    "\n",
    "z_D1.sort_values('z',ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Build proximal subnetwork by taking z> thresh \n",
    "\n",
    "Select genes which have z>threshold (default = 3). These genes are proximal in network space to the seed gene set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:26:32.041773Z",
     "start_time": "2022-06-24T21:26:31.863678Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes in proximal subgraph: 748\n",
      "Edges in proximal subgraph: 31926\n"
     ]
    }
   ],
   "source": [
    "# ----------- select thresholds for NetColoc -----------------\n",
    "zthresh=3 # default = 3\n",
    "\n",
    "# select the genes in the network intersection, make a subgraph\n",
    "\n",
    "G_prox = nx.subgraph(G_int,z_D1[z_D1['z']>zthresh].index.tolist()) \n",
    "print(\"Nodes in proximal subgraph:\", len(G_prox.nodes()))\n",
    "print(\"Edges in proximal subgraph:\", len(G_prox.edges()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Transform NetColoc proximal subnetwork edges to cosine similarities (OPTIONAL)\n",
    "\n",
    "Transform NetColoc subnetwork edges to cosine similarities, with the function ‘network_colocalization.transform_edges’. The cosine similarity score between two genes represents the extent to which those genes have similar interactors. In practice, the cosine similarity transformed score helps to visually reveal the underlying clustering structure present in a network. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:46.477275Z",
     "start_time": "2022-06-24T21:26:35.075775Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/karissabarthelson/miniconda3/lib/python3.11/site-packages/netcoloc/network_colocalization.py:273: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!\n",
      "You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.\n",
      "A typical example is when you are setting values in a column of a DataFrame, like:\n",
      "\n",
      "df[\"col\"][row_indexer] = value\n",
      "\n",
      "Use `df.loc[row_indexer, \"col\"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "  cos_pc.loc[n1][n2] = cosine_distance\n",
      "/Users/karissabarthelson/miniconda3/lib/python3.11/site-packages/netcoloc/network_colocalization.py:274: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!\n",
      "You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.\n",
      "A typical example is when you are setting values in a column of a DataFrame, like:\n",
      "\n",
      "df[\"col\"][row_indexer] = value\n",
      "\n",
      "Use `df.loc[row_indexer, \"col\"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "  cos_pc.loc[n2][n1] = cosine_distance\n",
      "/Users/karissabarthelson/miniconda3/lib/python3.11/site-packages/netcoloc/network_colocalization.py:258: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!\n",
      "You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.\n",
      "A typical example is when you are setting values in a column of a DataFrame, like:\n",
      "\n",
      "df[\"col\"][row_indexer] = value\n",
      "\n",
      "Use `df.loc[row_indexer, \"col\"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "  cos_pc.loc[n1][n2] = 1.0\n",
      "/Users/karissabarthelson/miniconda3/lib/python3.11/site-packages/netcoloc/network_colocalization.py:259: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!\n",
      "You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.\n",
      "A typical example is when you are setting values in a column of a DataFrame, like:\n",
      "\n",
      "df[\"col\"][row_indexer] = value\n",
      "\n",
      "Use `df.loc[row_indexer, \"col\"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "  cos_pc.loc[n2][n1] = 1.0\n"
     ]
    }
   ],
   "source": [
    "G_cosSim=network_colocalization.transform_edges(G_prox,method='cosine_sim',edge_weight_threshold=0.95)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Compute systems map from proximal subgraph_\n",
    "\n",
    "## 9. Convert network colocalization subnetwork to form used in community detection module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:46.499084Z",
     "start_time": "2022-06-24T21:27:46.478749Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d1_seeds</th>\n",
       "      <th>d1_name</th>\n",
       "      <th>z_d1</th>\n",
       "      <th>sum_seeds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DLAT</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>23.819753</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF2B1</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>23.796715</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRPPRC</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>23.454761</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDHB</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>23.177739</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLD</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>22.877058</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF2B3</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>22.639214</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDHA1</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>22.497164</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF2B5</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>21.419588</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTR</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>21.417352</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>21.376330</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMADHC</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>21.173681</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF2B4</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>21.109565</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR3B</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>20.900250</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBTF</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>20.668777</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDHX</th>\n",
       "      <td>1</td>\n",
       "      <td>ChildhoodDementia</td>\n",
       "      <td>20.436673</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        d1_seeds            d1_name       z_d1  sum_seeds\n",
       "DLAT           1  ChildhoodDementia  23.819753          1\n",
       "EIF2B1         1  ChildhoodDementia  23.796715          1\n",
       "LRPPRC         1  ChildhoodDementia  23.454761          1\n",
       "PDHB           1  ChildhoodDementia  23.177739          1\n",
       "DLD            1  ChildhoodDementia  22.877058          1\n",
       "EIF2B3         1  ChildhoodDementia  22.639214          1\n",
       "PDHA1          1  ChildhoodDementia  22.497164          1\n",
       "EIF2B5         1  ChildhoodDementia  21.419588          1\n",
       "MTR            1  ChildhoodDementia  21.417352          1\n",
       "DPM1           1  ChildhoodDementia  21.376330          1\n",
       "MMADHC         1  ChildhoodDementia  21.173681          1\n",
       "EIF2B4         1  ChildhoodDementia  21.109565          1\n",
       "POLR3B         1  ChildhoodDementia  20.900250          1\n",
       "UBTF           1  ChildhoodDementia  20.668777          1\n",
       "PDHX           1  ChildhoodDementia  20.436673          1"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compile dataframe of metadata for overlapping nodes\n",
    "node_df = pd.DataFrame(index=list(G_prox.nodes))\n",
    "node_df = node_df.assign(d1_seeds=0, d1_name=d1_name,)\n",
    "node_df.loc[list(np.intersect1d(D1_genes,node_df.index.tolist())), 'd1_seeds']=1\n",
    "node_df['z_d1']=z_D1.loc[list(G_prox.nodes)]['z']\n",
    "node_df['sum_seeds']=node_df['d1_seeds']\n",
    "\n",
    "node_df = node_df.sort_values('z_d1',ascending=False)\n",
    "node_df.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Run community detection on NetColoc subnetwork (recommend HiDef).\n",
    "\n",
    "Documentation for CDAPS utils to build multiscale systems map in notebook\n",
    "\n",
    "https://cdapsutil.readthedocs.io/en/latest/quicktutorial.html#example  \n",
    "https://cdapsutil.readthedocs.io/en/latest/cdapsutil.html#community-detection  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.356899Z",
     "start_time": "2022-06-24T21:27:46.500105Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes in overlap subgraph: 748\n",
      "Edges in overlap subgraph: 31926\n"
     ]
    }
   ],
   "source": [
    "print(\"Nodes in overlap subgraph:\", len(G_prox.nodes()))\n",
    "print(\"Edges in overlap subgraph:\", len(G_prox.edges()))\n",
    "# Create cx format of overlap subgraph\n",
    "G_prox_cx = ndex2.create_nice_cx_from_networkx(G_prox)\n",
    "G_prox_cx.set_name(d1_name+'_NetColoc_subgraph') \n",
    "for node_id, node in G_prox_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_prox_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "cd = cdapsutil.CommunityDetection()\n",
    "\n",
    "# Run HiDeF on CDAPS REST service\n",
    "G_hier = cd.run_community_detection(G_prox_cx, algorithm='hidefv1.1beta',arguments={'--maxres':'20'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.362414Z",
     "start_time": "2022-06-24T21:27:53.358041Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hierarchy name: hidefv1.1beta_(none)_ChildhoodDementia_NetColoc_subgraph\n",
      "# nodes: 65\n",
      "# edges: 64\n"
     ]
    }
   ],
   "source": [
    "# Print information about hierarchy\n",
    "print('Hierarchy name: ' + str(G_hier.get_name()))\n",
    "print('# nodes: ' + str(len(G_hier.get_nodes())))\n",
    "print('# edges: ' + str(len(G_hier.get_edges())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.374104Z",
     "start_time": "2022-06-24T21:27:53.363427Z"
    }
   },
   "outputs": [],
   "source": [
    "# sr = cdapsutil.ServiceRunner()\n",
    "# sr.get_algorithms() # this will print out available options for each algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11. Convert the NetColoc hierarchy to networkx format, and write out features of the hierarchy to a pandas dataframe, for easier access in Python.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.388170Z",
     "start_time": "2022-06-24T21:27:53.375241Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes: 65\n",
      "\n",
      "Number of edges: 64\n"
     ]
    }
   ],
   "source": [
    "G_hier = G_hier.to_networkx(mode='default')\n",
    "G_hier\n",
    "\n",
    "nodes = G_hier.nodes()\n",
    "\n",
    "# print the number of nodes and edges in the hierarchy for diagnostic purposes\n",
    "print('Number of nodes:', len(G_hier.nodes()))\n",
    "\n",
    "print('\\nNumber of edges:', len(G_hier.edges()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.409753Z",
     "start_time": "2022-06-24T21:27:53.389278Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>represents</th>\n",
       "      <th>name</th>\n",
       "      <th>system_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PEPD ATP7B EIF2B1 SLC3A2 SNAP29 TPH1 ECI2 PEX2...</td>\n",
       "      <td>747</td>\n",
       "      <td>False</td>\n",
       "      <td>9.545</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>122</td>\n",
       "      <td>C748</td>\n",
       "      <td>C748</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCP11X2 TUBB4A KIF1A TUBG2</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9</td>\n",
       "      <td>C800</td>\n",
       "      <td>C800</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLYR1 KCTD7 GAS2L2 XRN2</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13</td>\n",
       "      <td>C801</td>\n",
       "      <td>C801</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATP7B COMMD1 ATP7A CCS ATOX1 SLC20A2</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>2.585</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25</td>\n",
       "      <td>C776</td>\n",
       "      <td>C776</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STN1 MED27 TEN1 ACD CTC1 RPA4</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>2.585</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>C777</td>\n",
       "      <td>C777</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       CD_MemberList  CD_MemberList_Size  \\\n",
       "0  PEPD ATP7B EIF2B1 SLC3A2 SNAP29 TPH1 ECI2 PEX2...                 747   \n",
       "1                         TCP11X2 TUBB4A KIF1A TUBG2                   4   \n",
       "2                            GLYR1 KCTD7 GAS2L2 XRN2                   4   \n",
       "3               ATP7B COMMD1 ATP7A CCS ATOX1 SLC20A2                   6   \n",
       "4                      STN1 MED27 TEN1 ACD CTC1 RPA4                   6   \n",
       "\n",
       "  CD_Labeled CD_MemberList_LogSize CD_CommunityName CD_AnnotatedMembers  \\\n",
       "0      False                 9.545                                        \n",
       "1      False                   2.0                                        \n",
       "2      False                   2.0                                        \n",
       "3      False                 2.585                                        \n",
       "4      False                 2.585                                        \n",
       "\n",
       "  CD_AnnotatedMembers_Size CD_AnnotatedMembers_Overlap  \\\n",
       "0                        0                         0.0   \n",
       "1                        0                         0.0   \n",
       "2                        0                         0.0   \n",
       "3                        0                         0.0   \n",
       "4                        0                         0.0   \n",
       "\n",
       "  CD_AnnotatedMembers_Pvalue  HiDeF_persistence represents  name  system_ID  \n",
       "0                        0.0                122       C748  C748          0  \n",
       "1                        0.0                  9       C800  C800          1  \n",
       "2                        0.0                 13       C801  C801          2  \n",
       "3                        0.0                 25       C776  C776          3  \n",
       "4                        0.0                 14       C777  C777          4  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add node attributes to dataframe for easier access\n",
    "hier_df = pd.DataFrame.from_dict(dict(G_hier.nodes(data=True)), orient='index')\n",
    "hier_df['system_ID']=hier_df.index.tolist()\n",
    "# some columns are not the right type\n",
    "hier_df['CD_MemberList_Size']=[int(x) for x in hier_df['CD_MemberList_Size'].tolist()]\n",
    "hier_df['HiDeF_persistence']=[int(x) for x in hier_df['HiDeF_persistence'].tolist()]\n",
    "hier_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Remove systems with no seed genes (OPTIONAL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.450988Z",
     "start_time": "2022-06-24T21:27:53.410904Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes with seed genes: 63\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "      <th>CD_Labeled</th>\n",
       "      <th>CD_MemberList_LogSize</th>\n",
       "      <th>CD_CommunityName</th>\n",
       "      <th>CD_AnnotatedMembers</th>\n",
       "      <th>CD_AnnotatedMembers_Size</th>\n",
       "      <th>CD_AnnotatedMembers_Overlap</th>\n",
       "      <th>CD_AnnotatedMembers_Pvalue</th>\n",
       "      <th>HiDeF_persistence</th>\n",
       "      <th>represents</th>\n",
       "      <th>name</th>\n",
       "      <th>system_ID</th>\n",
       "      <th>num_d1_seeds</th>\n",
       "      <th>frac_d1_seeds</th>\n",
       "      <th>frac_no_seeds</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C748</th>\n",
       "      <td>PEPD ATP7B EIF2B1 SLC3A2 SNAP29 TPH1 ECI2 PEX2...</td>\n",
       "      <td>747</td>\n",
       "      <td>False</td>\n",
       "      <td>9.545</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>122</td>\n",
       "      <td>C748</td>\n",
       "      <td>C748</td>\n",
       "      <td>0</td>\n",
       "      <td>196</td>\n",
       "      <td>0.262383</td>\n",
       "      <td>0.737617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C800</th>\n",
       "      <td>TCP11X2 TUBB4A KIF1A TUBG2</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9</td>\n",
       "      <td>C800</td>\n",
       "      <td>C800</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C801</th>\n",
       "      <td>GLYR1 KCTD7 GAS2L2 XRN2</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13</td>\n",
       "      <td>C801</td>\n",
       "      <td>C801</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C776</th>\n",
       "      <td>ATP7B COMMD1 ATP7A CCS ATOX1 SLC20A2</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>2.585</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25</td>\n",
       "      <td>C776</td>\n",
       "      <td>C776</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C777</th>\n",
       "      <td>STN1 MED27 TEN1 ACD CTC1 RPA4</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>2.585</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>C777</td>\n",
       "      <td>C777</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CD_MemberList  CD_MemberList_Size  \\\n",
       "name                                                                          \n",
       "C748  PEPD ATP7B EIF2B1 SLC3A2 SNAP29 TPH1 ECI2 PEX2...                 747   \n",
       "C800                         TCP11X2 TUBB4A KIF1A TUBG2                   4   \n",
       "C801                            GLYR1 KCTD7 GAS2L2 XRN2                   4   \n",
       "C776               ATP7B COMMD1 ATP7A CCS ATOX1 SLC20A2                   6   \n",
       "C777                      STN1 MED27 TEN1 ACD CTC1 RPA4                   6   \n",
       "\n",
       "     CD_Labeled CD_MemberList_LogSize CD_CommunityName CD_AnnotatedMembers  \\\n",
       "name                                                                         \n",
       "C748      False                 9.545                                        \n",
       "C800      False                   2.0                                        \n",
       "C801      False                   2.0                                        \n",
       "C776      False                 2.585                                        \n",
       "C777      False                 2.585                                        \n",
       "\n",
       "     CD_AnnotatedMembers_Size CD_AnnotatedMembers_Overlap  \\\n",
       "name                                                        \n",
       "C748                        0                         0.0   \n",
       "C800                        0                         0.0   \n",
       "C801                        0                         0.0   \n",
       "C776                        0                         0.0   \n",
       "C777                        0                         0.0   \n",
       "\n",
       "     CD_AnnotatedMembers_Pvalue  HiDeF_persistence represents  name  \\\n",
       "name                                                                  \n",
       "C748                        0.0                122       C748  C748   \n",
       "C800                        0.0                  9       C800  C800   \n",
       "C801                        0.0                 13       C801  C801   \n",
       "C776                        0.0                 25       C776  C776   \n",
       "C777                        0.0                 14       C777  C777   \n",
       "\n",
       "      system_ID  num_d1_seeds  frac_d1_seeds  frac_no_seeds  \n",
       "name                                                         \n",
       "C748          0           196       0.262383       0.737617  \n",
       "C800          1             2       0.500000       0.500000  \n",
       "C801          2             1       0.250000       0.750000  \n",
       "C776          3             2       0.333333       0.666667  \n",
       "C777          4             2       0.333333       0.666667  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hier_df.index=hier_df['name']\n",
    "hier_df.head()\n",
    "\n",
    "num_d1_seeds = []\n",
    "frac_d1_seeds=[]\n",
    "\n",
    "systems_keep = []\n",
    "for c in hier_df.index.tolist():\n",
    "    system_genes = hier_df['CD_MemberList'].loc[c].split(' ')\n",
    "    d1_temp = list(np.intersect1d(system_genes,D1_genes))\n",
    "\n",
    "    num_d1_temp = len(d1_temp)\n",
    "    if (num_d1_temp)>0: # keep the system if it has at least 1 seed gene\n",
    "        systems_keep.append(c)\n",
    "        num_d1_seeds.append(num_d1_temp)\n",
    "        \n",
    "        frac_d1_seeds.append((num_d1_temp)/float(len(system_genes)))\n",
    "\n",
    "        \n",
    "frac_no_seeds = np.subtract(1.0,np.array([frac_d1_seeds]).sum(axis=0))\n",
    "\n",
    "hier_df = hier_df.loc[systems_keep]\n",
    "hier_df['num_d1_seeds']=num_d1_seeds\n",
    "hier_df['frac_d1_seeds']=frac_d1_seeds\n",
    "hier_df['frac_no_seeds']=frac_no_seeds\n",
    "print(\"Number of nodes with seed genes:\", len(hier_df))\n",
    "\n",
    "hier_df.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.456748Z",
     "start_time": "2022-06-24T21:27:53.451983Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes with seed genes: 63\n",
      "Number of edges remaining: 62\n"
     ]
    }
   ],
   "source": [
    "# prune G_hier--> only keep systems with at least one seed gene\n",
    "\n",
    "nkeep=[]\n",
    "for n in list(G_hier.nodes()):\n",
    "    if G_hier.nodes(data=True)[n]['name'] in systems_keep:\n",
    "        nkeep.append(n)\n",
    "        \n",
    "\n",
    "G_hier = nx.subgraph(G_hier, nkeep)\n",
    "print(\"Number of nodes with seed genes:\", len(G_hier.nodes()))\n",
    "print(\"Number of edges remaining:\", len(G_hier.edges()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 13. Sneak peek of hierarchy visualization (OPTIONAL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:27:53.850097Z",
     "start_time": "2022-06-24T21:27:53.457862Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0f87e6c49cac4590912e07abf3047fb2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "CytoscapeWidget(cytoscape_layout={'name': 'cose'}, cytoscape_style=[{'selector': 'node', 'css': {'content': 'd…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "network_colocalization.view_G_hier(G_hier)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 14. Annotate systems with gprofiler. \n",
    "\n",
    "Annotate moderately sized systems (between 50 to 1000 genes per system) if they are significantly enriched for a Gene Ontology biological process. Also require that the GO term is enriched with $p<10^{-5}$ and shares at least 3 genes with the system to annotate, to increase the stringency of the annotation. Label the system using the GO term that meets these criteria, and has the highest sum of precision and recall. Systems which have no GO terms meeting these criteria are labeled with their unique system ID.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:28:31.074583Z",
     "start_time": "2022-06-24T21:27:53.851319Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C748\n",
      "747\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pyruvate metabolism and Citric Acid (TCA) cycle</td>\n",
       "      <td>10842</td>\n",
       "      <td>26</td>\n",
       "      <td>Pyruvate metabolism and Citric Acid (TCA) cycle</td>\n",
       "      <td>REAC:R-HSA-71406</td>\n",
       "      <td>1.162152e-17</td>\n",
       "      <td>[REAC:R-HSA-1428517]</td>\n",
       "      <td>0.047016</td>\n",
       "      <td>query_1</td>\n",
       "      <td>553</td>\n",
       "      <td>0.481481</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>10842</td>\n",
       "      <td>25</td>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>REAC:R-HSA-9033241</td>\n",
       "      <td>2.169400e-15</td>\n",
       "      <td>[REAC:R-HSA-9609507]</td>\n",
       "      <td>0.045208</td>\n",
       "      <td>query_1</td>\n",
       "      <td>553</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>31</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>6.387769e-17</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.056058</td>\n",
       "      <td>query_1</td>\n",
       "      <td>553</td>\n",
       "      <td>0.360465</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>10842</td>\n",
       "      <td>59</td>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>REAC:R-HSA-1428517</td>\n",
       "      <td>3.049943e-30</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.106691</td>\n",
       "      <td>query_1</td>\n",
       "      <td>553</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism of water-soluble vitamins and cofac...</td>\n",
       "      <td>10842</td>\n",
       "      <td>37</td>\n",
       "      <td>Metabolism of water-soluble vitamins and cofac...</td>\n",
       "      <td>REAC:R-HSA-196849</td>\n",
       "      <td>2.183299e-17</td>\n",
       "      <td>[REAC:R-HSA-196854]</td>\n",
       "      <td>0.066908</td>\n",
       "      <td>query_1</td>\n",
       "      <td>553</td>\n",
       "      <td>0.305785</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "5     Pyruvate metabolism and Citric Acid (TCA) cycle                  10842   \n",
       "12                         Peroxisomal protein import                  10842   \n",
       "9                             Sphingolipid metabolism                  10842   \n",
       "1   The citric acid (TCA) cycle and respiratory el...                  10842   \n",
       "6   Metabolism of water-soluble vitamins and cofac...                  10842   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "5                  26    Pyruvate metabolism and Citric Acid (TCA) cycle   \n",
       "12                 25                         Peroxisomal protein import   \n",
       "9                  31                            Sphingolipid metabolism   \n",
       "1                  59  The citric acid (TCA) cycle and respiratory el...   \n",
       "6                  37  Metabolism of water-soluble vitamins and cofac...   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "5     REAC:R-HSA-71406  1.162152e-17  [REAC:R-HSA-1428517]   0.047016   \n",
       "12  REAC:R-HSA-9033241  2.169400e-15  [REAC:R-HSA-9609507]   0.045208   \n",
       "9    REAC:R-HSA-428157  6.387769e-17   [REAC:R-HSA-556833]   0.056058   \n",
       "1   REAC:R-HSA-1428517  3.049943e-30  [REAC:R-HSA-1430728]   0.106691   \n",
       "6    REAC:R-HSA-196849  2.183299e-17   [REAC:R-HSA-196854]   0.066908   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "5   query_1         553  0.481481         True   REAC         54  \n",
       "12  query_1         553  0.416667         True   REAC         60  \n",
       "9   query_1         553  0.360465         True   REAC         86  \n",
       "1   query_1         553  0.333333         True   REAC        177  \n",
       "6   query_1         553  0.305785         True   REAC        121  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C800\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C801\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C776\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C777\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Extension of Telomeres</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Extension of Telomeres</td>\n",
       "      <td>REAC:R-HSA-180786</td>\n",
       "      <td>7.116160e-08</td>\n",
       "      <td>[REAC:R-HSA-157579]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.078431</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Telomere Maintenance</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Telomere Maintenance</td>\n",
       "      <td>REAC:R-HSA-157579</td>\n",
       "      <td>1.352247e-06</td>\n",
       "      <td>[REAC:R-HSA-73886]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.036036</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chromosome Maintenance</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromosome Maintenance</td>\n",
       "      <td>REAC:R-HSA-73886</td>\n",
       "      <td>2.710411e-06</td>\n",
       "      <td>[REAC:R-HSA-1640170]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.028986</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              description  effective_domain_size  intersection_size  \\\n",
       "3  Extension of Telomeres                  10842                  4   \n",
       "4    Telomere Maintenance                  10842                  4   \n",
       "5  Chromosome Maintenance                  10842                  4   \n",
       "\n",
       "                     name             native       p_value  \\\n",
       "3  Extension of Telomeres  REAC:R-HSA-180786  7.116160e-08   \n",
       "4    Telomere Maintenance  REAC:R-HSA-157579  1.352247e-06   \n",
       "5  Chromosome Maintenance   REAC:R-HSA-73886  2.710411e-06   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "3   [REAC:R-HSA-157579]   0.666667  query_1           6  0.078431   \n",
       "4    [REAC:R-HSA-73886]   0.666667  query_1           6  0.036036   \n",
       "5  [REAC:R-HSA-1640170]   0.666667  query_1           6  0.028986   \n",
       "\n",
       "   significant source  term_size  \n",
       "3         True   REAC         51  \n",
       "4         True   REAC        111  \n",
       "5         True   REAC        138  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C778\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Extension of Telomeres</td>\n",
       "      <td>10842</td>\n",
       "      <td>3</td>\n",
       "      <td>Extension of Telomeres</td>\n",
       "      <td>REAC:R-HSA-180786</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>[REAC:R-HSA-157579]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              description  effective_domain_size  intersection_size  \\\n",
       "3  Extension of Telomeres                  10842                  3   \n",
       "\n",
       "                     name             native   p_value              parents  \\\n",
       "3  Extension of Telomeres  REAC:R-HSA-180786  0.000001  [REAC:R-HSA-157579]   \n",
       "\n",
       "   precision    query  query_size    recall  significant source  term_size  \n",
       "3       0.75  query_1           4  0.058824         True   REAC         51  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C809\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C749\n",
      "458\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pyruvate metabolism and Citric Acid (TCA) cycle</td>\n",
       "      <td>10842</td>\n",
       "      <td>26</td>\n",
       "      <td>Pyruvate metabolism and Citric Acid (TCA) cycle</td>\n",
       "      <td>REAC:R-HSA-71406</td>\n",
       "      <td>1.883569e-22</td>\n",
       "      <td>[REAC:R-HSA-1428517]</td>\n",
       "      <td>0.071625</td>\n",
       "      <td>query_1</td>\n",
       "      <td>363</td>\n",
       "      <td>0.481481</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>10842</td>\n",
       "      <td>23</td>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>REAC:R-HSA-9033241</td>\n",
       "      <td>2.219376e-17</td>\n",
       "      <td>[REAC:R-HSA-9609507]</td>\n",
       "      <td>0.063361</td>\n",
       "      <td>query_1</td>\n",
       "      <td>363</td>\n",
       "      <td>0.383333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>10842</td>\n",
       "      <td>59</td>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>REAC:R-HSA-1428517</td>\n",
       "      <td>4.563020e-41</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.162534</td>\n",
       "      <td>query_1</td>\n",
       "      <td>363</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Protein localization</td>\n",
       "      <td>10842</td>\n",
       "      <td>45</td>\n",
       "      <td>Protein localization</td>\n",
       "      <td>REAC:R-HSA-9609507</td>\n",
       "      <td>1.070751e-27</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.123967</td>\n",
       "      <td>query_1</td>\n",
       "      <td>363</td>\n",
       "      <td>0.284810</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>10842</td>\n",
       "      <td>35</td>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>REAC:R-HSA-163200</td>\n",
       "      <td>7.231846e-21</td>\n",
       "      <td>[REAC:R-HSA-1428517]</td>\n",
       "      <td>0.096419</td>\n",
       "      <td>query_1</td>\n",
       "      <td>363</td>\n",
       "      <td>0.275591</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "4     Pyruvate metabolism and Citric Acid (TCA) cycle                  10842   \n",
       "12                         Peroxisomal protein import                  10842   \n",
       "1   The citric acid (TCA) cycle and respiratory el...                  10842   \n",
       "3                                Protein localization                  10842   \n",
       "5   Respiratory electron transport, ATP synthesis ...                  10842   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "4                  26    Pyruvate metabolism and Citric Acid (TCA) cycle   \n",
       "12                 23                         Peroxisomal protein import   \n",
       "1                  59  The citric acid (TCA) cycle and respiratory el...   \n",
       "3                  45                               Protein localization   \n",
       "5                  35  Respiratory electron transport, ATP synthesis ...   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "4     REAC:R-HSA-71406  1.883569e-22  [REAC:R-HSA-1428517]   0.071625   \n",
       "12  REAC:R-HSA-9033241  2.219376e-17  [REAC:R-HSA-9609507]   0.063361   \n",
       "1   REAC:R-HSA-1428517  4.563020e-41  [REAC:R-HSA-1430728]   0.162534   \n",
       "3   REAC:R-HSA-9609507  1.070751e-27        [REAC:0000000]   0.123967   \n",
       "5    REAC:R-HSA-163200  7.231846e-21  [REAC:R-HSA-1428517]   0.096419   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "4   query_1         363  0.481481         True   REAC         54  \n",
       "12  query_1         363  0.383333         True   REAC         60  \n",
       "1   query_1         363  0.333333         True   REAC        177  \n",
       "3   query_1         363  0.284810         True   REAC        158  \n",
       "5   query_1         363  0.275591         True   REAC        127  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C751\n",
      "191\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>29</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>1.015042e-32</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.223077</td>\n",
       "      <td>query_1</td>\n",
       "      <td>130</td>\n",
       "      <td>0.337209</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Glycosaminoglycan metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>11</td>\n",
       "      <td>Glycosaminoglycan metabolism</td>\n",
       "      <td>REAC:R-HSA-1630316</td>\n",
       "      <td>5.477562e-06</td>\n",
       "      <td>[REAC:R-HSA-71387]</td>\n",
       "      <td>0.084615</td>\n",
       "      <td>query_1</td>\n",
       "      <td>130</td>\n",
       "      <td>0.089431</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neutrophil degranulation</td>\n",
       "      <td>10842</td>\n",
       "      <td>37</td>\n",
       "      <td>Neutrophil degranulation</td>\n",
       "      <td>REAC:R-HSA-6798695</td>\n",
       "      <td>1.734152e-18</td>\n",
       "      <td>[REAC:R-HSA-168249]</td>\n",
       "      <td>0.284615</td>\n",
       "      <td>query_1</td>\n",
       "      <td>130</td>\n",
       "      <td>0.077731</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Diseases of metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>15</td>\n",
       "      <td>Diseases of metabolism</td>\n",
       "      <td>REAC:R-HSA-5668914</td>\n",
       "      <td>5.477562e-06</td>\n",
       "      <td>[REAC:R-HSA-1643685]</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>query_1</td>\n",
       "      <td>130</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Metabolism of carbohydrates</td>\n",
       "      <td>10842</td>\n",
       "      <td>17</td>\n",
       "      <td>Metabolism of carbohydrates</td>\n",
       "      <td>REAC:R-HSA-71387</td>\n",
       "      <td>1.325420e-06</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.130769</td>\n",
       "      <td>query_1</td>\n",
       "      <td>130</td>\n",
       "      <td>0.059441</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     description  effective_domain_size  intersection_size  \\\n",
       "0        Sphingolipid metabolism                  10842                 29   \n",
       "15  Glycosaminoglycan metabolism                  10842                 11   \n",
       "2       Neutrophil degranulation                  10842                 37   \n",
       "14        Diseases of metabolism                  10842                 15   \n",
       "13   Metabolism of carbohydrates                  10842                 17   \n",
       "\n",
       "                            name              native       p_value  \\\n",
       "0        Sphingolipid metabolism   REAC:R-HSA-428157  1.015042e-32   \n",
       "15  Glycosaminoglycan metabolism  REAC:R-HSA-1630316  5.477562e-06   \n",
       "2       Neutrophil degranulation  REAC:R-HSA-6798695  1.734152e-18   \n",
       "14        Diseases of metabolism  REAC:R-HSA-5668914  5.477562e-06   \n",
       "13   Metabolism of carbohydrates    REAC:R-HSA-71387  1.325420e-06   \n",
       "\n",
       "                 parents  precision    query  query_size    recall  \\\n",
       "0    [REAC:R-HSA-556833]   0.223077  query_1         130  0.337209   \n",
       "15    [REAC:R-HSA-71387]   0.084615  query_1         130  0.089431   \n",
       "2    [REAC:R-HSA-168249]   0.284615  query_1         130  0.077731   \n",
       "14  [REAC:R-HSA-1643685]   0.115385  query_1         130  0.061224   \n",
       "13  [REAC:R-HSA-1430728]   0.130769  query_1         130  0.059441   \n",
       "\n",
       "    significant source  term_size  \n",
       "0          True   REAC         86  \n",
       "15         True   REAC        123  \n",
       "2          True   REAC        476  \n",
       "14         True   REAC        245  \n",
       "13         True   REAC        286  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C761\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dual incision in TC-NER</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Dual incision in TC-NER</td>\n",
       "      <td>REAC:R-HSA-6782135</td>\n",
       "      <td>1.403451e-10</td>\n",
       "      <td>[REAC:R-HSA-6781827]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transcription-Coupled Nucleotide Excision Repa...</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Transcription-Coupled Nucleotide Excision Repa...</td>\n",
       "      <td>REAC:R-HSA-6781827</td>\n",
       "      <td>2.214507e-10</td>\n",
       "      <td>[REAC:R-HSA-5696398]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Formation of TC-NER Pre-Incision Complex</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Formation of TC-NER Pre-Incision Complex</td>\n",
       "      <td>REAC:R-HSA-6781823</td>\n",
       "      <td>8.225598e-07</td>\n",
       "      <td>[REAC:R-HSA-6781827]</td>\n",
       "      <td>0.50</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
       "      <td>REAC:R-HSA-6782210</td>\n",
       "      <td>1.546247e-06</td>\n",
       "      <td>[REAC:R-HSA-6781827]</td>\n",
       "      <td>0.50</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.063492</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nucleotide Excision Repair</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Nucleotide Excision Repair</td>\n",
       "      <td>REAC:R-HSA-5696398</td>\n",
       "      <td>1.254533e-09</td>\n",
       "      <td>[REAC:R-HSA-73894]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.055046</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "0                            Dual incision in TC-NER                  10842   \n",
       "1  Transcription-Coupled Nucleotide Excision Repa...                  10842   \n",
       "5           Formation of TC-NER Pre-Incision Complex                  10842   \n",
       "6  Gap-filling DNA repair synthesis and ligation ...                  10842   \n",
       "2                         Nucleotide Excision Repair                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "0                  6                            Dual incision in TC-NER   \n",
       "1                  6  Transcription-Coupled Nucleotide Excision Repa...   \n",
       "5                  4           Formation of TC-NER Pre-Incision Complex   \n",
       "6                  4  Gap-filling DNA repair synthesis and ligation ...   \n",
       "2                  6                         Nucleotide Excision Repair   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "0  REAC:R-HSA-6782135  1.403451e-10  [REAC:R-HSA-6781827]       0.75  query_1   \n",
       "1  REAC:R-HSA-6781827  2.214507e-10  [REAC:R-HSA-5696398]       0.75  query_1   \n",
       "5  REAC:R-HSA-6781823  8.225598e-07  [REAC:R-HSA-6781827]       0.50  query_1   \n",
       "6  REAC:R-HSA-6782210  1.546247e-06  [REAC:R-HSA-6781827]       0.50  query_1   \n",
       "2  REAC:R-HSA-5696398  1.254533e-09    [REAC:R-HSA-73894]       0.75  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           8  0.093750         True   REAC         64  \n",
       "1           8  0.077922         True   REAC         77  \n",
       "5           8  0.076923         True   REAC         52  \n",
       "6           8  0.063492         True   REAC         63  \n",
       "2           8  0.055046         True   REAC        109  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C762\n",
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C765\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C766\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C767\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Diseases of metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Diseases of metabolism</td>\n",
       "      <td>REAC:R-HSA-5668914</td>\n",
       "      <td>2.336764e-09</td>\n",
       "      <td>[REAC:R-HSA-1643685]</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>query_1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.024490</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism of carbohydrates</td>\n",
       "      <td>10842</td>\n",
       "      <td>7</td>\n",
       "      <td>Metabolism of carbohydrates</td>\n",
       "      <td>REAC:R-HSA-71387</td>\n",
       "      <td>2.618828e-11</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.024476</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   description  effective_domain_size  intersection_size  \\\n",
       "6       Diseases of metabolism                  10842                  6   \n",
       "5  Metabolism of carbohydrates                  10842                  7   \n",
       "\n",
       "                          name              native       p_value  \\\n",
       "6       Diseases of metabolism  REAC:R-HSA-5668914  2.336764e-09   \n",
       "5  Metabolism of carbohydrates    REAC:R-HSA-71387  2.618828e-11   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "6  [REAC:R-HSA-1643685]   0.857143  query_1           7  0.024490   \n",
       "5  [REAC:R-HSA-1430728]   1.000000  query_1           7  0.024476   \n",
       "\n",
       "   significant source  term_size  \n",
       "6         True   REAC        245  \n",
       "5         True   REAC        286  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C769\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C771\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>REAC:R-HSA-71291</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.013514</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 description  effective_domain_size  \\\n",
       "1  Metabolism of amino acids and derivatives                  10842   \n",
       "\n",
       "   intersection_size                                       name  \\\n",
       "1                  5  Metabolism of amino acids and derivatives   \n",
       "\n",
       "             native   p_value               parents  precision    query  \\\n",
       "1  REAC:R-HSA-71291  0.000002  [REAC:R-HSA-1430728]   0.833333  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1           6  0.013514         True   REAC        370  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C804\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C805\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C807\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C810\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C780\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>REAC:R-HSA-71291</td>\n",
       "      <td>4.555979e-09</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.016216</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 description  effective_domain_size  \\\n",
       "1  Metabolism of amino acids and derivatives                  10842   \n",
       "\n",
       "   intersection_size                                       name  \\\n",
       "1                  6  Metabolism of amino acids and derivatives   \n",
       "\n",
       "             native       p_value               parents  precision    query  \\\n",
       "1  REAC:R-HSA-71291  4.555979e-09  [REAC:R-HSA-1430728]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1           6  0.016216         True   REAC        370  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C812\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C750\n",
      "236\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>10842</td>\n",
       "      <td>31</td>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>REAC:R-HSA-163200</td>\n",
       "      <td>5.328602e-26</td>\n",
       "      <td>[REAC:R-HSA-1428517]</td>\n",
       "      <td>0.174157</td>\n",
       "      <td>query_1</td>\n",
       "      <td>178</td>\n",
       "      <td>0.244094</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Respiratory electron transport</td>\n",
       "      <td>10842</td>\n",
       "      <td>23</td>\n",
       "      <td>Respiratory electron transport</td>\n",
       "      <td>REAC:R-HSA-611105</td>\n",
       "      <td>4.628504e-18</td>\n",
       "      <td>[REAC:R-HSA-163200]</td>\n",
       "      <td>0.129213</td>\n",
       "      <td>query_1</td>\n",
       "      <td>178</td>\n",
       "      <td>0.223301</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>10842</td>\n",
       "      <td>38</td>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>REAC:R-HSA-1428517</td>\n",
       "      <td>1.363242e-29</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.213483</td>\n",
       "      <td>query_1</td>\n",
       "      <td>178</td>\n",
       "      <td>0.214689</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Complex I biogenesis</td>\n",
       "      <td>10842</td>\n",
       "      <td>11</td>\n",
       "      <td>Complex I biogenesis</td>\n",
       "      <td>REAC:R-HSA-6799198</td>\n",
       "      <td>1.036751e-07</td>\n",
       "      <td>[REAC:R-HSA-611105]</td>\n",
       "      <td>0.061798</td>\n",
       "      <td>query_1</td>\n",
       "      <td>178</td>\n",
       "      <td>0.192982</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Mitochondrial protein import</td>\n",
       "      <td>10842</td>\n",
       "      <td>12</td>\n",
       "      <td>Mitochondrial protein import</td>\n",
       "      <td>REAC:R-HSA-1268020</td>\n",
       "      <td>3.606188e-08</td>\n",
       "      <td>[REAC:R-HSA-9609507]</td>\n",
       "      <td>0.067416</td>\n",
       "      <td>query_1</td>\n",
       "      <td>178</td>\n",
       "      <td>0.184615</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "2  Respiratory electron transport, ATP synthesis ...                  10842   \n",
       "3                     Respiratory electron transport                  10842   \n",
       "0  The citric acid (TCA) cycle and respiratory el...                  10842   \n",
       "7                               Complex I biogenesis                  10842   \n",
       "6                       Mitochondrial protein import                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "2                 31  Respiratory electron transport, ATP synthesis ...   \n",
       "3                 23                     Respiratory electron transport   \n",
       "0                 38  The citric acid (TCA) cycle and respiratory el...   \n",
       "7                 11                               Complex I biogenesis   \n",
       "6                 12                       Mitochondrial protein import   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "2   REAC:R-HSA-163200  5.328602e-26  [REAC:R-HSA-1428517]   0.174157  query_1   \n",
       "3   REAC:R-HSA-611105  4.628504e-18   [REAC:R-HSA-163200]   0.129213  query_1   \n",
       "0  REAC:R-HSA-1428517  1.363242e-29  [REAC:R-HSA-1430728]   0.213483  query_1   \n",
       "7  REAC:R-HSA-6799198  1.036751e-07   [REAC:R-HSA-611105]   0.061798  query_1   \n",
       "6  REAC:R-HSA-1268020  3.606188e-08  [REAC:R-HSA-9609507]   0.067416  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "2         178  0.244094         True   REAC        127  \n",
       "3         178  0.223301         True   REAC        103  \n",
       "0         178  0.214689         True   REAC        177  \n",
       "7         178  0.192982         True   REAC         57  \n",
       "6         178  0.184615         True   REAC         65  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C784\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C785\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C786\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C755\n",
      "22\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>10842</td>\n",
       "      <td>12</td>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>REAC:R-HSA-9033241</td>\n",
       "      <td>2.479505e-23</td>\n",
       "      <td>[REAC:R-HSA-9609507]</td>\n",
       "      <td>0.705882</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Protein localization</td>\n",
       "      <td>10842</td>\n",
       "      <td>15</td>\n",
       "      <td>Protein localization</td>\n",
       "      <td>REAC:R-HSA-9609507</td>\n",
       "      <td>6.129413e-25</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.882353</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.094937</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E3 ubiquitin ligases ubiquitinate target proteins</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>E3 ubiquitin ligases ubiquitinate target proteins</td>\n",
       "      <td>REAC:R-HSA-8866654</td>\n",
       "      <td>1.731659e-07</td>\n",
       "      <td>[REAC:R-HSA-8852135]</td>\n",
       "      <td>0.294118</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.086207</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Protein ubiquitination</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>Protein ubiquitination</td>\n",
       "      <td>REAC:R-HSA-8852135</td>\n",
       "      <td>6.265445e-07</td>\n",
       "      <td>[REAC:R-HSA-597592]</td>\n",
       "      <td>0.294118</td>\n",
       "      <td>query_1</td>\n",
       "      <td>17</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "1                         Peroxisomal protein import                  10842   \n",
       "0                               Protein localization                  10842   \n",
       "3  E3 ubiquitin ligases ubiquitinate target proteins                  10842   \n",
       "4                             Protein ubiquitination                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "1                 12                         Peroxisomal protein import   \n",
       "0                 15                               Protein localization   \n",
       "3                  5  E3 ubiquitin ligases ubiquitinate target proteins   \n",
       "4                  5                             Protein ubiquitination   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "1  REAC:R-HSA-9033241  2.479505e-23  [REAC:R-HSA-9609507]   0.705882  query_1   \n",
       "0  REAC:R-HSA-9609507  6.129413e-25        [REAC:0000000]   0.882353  query_1   \n",
       "3  REAC:R-HSA-8866654  1.731659e-07  [REAC:R-HSA-8852135]   0.294118  query_1   \n",
       "4  REAC:R-HSA-8852135  6.265445e-07   [REAC:R-HSA-597592]   0.294118  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1          17  0.200000         True   REAC         60  \n",
       "0          17  0.094937         True   REAC        158  \n",
       "3          17  0.086207         True   REAC         58  \n",
       "4          17  0.064103         True   REAC         78  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C763\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Metabolism of water-soluble vitamins and cofac...</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Metabolism of water-soluble vitamins and cofac...</td>\n",
       "      <td>REAC:R-HSA-196849</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>[REAC:R-HSA-196854]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.033058</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Metabolism of vitamins and cofactors</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Metabolism of vitamins and cofactors</td>\n",
       "      <td>REAC:R-HSA-196854</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.021505</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "2  Metabolism of water-soluble vitamins and cofac...                  10842   \n",
       "3               Metabolism of vitamins and cofactors                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "2                  4  Metabolism of water-soluble vitamins and cofac...   \n",
       "3                  4               Metabolism of vitamins and cofactors   \n",
       "\n",
       "              native   p_value               parents  precision    query  \\\n",
       "2  REAC:R-HSA-196849  0.000001   [REAC:R-HSA-196854]   0.666667  query_1   \n",
       "3  REAC:R-HSA-196854  0.000004  [REAC:R-HSA-1430728]   0.666667  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "2           6  0.033058         True   REAC        121  \n",
       "3           6  0.021505         True   REAC        186  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C764\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>REAC:R-HSA-71291</td>\n",
       "      <td>7.214341e-07</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.013514</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 description  effective_domain_size  \\\n",
       "0  Metabolism of amino acids and derivatives                  10842   \n",
       "\n",
       "   intersection_size                                       name  \\\n",
       "0                  5  Metabolism of amino acids and derivatives   \n",
       "\n",
       "             native       p_value               parents  precision    query  \\\n",
       "0  REAC:R-HSA-71291  7.214341e-07  [REAC:R-HSA-1430728]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           5  0.013514         True   REAC        370  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C802\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C803\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C808\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C754\n",
      "23\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Complex I biogenesis</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Complex I biogenesis</td>\n",
       "      <td>REAC:R-HSA-6799198</td>\n",
       "      <td>2.150181e-13</td>\n",
       "      <td>[REAC:R-HSA-611105]</td>\n",
       "      <td>0.421053</td>\n",
       "      <td>query_1</td>\n",
       "      <td>19</td>\n",
       "      <td>0.140351</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Respiratory electron transport</td>\n",
       "      <td>10842</td>\n",
       "      <td>13</td>\n",
       "      <td>Respiratory electron transport</td>\n",
       "      <td>REAC:R-HSA-611105</td>\n",
       "      <td>6.889369e-22</td>\n",
       "      <td>[REAC:R-HSA-163200]</td>\n",
       "      <td>0.684211</td>\n",
       "      <td>query_1</td>\n",
       "      <td>19</td>\n",
       "      <td>0.126214</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>10842</td>\n",
       "      <td>16</td>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>REAC:R-HSA-163200</td>\n",
       "      <td>1.516828e-27</td>\n",
       "      <td>[REAC:R-HSA-1428517]</td>\n",
       "      <td>0.842105</td>\n",
       "      <td>query_1</td>\n",
       "      <td>19</td>\n",
       "      <td>0.125984</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>10842</td>\n",
       "      <td>16</td>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>REAC:R-HSA-1428517</td>\n",
       "      <td>2.021200e-25</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.842105</td>\n",
       "      <td>query_1</td>\n",
       "      <td>19</td>\n",
       "      <td>0.090395</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "3                               Complex I biogenesis                  10842   \n",
       "2                     Respiratory electron transport                  10842   \n",
       "0  Respiratory electron transport, ATP synthesis ...                  10842   \n",
       "1  The citric acid (TCA) cycle and respiratory el...                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "3                  8                               Complex I biogenesis   \n",
       "2                 13                     Respiratory electron transport   \n",
       "0                 16  Respiratory electron transport, ATP synthesis ...   \n",
       "1                 16  The citric acid (TCA) cycle and respiratory el...   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "3  REAC:R-HSA-6799198  2.150181e-13   [REAC:R-HSA-611105]   0.421053  query_1   \n",
       "2   REAC:R-HSA-611105  6.889369e-22   [REAC:R-HSA-163200]   0.684211  query_1   \n",
       "0   REAC:R-HSA-163200  1.516828e-27  [REAC:R-HSA-1428517]   0.842105  query_1   \n",
       "1  REAC:R-HSA-1428517  2.021200e-25  [REAC:R-HSA-1430728]   0.842105  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "3          19  0.140351         True   REAC         57  \n",
       "2          19  0.126214         True   REAC        103  \n",
       "0          19  0.125984         True   REAC        127  \n",
       "1          19  0.090395         True   REAC        177  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C788\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C759\n",
      "12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C792\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C797\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C798\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C774\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transport of bile salts and organic acids, met...</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Transport of bile salts and organic acids, met...</td>\n",
       "      <td>REAC:R-HSA-425366</td>\n",
       "      <td>1.988229e-07</td>\n",
       "      <td>[REAC:R-HSA-425407]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glycerophospholipid biosynthesis</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Glycerophospholipid biosynthesis</td>\n",
       "      <td>REAC:R-HSA-1483206</td>\n",
       "      <td>1.925997e-11</td>\n",
       "      <td>[REAC:R-HSA-1483257]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Phospholipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Phospholipid metabolism</td>\n",
       "      <td>REAC:R-HSA-1483257</td>\n",
       "      <td>2.623999e-10</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SLC-mediated transmembrane transport</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>SLC-mediated transmembrane transport</td>\n",
       "      <td>REAC:R-HSA-425407</td>\n",
       "      <td>9.823245e-06</td>\n",
       "      <td>[REAC:R-HSA-382551]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.016327</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>REAC:R-HSA-556833</td>\n",
       "      <td>2.973303e-07</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>query_1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>732</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "3  Transport of bile salts and organic acids, met...                  10842   \n",
       "1                   Glycerophospholipid biosynthesis                  10842   \n",
       "2                            Phospholipid metabolism                  10842   \n",
       "5               SLC-mediated transmembrane transport                  10842   \n",
       "4                               Metabolism of lipids                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "3                  4  Transport of bile salts and organic acids, met...   \n",
       "1                  6                   Glycerophospholipid biosynthesis   \n",
       "2                  6                            Phospholipid metabolism   \n",
       "5                  4               SLC-mediated transmembrane transport   \n",
       "4                  6                               Metabolism of lipids   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "3   REAC:R-HSA-425366  1.988229e-07   [REAC:R-HSA-425407]   0.666667  query_1   \n",
       "1  REAC:R-HSA-1483206  1.925997e-11  [REAC:R-HSA-1483257]   1.000000  query_1   \n",
       "2  REAC:R-HSA-1483257  2.623999e-10   [REAC:R-HSA-556833]   1.000000  query_1   \n",
       "5   REAC:R-HSA-425407  9.823245e-06   [REAC:R-HSA-382551]   0.666667  query_1   \n",
       "4   REAC:R-HSA-556833  2.973303e-07  [REAC:R-HSA-1430728]   1.000000  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "3           6  0.047619         True   REAC         84  \n",
       "1           6  0.046875         True   REAC        128  \n",
       "2           6  0.028571         True   REAC        210  \n",
       "5           6  0.016327         True   REAC        245  \n",
       "4           6  0.008197         True   REAC        732  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C775\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clathrin-mediated endocytosis</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Clathrin-mediated endocytosis</td>\n",
       "      <td>REAC:R-HSA-8856828</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>[REAC:R-HSA-199991]</td>\n",
       "      <td>0.8</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     description  effective_domain_size  intersection_size  \\\n",
       "0  Clathrin-mediated endocytosis                  10842                  4   \n",
       "\n",
       "                            name              native   p_value  \\\n",
       "0  Clathrin-mediated endocytosis  REAC:R-HSA-8856828  0.000003   \n",
       "\n",
       "               parents  precision    query  query_size    recall  significant  \\\n",
       "0  [REAC:R-HSA-199991]        0.8  query_1           5  0.027778         True   \n",
       "\n",
       "  source  term_size  \n",
       "0   REAC        144  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C806\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C781\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C752\n",
      "99\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>18</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>1.394247e-20</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.243243</td>\n",
       "      <td>query_1</td>\n",
       "      <td>74</td>\n",
       "      <td>0.209302</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Synthesis of substrates in N-glycan biosythesis</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Synthesis of substrates in N-glycan biosythesis</td>\n",
       "      <td>REAC:R-HSA-446219</td>\n",
       "      <td>1.838320e-07</td>\n",
       "      <td>[REAC:R-HSA-446193]</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>query_1</td>\n",
       "      <td>74</td>\n",
       "      <td>0.126984</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Biosynthesis of the N-glycan precursor (dolich...</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Biosynthesis of the N-glycan precursor (dolich...</td>\n",
       "      <td>REAC:R-HSA-446193</td>\n",
       "      <td>7.911541e-07</td>\n",
       "      <td>[REAC:R-HSA-446203]</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>query_1</td>\n",
       "      <td>74</td>\n",
       "      <td>0.102564</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Glycosaminoglycan metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>10</td>\n",
       "      <td>Glycosaminoglycan metabolism</td>\n",
       "      <td>REAC:R-HSA-1630316</td>\n",
       "      <td>1.838320e-07</td>\n",
       "      <td>[REAC:R-HSA-71387]</td>\n",
       "      <td>0.135135</td>\n",
       "      <td>query_1</td>\n",
       "      <td>74</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neutrophil degranulation</td>\n",
       "      <td>10842</td>\n",
       "      <td>29</td>\n",
       "      <td>Neutrophil degranulation</td>\n",
       "      <td>REAC:R-HSA-6798695</td>\n",
       "      <td>7.185903e-19</td>\n",
       "      <td>[REAC:R-HSA-168249]</td>\n",
       "      <td>0.391892</td>\n",
       "      <td>query_1</td>\n",
       "      <td>74</td>\n",
       "      <td>0.060924</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "1                             Sphingolipid metabolism                  10842   \n",
       "10    Synthesis of substrates in N-glycan biosythesis                  10842   \n",
       "14  Biosynthesis of the N-glycan precursor (dolich...                  10842   \n",
       "11                       Glycosaminoglycan metabolism                  10842   \n",
       "2                            Neutrophil degranulation                  10842   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "1                  18                            Sphingolipid metabolism   \n",
       "10                  8    Synthesis of substrates in N-glycan biosythesis   \n",
       "14                  8  Biosynthesis of the N-glycan precursor (dolich...   \n",
       "11                 10                       Glycosaminoglycan metabolism   \n",
       "2                  29                           Neutrophil degranulation   \n",
       "\n",
       "                native       p_value              parents  precision    query  \\\n",
       "1    REAC:R-HSA-428157  1.394247e-20  [REAC:R-HSA-556833]   0.243243  query_1   \n",
       "10   REAC:R-HSA-446219  1.838320e-07  [REAC:R-HSA-446193]   0.108108  query_1   \n",
       "14   REAC:R-HSA-446193  7.911541e-07  [REAC:R-HSA-446203]   0.108108  query_1   \n",
       "11  REAC:R-HSA-1630316  1.838320e-07   [REAC:R-HSA-71387]   0.135135  query_1   \n",
       "2   REAC:R-HSA-6798695  7.185903e-19  [REAC:R-HSA-168249]   0.391892  query_1   \n",
       "\n",
       "    query_size    recall  significant source  term_size  \n",
       "1           74  0.209302         True   REAC         86  \n",
       "10          74  0.126984         True   REAC         63  \n",
       "14          74  0.102564         True   REAC         78  \n",
       "11          74  0.081301         True   REAC        123  \n",
       "2           74  0.060924         True   REAC        476  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C756\n",
      "19\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>14</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>2.876732e-28</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.162791</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>10842</td>\n",
       "      <td>14</td>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>REAC:R-HSA-556833</td>\n",
       "      <td>2.560121e-15</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>query_1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.019126</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>732</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               description  effective_domain_size  intersection_size  \\\n",
       "0  Sphingolipid metabolism                  10842                 14   \n",
       "2     Metabolism of lipids                  10842                 14   \n",
       "\n",
       "                      name             native       p_value  \\\n",
       "0  Sphingolipid metabolism  REAC:R-HSA-428157  2.876732e-28   \n",
       "2     Metabolism of lipids  REAC:R-HSA-556833  2.560121e-15   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0   [REAC:R-HSA-556833]   0.933333  query_1          15  0.162791   \n",
       "2  [REAC:R-HSA-1430728]   0.933333  query_1          15  0.019126   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         86  \n",
       "2         True   REAC        732  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C791\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C753\n",
      "56\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>16</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>1.054031e-21</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.186047</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Synthesis of substrates in N-glycan biosythesis</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Synthesis of substrates in N-glycan biosythesis</td>\n",
       "      <td>REAC:R-HSA-446219</td>\n",
       "      <td>1.321256e-09</td>\n",
       "      <td>[REAC:R-HSA-446193]</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.126984</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Heparan sulfate/heparin (HS-GAG) metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Heparan sulfate/heparin (HS-GAG) metabolism</td>\n",
       "      <td>REAC:R-HSA-1638091</td>\n",
       "      <td>4.413757e-07</td>\n",
       "      <td>[REAC:R-HSA-1630316]</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.109091</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Biosynthesis of the N-glycan precursor (dolich...</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Biosynthesis of the N-glycan precursor (dolich...</td>\n",
       "      <td>REAC:R-HSA-446193</td>\n",
       "      <td>5.464923e-09</td>\n",
       "      <td>[REAC:R-HSA-446203]</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.102564</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Glycosaminoglycan metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>10</td>\n",
       "      <td>Glycosaminoglycan metabolism</td>\n",
       "      <td>REAC:R-HSA-1630316</td>\n",
       "      <td>5.846141e-10</td>\n",
       "      <td>[REAC:R-HSA-71387]</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>query_1</td>\n",
       "      <td>44</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          description  effective_domain_size  \\\n",
       "1                             Sphingolipid metabolism                  10842   \n",
       "9     Synthesis of substrates in N-glycan biosythesis                  10842   \n",
       "18        Heparan sulfate/heparin (HS-GAG) metabolism                  10842   \n",
       "13  Biosynthesis of the N-glycan precursor (dolich...                  10842   \n",
       "7                        Glycosaminoglycan metabolism                  10842   \n",
       "\n",
       "    intersection_size                                               name  \\\n",
       "1                  16                            Sphingolipid metabolism   \n",
       "9                   8    Synthesis of substrates in N-glycan biosythesis   \n",
       "18                  6        Heparan sulfate/heparin (HS-GAG) metabolism   \n",
       "13                  8  Biosynthesis of the N-glycan precursor (dolich...   \n",
       "7                  10                       Glycosaminoglycan metabolism   \n",
       "\n",
       "                native       p_value               parents  precision  \\\n",
       "1    REAC:R-HSA-428157  1.054031e-21   [REAC:R-HSA-556833]   0.363636   \n",
       "9    REAC:R-HSA-446219  1.321256e-09   [REAC:R-HSA-446193]   0.181818   \n",
       "18  REAC:R-HSA-1638091  4.413757e-07  [REAC:R-HSA-1630316]   0.136364   \n",
       "13   REAC:R-HSA-446193  5.464923e-09   [REAC:R-HSA-446203]   0.181818   \n",
       "7   REAC:R-HSA-1630316  5.846141e-10    [REAC:R-HSA-71387]   0.227273   \n",
       "\n",
       "      query  query_size    recall  significant source  term_size  \n",
       "1   query_1          44  0.186047         True   REAC         86  \n",
       "9   query_1          44  0.126984         True   REAC         63  \n",
       "18  query_1          44  0.109091         True   REAC         55  \n",
       "13  query_1          44  0.102564         True   REAC         78  \n",
       "7   query_1          44  0.081301         True   REAC        123  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C773\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C811\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C779\n",
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Synthesis of substrates in N-glycan biosythesis</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Synthesis of substrates in N-glycan biosythesis</td>\n",
       "      <td>REAC:R-HSA-446219</td>\n",
       "      <td>1.906744e-07</td>\n",
       "      <td>[REAC:R-HSA-446193]</td>\n",
       "      <td>0.8</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.063492</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Biosynthesis of the N-glycan precursor (dolich...</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Biosynthesis of the N-glycan precursor (dolich...</td>\n",
       "      <td>REAC:R-HSA-446193</td>\n",
       "      <td>2.280478e-07</td>\n",
       "      <td>[REAC:R-HSA-446203]</td>\n",
       "      <td>0.8</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "0    Synthesis of substrates in N-glycan biosythesis                  10842   \n",
       "1  Biosynthesis of the N-glycan precursor (dolich...                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "0                  4    Synthesis of substrates in N-glycan biosythesis   \n",
       "1                  4  Biosynthesis of the N-glycan precursor (dolich...   \n",
       "\n",
       "              native       p_value              parents  precision    query  \\\n",
       "0  REAC:R-HSA-446219  1.906744e-07  [REAC:R-HSA-446193]        0.8  query_1   \n",
       "1  REAC:R-HSA-446193  2.280478e-07  [REAC:R-HSA-446203]        0.8  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           5  0.063492         True   REAC         63  \n",
       "1           5  0.051282         True   REAC         78  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C757\n",
      "16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>4.335902e-14</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.093023</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>10842</td>\n",
       "      <td>8</td>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>REAC:R-HSA-556833</td>\n",
       "      <td>7.775659e-07</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.010929</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>732</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               description  effective_domain_size  intersection_size  \\\n",
       "3  Sphingolipid metabolism                  10842                  8   \n",
       "5     Metabolism of lipids                  10842                  8   \n",
       "\n",
       "                      name             native       p_value  \\\n",
       "3  Sphingolipid metabolism  REAC:R-HSA-428157  4.335902e-14   \n",
       "5     Metabolism of lipids  REAC:R-HSA-556833  7.775659e-07   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "3   [REAC:R-HSA-556833]   0.666667  query_1          12  0.093023   \n",
       "5  [REAC:R-HSA-1430728]   0.666667  query_1          12  0.010929   \n",
       "\n",
       "   significant source  term_size  \n",
       "3         True   REAC         86  \n",
       "5         True   REAC        732  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C768\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>Respiratory electron transport, ATP synthesis ...</td>\n",
       "      <td>REAC:R-HSA-163200</td>\n",
       "      <td>3.872739e-09</td>\n",
       "      <td>[REAC:R-HSA-1428517]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.039370</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Respiratory electron transport</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Respiratory electron transport</td>\n",
       "      <td>REAC:R-HSA-611105</td>\n",
       "      <td>2.415342e-07</td>\n",
       "      <td>[REAC:R-HSA-163200]</td>\n",
       "      <td>0.8</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>The citric acid (TCA) cycle and respiratory el...</td>\n",
       "      <td>REAC:R-HSA-1428517</td>\n",
       "      <td>1.041589e-08</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.028249</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         description  effective_domain_size  \\\n",
       "0  Respiratory electron transport, ATP synthesis ...                  10842   \n",
       "2                     Respiratory electron transport                  10842   \n",
       "1  The citric acid (TCA) cycle and respiratory el...                  10842   \n",
       "\n",
       "   intersection_size                                               name  \\\n",
       "0                  5  Respiratory electron transport, ATP synthesis ...   \n",
       "2                  4                     Respiratory electron transport   \n",
       "1                  5  The citric acid (TCA) cycle and respiratory el...   \n",
       "\n",
       "               native       p_value               parents  precision    query  \\\n",
       "0   REAC:R-HSA-163200  3.872739e-09  [REAC:R-HSA-1428517]        1.0  query_1   \n",
       "2   REAC:R-HSA-611105  2.415342e-07   [REAC:R-HSA-163200]        0.8  query_1   \n",
       "1  REAC:R-HSA-1428517  1.041589e-08  [REAC:R-HSA-1430728]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "0           5  0.039370         True   REAC        127  \n",
       "2           5  0.038835         True   REAC        103  \n",
       "1           5  0.028249         True   REAC        177  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C760\n",
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Peroxisomal protein import</td>\n",
       "      <td>REAC:R-HSA-9033241</td>\n",
       "      <td>1.665732e-11</td>\n",
       "      <td>[REAC:R-HSA-9609507]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Protein localization</td>\n",
       "      <td>10842</td>\n",
       "      <td>6</td>\n",
       "      <td>Protein localization</td>\n",
       "      <td>REAC:R-HSA-9609507</td>\n",
       "      <td>3.214813e-09</td>\n",
       "      <td>[REAC:0000000]</td>\n",
       "      <td>0.75</td>\n",
       "      <td>query_1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.037975</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  description  effective_domain_size  intersection_size  \\\n",
       "0  Peroxisomal protein import                  10842                  6   \n",
       "1        Protein localization                  10842                  6   \n",
       "\n",
       "                         name              native       p_value  \\\n",
       "0  Peroxisomal protein import  REAC:R-HSA-9033241  1.665732e-11   \n",
       "1        Protein localization  REAC:R-HSA-9609507  3.214813e-09   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0  [REAC:R-HSA-9609507]       0.75  query_1           8  0.100000   \n",
       "1        [REAC:0000000]       0.75  query_1           8  0.037975   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         60  \n",
       "1         True   REAC        158  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C758\n",
      "12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>10842</td>\n",
       "      <td>10</td>\n",
       "      <td>Sphingolipid metabolism</td>\n",
       "      <td>REAC:R-HSA-428157</td>\n",
       "      <td>7.473576e-21</td>\n",
       "      <td>[REAC:R-HSA-556833]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.116279</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>10842</td>\n",
       "      <td>10</td>\n",
       "      <td>Metabolism of lipids</td>\n",
       "      <td>REAC:R-HSA-556833</td>\n",
       "      <td>1.207587e-11</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.013661</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>732</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               description  effective_domain_size  intersection_size  \\\n",
       "0  Sphingolipid metabolism                  10842                 10   \n",
       "1     Metabolism of lipids                  10842                 10   \n",
       "\n",
       "                      name             native       p_value  \\\n",
       "0  Sphingolipid metabolism  REAC:R-HSA-428157  7.473576e-21   \n",
       "1     Metabolism of lipids  REAC:R-HSA-556833  1.207587e-11   \n",
       "\n",
       "                parents  precision    query  query_size    recall  \\\n",
       "0   [REAC:R-HSA-556833]        1.0  query_1          10  0.116279   \n",
       "1  [REAC:R-HSA-1430728]        1.0  query_1          10  0.013661   \n",
       "\n",
       "   significant source  term_size  \n",
       "0         True   REAC         86  \n",
       "1         True   REAC        732  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C770\n",
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C794\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C787\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C795\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C790\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C782\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C789\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>10842</td>\n",
       "      <td>5</td>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>REAC:R-HSA-71291</td>\n",
       "      <td>1.127241e-07</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.013514</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 description  effective_domain_size  \\\n",
       "1  Metabolism of amino acids and derivatives                  10842   \n",
       "\n",
       "   intersection_size                                       name  \\\n",
       "1                  5  Metabolism of amino acids and derivatives   \n",
       "\n",
       "             native       p_value               parents  precision    query  \\\n",
       "1  REAC:R-HSA-71291  1.127241e-07  [REAC:R-HSA-1430728]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1           5  0.013514         True   REAC        370  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C783\n",
      "5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C793\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>10842</td>\n",
       "      <td>4</td>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "      <td>REAC:R-HSA-71291</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>[REAC:R-HSA-1430728]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.010811</td>\n",
       "      <td>True</td>\n",
       "      <td>REAC</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 description  effective_domain_size  \\\n",
       "1  Metabolism of amino acids and derivatives                  10842   \n",
       "\n",
       "   intersection_size                                       name  \\\n",
       "1                  4  Metabolism of amino acids and derivatives   \n",
       "\n",
       "             native   p_value               parents  precision    query  \\\n",
       "1  REAC:R-HSA-71291  0.000003  [REAC:R-HSA-1430728]        1.0  query_1   \n",
       "\n",
       "   query_size    recall  significant source  term_size  \n",
       "1           4  0.010811         True   REAC        370  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C796\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>name</th>\n",
       "      <th>native</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parents</th>\n",
       "      <th>precision</th>\n",
       "      <th>query</th>\n",
       "      <th>query_size</th>\n",
       "      <th>recall</th>\n",
       "      <th>significant</th>\n",
       "      <th>source</th>\n",
       "      <th>term_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description, effective_domain_size, intersection_size, name, native, p_value, parents, precision, query, query_size, recall, significant, source, term_size]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# gprofiler annotation of clusters\n",
    "\n",
    "system_name_list = []\n",
    "for p in hier_df.index.tolist():\n",
    "    focal_genes=hier_df['CD_MemberList'].loc[p].split(' ')\n",
    "    print(p)\n",
    "    print(len(focal_genes))\n",
    "    if len(focal_genes)>2:\n",
    "        gp_temp = pd.DataFrame(gp.profile(focal_genes,significance_threshold_method='fdr',\n",
    "                                               sources=['REAC']))\n",
    "        if len(gp_temp)>0: # make sure data is not empty\n",
    "            \n",
    "            # make sure terms are specific, and overlap with at least 3 genes\n",
    "            gp_temp = gp_temp[(gp_temp['term_size']<1000)&(gp_temp['term_size']>50)]\n",
    "            gp_temp = gp_temp[gp_temp['intersection_size']>=3]\n",
    "            \n",
    "            gp_temp = gp_temp[gp_temp['p_value']<1E-5] # set a stringent pvalue threshold\n",
    "            \n",
    "            gp_temp = gp_temp.sort_values('recall',ascending=False)\n",
    "            \n",
    "            if len(gp_temp)>1:\n",
    "                system_name_list.append(gp_temp.head(1)['name'].tolist()[0])\n",
    "            else:\n",
    "                system_name_list.append(p)\n",
    "        else:\n",
    "            system_name_list.append(p)\n",
    "            \n",
    "\n",
    "        display(gp_temp.head())\n",
    "        \n",
    "    else:\n",
    "        system_name_list.append(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _Validate identified genes and systems_\n",
    "\n",
    "## 15. Load and parse mouse variant database\n",
    "\n",
    "Requires DDOT for ontology parsing https://github.com/michaelkyu/ddot/blob/master/examples/Tutorial.ipynb\n",
    "\n",
    "Parse the ontology, data from http://www.informatics.jax.org/vocab/mp_ontology\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:29:19.110315Z",
     "start_time": "2022-06-24T21:28:31.080464Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-1000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 1001-2000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 2001-3000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 3001-4000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 4001-5000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 5001-6000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 6001-7000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 7001-8000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 8001-9000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 9001-10000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 10001-11000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 11001-12000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 12001-13000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 13001-14000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 14001-15000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 15001-16000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 16001-17000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 17001-18000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 18001-19000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 19001-20000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 20001-21000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 21001-22000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 22001-23000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 23001-24000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 24001-25000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 25001-25356...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:Finished.\n",
      "WARNING:biothings.client:29 input query terms found dup hits:\t[('Adam1a', 2), ('Betalinc1', 2), ('Chaserr', 2), ('Dnajb3', 2), ('Dnm3os', 2), ('Eldr', 2), ('Ggnbp\n",
      "WARNING:biothings.client:12568 input query terms found no hit:\t['+|Cav1', '+|Cm', '+|Del(10Prmt2-Pdxk)4Yey', '+|Del(11Mpo-Chad)18Brd', '+|Del(11Rr325)1Smun', '+|De\n",
      "INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25385\n",
      "12817\n",
      "12788\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI_Allele_Accession_ID</th>\n",
       "      <th>Allele symbol</th>\n",
       "      <th>involves</th>\n",
       "      <th>MP</th>\n",
       "      <th>PMID</th>\n",
       "      <th>MGI_marker_accession_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>human_ortholog</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000600</td>\n",
       "      <td>12529408</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001716</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001698</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001092</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rb1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000961</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           MGI_Allele_Accession_ID Allele symbol               involves  \\\n",
       "gene_name                                                                 \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "Rb1        Rb1<tm1Tyj>/Rb1<tm1Tyj>   Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "\n",
       "                   MP      PMID MGI_marker_accession_ID gene_name  \\\n",
       "gene_name                                                           \n",
       "Rb1        MP:0000600  12529408               MGI:97874       Rb1   \n",
       "Rb1        MP:0001716  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0001698  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0001092  16449662               MGI:97874       Rb1   \n",
       "Rb1        MP:0000961  16449662               MGI:97874       Rb1   \n",
       "\n",
       "          human_ortholog  \n",
       "gene_name                 \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  \n",
       "Rb1                  RB1  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_df = validation.load_MGI_mouseKO_data()\n",
    "mgi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:29:20.721839Z",
     "start_time": "2022-06-24T21:29:19.111386Z"
    }
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "ddot package is required to use this method",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[47], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m MPO \u001b[38;5;241m=\u001b[39m \u001b[43mvalidation\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload_MPO\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m MPO\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.11/site-packages/netcoloc/validation.py:86\u001b[0m, in \u001b[0;36mload_MPO\u001b[0;34m(url)\u001b[0m\n\u001b[1;32m     84\u001b[0m \u001b[38;5;28mopen\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMPheno_OBO.ontology\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mwrite(r\u001b[38;5;241m.\u001b[39mcontent)\n\u001b[1;32m     85\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m DDOT_LOADED \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m:\n\u001b[0;32m---> 86\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mddot package is required to use this method\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     87\u001b[0m ddot\u001b[38;5;241m.\u001b[39mparse_obo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMPheno_OBO.ontology\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     88\u001b[0m                \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mparsed_mp.txt\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     89\u001b[0m               \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mid2name_mp.txt\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     90\u001b[0m               \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mid2namespace_mp.txt\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     91\u001b[0m               \u001b[38;5;124m'\u001b[39m\u001b[38;5;124maltID_mp.txt\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     94\u001b[0m MP2desc \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mid2name_mp.txt\u001b[39m\u001b[38;5;124m'\u001b[39m,sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     95\u001b[0m                       names\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMP\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdescription\u001b[39m\u001b[38;5;124m'\u001b[39m],index_col\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMP\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mImportError\u001b[0m: ddot package is required to use this method"
     ]
    }
   ],
   "source": [
    "MPO = validation.load_MPO()\n",
    "\n",
    "MPO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 16. Identify phenotype(s) of interest. \n",
    "\n",
    "Recommend including a negative control, a phenotype that is not expected to overlap with the two phenotypes of interest.\n",
    "\n",
    "### Here we find MPO terms related to brain or heart\n",
    "\n",
    "Modify as needed for each specific project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:29:20.726414Z",
     "start_time": "2022-06-24T21:29:20.723018Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'MPO' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[37], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_MP_description\u001b[39m(term_id, ontology\u001b[38;5;241m=\u001b[39m\u001b[43mMPO\u001b[49m):\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ontology\u001b[38;5;241m.\u001b[39mnode_attr\u001b[38;5;241m.\u001b[39mloc[term_id, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdescription\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'MPO' is not defined"
     ]
    }
   ],
   "source": [
    "def get_MP_description(term_id, ontology=MPO):\n",
    "    return ontology.node_attr.loc[term_id, \"description\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:29:21.532067Z",
     "start_time": "2022-06-24T21:29:20.727873Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of brain phenotypes 200\n",
      "Example brain phenotypes:\n",
      "MP:0000778 - abnormal nervous system tract morphology\n",
      "MP:0000811 - hippocampal neuron degeneration\n",
      "MP:0000937 - abnormal motor neuron morphology\n",
      "MP:0000938 - motor neuron degeneration\n",
      "MP:0000939 - decreased motor neuron number\n",
      "MP:0000940 - abnormal motor neuron innervation pattern\n",
      "MP:0000958 - peripheral nervous system degeneration\n",
      "MP:0000965 - abnormal sensory neuron morphology\n",
      "MP:0000966 - decreased sensory neuron number\n",
      "MP:0000968 - abnormal sensory neuron innervation pattern\n"
     ]
    }
   ],
   "source": [
    "# find terms related to brain\n",
    "# ---- modify this part as needed for your project -----\n",
    "MP_focal_brain_list = []\n",
    "for t in MPO.node_attr.index.tolist():\n",
    "    descr_temp = MPO.node_attr.loc[t]['description']\n",
    "    if descr_temp.find('nervous')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "    elif descr_temp.find('neuron')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "    elif descr_temp.find('synapt')>-1:\n",
    "        MP_focal_brain_list.append(t)\n",
    "        \n",
    "print(\"Number of brain phenotypes\", len(MP_focal_brain_list))\n",
    "print(\"Example brain phenotypes:\")\n",
    "print(\"\\n\".join([mp+\" - \"+get_MP_description(mp) for mp in MP_focal_brain_list[0:10]]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 17. Compute the enrichment of selected phenotype(s) in NetColoc subnetwork as a whole.\n",
    "\n",
    "By computing the enrichment in the entire NetColoc subnetwork, we identify the phenotypes with the strongest association with the entire set of genes identified to be related to both input sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:30:18.381323Z",
     "start_time": "2022-06-24T21:29:22.345431Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal innate immunity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 606\n",
      "number overlapping genes = 9\n",
      "0.14110702011931123\n",
      "(-1.1654239055794087, 0.1658261748879888)\n",
      "-0.49979886534570994\n",
      "\n",
      "abnormal nervous system tract morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 215\n",
      "number overlapping genes = 21\n",
      "8.846118167068254e-11\n",
      "(1.0651571336135097, 1.9877828028378683)\n",
      "1.526469968225689\n",
      "\n",
      "hippocampal neuron degeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 4\n",
      "0.0002267526170912075\n",
      "(0.942831168199129, 3.0825921426594665)\n",
      "2.012711655429298\n",
      "\n",
      "abnormal motor neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 167\n",
      "number overlapping genes = 8\n",
      "0.04609268201409096\n",
      "(0.01265022122823889, 1.446673695374994)\n",
      "0.7296619583016164\n",
      "\n",
      "motor neuron degeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 44\n",
      "number overlapping genes = 3\n",
      "0.06711206560651001\n",
      "(-0.07744989967973681, 2.2752721877881648)\n",
      "1.098911144054214\n",
      "\n",
      "decreased motor neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 77\n",
      "number overlapping genes = 4\n",
      "0.11633072156452744\n",
      "(-0.20101890209091322, 1.8215101248519596)\n",
      "0.8102456113805232\n",
      "\n",
      "abnormal motor neuron innervation pattern\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 41\n",
      "number overlapping genes = 3\n",
      "0.05086560550706154\n",
      "(-0.004424491688665988, 2.3545771344930904)\n",
      "1.1750763214022122\n",
      "\n",
      "peripheral nervous system degeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.6696387209853685\n",
      "(-2.218498432362735, 3.4530858079161892)\n",
      "0.6172936877767272\n",
      "\n",
      "abnormal sensory neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 494\n",
      "number overlapping genes = 18\n",
      "0.06682061529652303\n",
      "(-0.03119738047360937, 0.931081323824906)\n",
      "0.4499419716756483\n",
      "\n",
      "decreased sensory neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 204\n",
      "number overlapping genes = 6\n",
      "0.6059826488045965\n",
      "(-0.6028285047081752, 1.0334615376805567)\n",
      "0.21531651648619077\n",
      "\n",
      "abnormal sensory neuron innervation pattern\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 98\n",
      "number overlapping genes = 4\n",
      "0.2783003306798568\n",
      "(-0.44928319235519876, 1.56158733099526)\n",
      "0.5561520693200306\n",
      "\n",
      "abnormal sympathetic neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.2394509045294989\n",
      "(-0.8165405499709597, 3.268152328315956)\n",
      "1.225805889172498\n",
      "\n",
      "abnormal cholinergic neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.8660990604072069\n",
      "(-2.5741825333772765, 3.0587863691318202)\n",
      "0.24230191787727184\n",
      "\n",
      "abnormal enteric nervous system morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 35\n",
      "number overlapping genes = 1\n",
      "0.8570063009352193\n",
      "(-1.8079514942497954, 2.174028813707698)\n",
      "0.18303865972895128\n",
      "\n",
      "abnormal enteric neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.6026655635990128\n",
      "(-1.4708351133117534, 2.5347202077255795)\n",
      "0.531942547206913\n",
      "\n",
      "impaired synaptic plasticity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 6\n",
      "1.750942822155272e-05\n",
      "(1.034126013622632, 2.7705371983592273)\n",
      "1.9023316059909297\n",
      "\n",
      "abnormal synaptic neurotransmitter level\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.7135224554861573\n",
      "(-2.3004336307867996, 3.3608789384161444)\n",
      "0.5302226538146724\n",
      "\n",
      "abnormal CNS synaptic transmission\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 866\n",
      "number overlapping genes = 51\n",
      "3.33202314852535e-11\n",
      "(0.7187365102774576, 1.3218868905875292)\n",
      "1.0203117004324933\n",
      "\n",
      "abnormal nervous system electrophysiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 335\n",
      "number overlapping genes = 27\n",
      "1.665167192265028e-10\n",
      "(0.9179632427698274, 1.7303316738329715)\n",
      "1.3241474583013995\n",
      "\n",
      "abnormal autonomic nervous system morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 94\n",
      "number overlapping genes = 6\n",
      "0.015069577908850872\n",
      "(0.20002756528637988, 1.8656942569242778)\n",
      "1.0328609111053288\n",
      "\n",
      "abnormal somatic nervous system morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 1040\n",
      "number overlapping genes = 39\n",
      "0.0034119975650116438\n",
      "(0.1659521147300705, 0.837983703327436)\n",
      "0.5019679090287532\n",
      "\n",
      "abnormal neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 1555\n",
      "number overlapping genes = 72\n",
      "3.3754078609776067e-09\n",
      "(0.524004012698436, 1.0437227243132126)\n",
      "0.7838633685058243\n",
      "\n",
      "abnormal AMPA-mediated synaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 30\n",
      "number overlapping genes = 5\n",
      "1.8259576816709434e-05\n",
      "(1.145087142136498, 3.0755183817487772)\n",
      "2.1103027619426378\n",
      "\n",
      "abnormal NMDA-mediated synaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 2\n",
      "0.0961035732390383\n",
      "(-0.2182939341482235, 2.673357396993044)\n",
      "1.2275317314224101\n",
      "\n",
      "abnormal excitatory postsynaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 139\n",
      "number overlapping genes = 15\n",
      "4.928708243556163e-09\n",
      "(1.081535394030102, 2.1715044452724634)\n",
      "1.6265199196512827\n",
      "\n",
      "abnormal excitatory postsynaptic potential\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 97\n",
      "number overlapping genes = 6\n",
      "0.018581043382423\n",
      "(0.16718947252254057, 1.8311273503981402)\n",
      "0.9991584114603405\n",
      "\n",
      "abnormal PNS synaptic transmission\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 39\n",
      "number overlapping genes = 0\n",
      "0.6480652698827825\n",
      "(-3.4412794135337266, 2.1411905584529287)\n",
      "-0.6500444275403989\n",
      "\n",
      "abnormal synaptic depression\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 116\n",
      "number overlapping genes = 12\n",
      "3.591003647348888e-07\n",
      "(0.9671837081476342, 2.1786569108555742)\n",
      "1.5729203095016042\n",
      "\n",
      "increased synaptic depression\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 0\n",
      "0.7940477779774581\n",
      "(-2.446614156979289, 3.198519455708561)\n",
      "0.3759526493646361\n",
      "\n",
      "decreased synaptic depression\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.7549294929290006\n",
      "(-2.3761830029985522, 3.276423574217636)\n",
      "0.45012028560954187\n",
      "\n",
      "abnormal inhibitory postsynaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 81\n",
      "number overlapping genes = 4\n",
      "0.1419763290694368\n",
      "(-0.25325194205044355, 1.7665721327955808)\n",
      "0.7566600953725686\n",
      "\n",
      "abnormal neuron specification\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 1\n",
      "0.3322880841201836\n",
      "(-1.0241282858903165, 3.029094977058936)\n",
      "1.0024833455843098\n",
      "\n",
      "abnormal neuron apoptosis\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 235\n",
      "number overlapping genes = 11\n",
      "0.02349142588957176\n",
      "(0.09556754752058871, 1.3222845617605286)\n",
      "0.7089260546405587\n",
      "\n",
      "increased neuron apoptosis\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 169\n",
      "number overlapping genes = 8\n",
      "0.04992366791515487\n",
      "(0.00023897844770026655, 1.4338438518226386)\n",
      "0.7170414151351694\n",
      "\n",
      "decreased neuron apoptosis\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 71\n",
      "number overlapping genes = 4\n",
      "0.08304280604386703\n",
      "(-0.11722016123590806, 1.9099634183241252)\n",
      "0.8963716285441086\n",
      "\n",
      "neuron degeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 413\n",
      "number overlapping genes = 15\n",
      "0.09762453516312651\n",
      "(-0.0813004502533019, 0.9688126740621366)\n",
      "0.4437561119044173\n",
      "\n",
      "loss of hippocampal neurons\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 3\n",
      "0.00032101170433086366\n",
      "(1.0603509696717768, 3.5982653565978233)\n",
      "2.3293081631348\n",
      "\n",
      "loss of cortex neurons\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 3\n",
      "0.0012189041567640508\n",
      "(0.8043659959172533, 3.2784088584775457)\n",
      "2.0413874271973995\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal dopaminergic neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 2\n",
      "0.392125087479101\n",
      "(-0.8002414321713913, 2.0406438013308947)\n",
      "0.6202011845797517\n",
      "\n",
      "loss of dopaminergic neurons\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.8660990604072069\n",
      "(-2.5741825333772765, 3.0587863691318202)\n",
      "0.24230191787727184\n",
      "\n",
      "abnormal GABAergic neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 0\n",
      "0.8045523730056512\n",
      "(-3.1504938364139106, 2.4441313567852125)\n",
      "-0.353181239814349\n",
      "\n",
      "loss of GABAergic neurons\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.7135224554861573\n",
      "(-2.3004336307867996, 3.3608789384161444)\n",
      "0.5302226538146724\n",
      "\n",
      "abnormal glutaminergic neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 1\n",
      "0.17985935904760075\n",
      "(-0.6497141917265497, 3.4662077044316737)\n",
      "1.408246756352562\n",
      "\n",
      "abnormal central nervous system regeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.6696387209853685\n",
      "(-2.218498432362735, 3.4530858079161892)\n",
      "0.6172936877767272\n",
      "\n",
      "premature neuronal precursor differentiation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 3\n",
      "0.0021083493228830938\n",
      "(0.6972936783927481, 3.1496763984766716)\n",
      "1.9234850384347097\n",
      "\n",
      "abnormal synaptic transmission\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 956\n",
      "number overlapping genes = 55\n",
      "1.7486912480788608e-11\n",
      "(0.7094829213786688, 1.2930611353100208)\n",
      "1.0012720283443448\n",
      "\n",
      "abnormal nervous system development\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 1234\n",
      "number overlapping genes = 68\n",
      "7.653100275716135e-13\n",
      "(0.7090879506576702, 1.2428755961476838)\n",
      "0.975981773402677\n",
      "\n",
      "abnormal spinal cord interneuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 33\n",
      "number overlapping genes = 1\n",
      "0.8105082969576111\n",
      "(-1.7489799744872367, 2.2365454719310804)\n",
      "0.24378274872192174\n",
      "\n",
      "abnormal brain interneuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 42\n",
      "number overlapping genes = 6\n",
      "1.3727879379191806e-05\n",
      "(1.0599254430719038, 2.799655209444546)\n",
      "1.929790326258225\n",
      "\n",
      "abnormal sympathetic nervous system physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 29\n",
      "number overlapping genes = 0\n",
      "0.8045523730056512\n",
      "(-3.1504938364139106, 2.4441313567852125)\n",
      "-0.353181239814349\n",
      "\n",
      "neuronal intranuclear inclusions\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.2394509045294989\n",
      "(-0.8165405499709597, 3.268152328315956)\n",
      "1.225805889172498\n",
      "\n",
      "abnormal peripheral nervous system regeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.6696387209853685\n",
      "(-2.218498432362735, 3.4530858079161892)\n",
      "0.6172936877767272\n",
      "\n",
      "abnormal proprioceptive neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 16\n",
      "number overlapping genes = 0\n",
      "0.8660990604072069\n",
      "(-2.5741825333772765, 3.0587863691318202)\n",
      "0.24230191787727184\n",
      "\n",
      "abnormal synaptic glutamate release\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 3\n",
      "0.00021650554963521855\n",
      "(1.1360064756079873, 3.6967519185857234)\n",
      "2.4163791970968553\n",
      "\n",
      "abnormal miniature excitatory postsynaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 120\n",
      "number overlapping genes = 12\n",
      "6.496383234580731e-07\n",
      "(0.9303336456076765, 2.1395463558244616)\n",
      "1.5349400007160692\n",
      "\n",
      "abnormal synaptic vesicle morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 57\n",
      "number overlapping genes = 1\n",
      "0.7534550242803121\n",
      "(-2.2970795719749653, 1.6625443956261596)\n",
      "-0.3172675881744027\n",
      "\n",
      "abnormal synaptic vesicle recycling\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 20\n",
      "number overlapping genes = 0\n",
      "0.9894634209927645\n",
      "(-2.7890273768348304, 2.826866669041998)\n",
      "0.018919646103583876\n",
      "\n",
      "abnormal synaptic vesicle number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 49\n",
      "number overlapping genes = 1\n",
      "0.8722716294598836\n",
      "(-2.1453357237910162, 1.82005873046147)\n",
      "-0.16263849666477315\n",
      "\n",
      "abnormal neuron physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 640\n",
      "number overlapping genes = 37\n",
      "3.702989412117251e-08\n",
      "(0.6282968825909372, 1.3231425415699627)\n",
      "0.9757197120804499\n",
      "\n",
      "abnormal nervous system regeneration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 22\n",
      "number overlapping genes = 0\n",
      "0.957362519310701\n",
      "(-2.8813459772773506, 2.7283261466689948)\n",
      "-0.07650991530417794\n",
      "\n",
      "abnormal synaptic plasticity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 68\n",
      "number overlapping genes = 7\n",
      "0.00010778648256011189\n",
      "(0.7692152873629252, 2.3459118038566267)\n",
      "1.557563545609776\n",
      "\n",
      "abnormal neuronal precursor proliferation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 95\n",
      "number overlapping genes = 11\n",
      "1.6823907644820236e-07\n",
      "(1.0619486439015606, 2.3344092431043535)\n",
      "1.698178943502957\n",
      "\n",
      "abnormal neuronal precursor cell number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 69\n",
      "number overlapping genes = 7\n",
      "0.00012560510168537804\n",
      "(0.7535393943816573, 2.3289469841892876)\n",
      "1.5412431892854723\n",
      "\n",
      "increased neuronal precursor cell number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 21\n",
      "number overlapping genes = 2\n",
      "0.049823681565847995\n",
      "(0.001125544254639399, 2.921764546789589)\n",
      "1.4614450455221144\n",
      "\n",
      "decreased neuronal precursor cell number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 4\n",
      "0.009418188003101931\n",
      "(0.33460057537727494, 2.3951792045347595)\n",
      "1.3648898899560171\n",
      "\n",
      "increased motor neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.6696387209853685\n",
      "(-2.218498432362735, 3.4530858079161892)\n",
      "0.6172936877767272\n",
      "\n",
      "abnormal somatic nervous system physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 76\n",
      "number overlapping genes = 2\n",
      "0.8909012024465022\n",
      "(-1.3093978196730758, 1.5064596791087146)\n",
      "0.0985309297178194\n",
      "\n",
      "increased sensory neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 51\n",
      "number overlapping genes = 7\n",
      "4.26688566749744e-06\n",
      "(1.0816327429774983, 2.688896119336044)\n",
      "1.8852644311567712\n",
      "\n",
      "abnormal neuronal migration\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 129\n",
      "number overlapping genes = 9\n",
      "0.0011710438846382338\n",
      "(0.4491227630422332, 1.8182501424196011)\n",
      "1.1336864527309172\n",
      "\n",
      "abnormal retina neuronal layer morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 429\n",
      "number overlapping genes = 9\n",
      "0.6870055541652933\n",
      "(-0.8049114133008175, 0.5304025576242319)\n",
      "-0.13725442783829278\n",
      "\n",
      "abnormal olfactory sensory neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 28\n",
      "number overlapping genes = 1\n",
      "0.684719233538055\n",
      "(-1.5843523808004287, 2.41231316286402)\n",
      "0.41398039103179585\n",
      "\n",
      "abnormal autonomic nervous system physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 46\n",
      "number overlapping genes = 2\n",
      "0.392125087479101\n",
      "(-0.8002414321713913, 2.0406438013308947)\n",
      "0.6202011845797517\n",
      "\n",
      "abnormal parasympathetic nervous system physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.20928029732403952\n",
      "(-0.7365817674891977, 3.3623356137583036)\n",
      "1.312876923134553\n",
      "\n",
      "increased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 36\n",
      "number overlapping genes = 1\n",
      "0.8794602708365866\n",
      "(-1.8361879271758437, 2.144170695005549)\n",
      "0.1539913839148528\n",
      "\n",
      "decreased susceptibility to neuronal excitotoxicity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 31\n",
      "number overlapping genes = 2\n",
      "0.1566339260081966\n",
      "(-0.3982592290391864, 2.474241588557133)\n",
      "1.0379911797589734\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "abnormal synaptic bouton morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 26\n",
      "number overlapping genes = 1\n",
      "0.6305539129588807\n",
      "(-1.5101177882408057, 2.4922394870646385)\n",
      "0.4910608494119164\n",
      "\n",
      "abnormal neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 523\n",
      "number overlapping genes = 30\n",
      "9.619697871835968e-07\n",
      "(0.5733352318011014, 1.3379505586474554)\n",
      "0.9556428952242784\n",
      "\n",
      "increased neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 83\n",
      "number overlapping genes = 9\n",
      "5.561932559155366e-06\n",
      "(0.9209248589615724, 2.318817928874177)\n",
      "1.6198713939178746\n",
      "\n",
      "decreased neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 439\n",
      "number overlapping genes = 21\n",
      "0.001161013639274059\n",
      "(0.2955761774446567, 1.1949125176743562)\n",
      "0.7452443475595065\n",
      "\n",
      "abnormal miniature inhibitory postsynaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 50\n",
      "number overlapping genes = 5\n",
      "0.0013317643014288465\n",
      "(0.5921555998259623, 2.450486964960695)\n",
      "1.5213212823933286\n",
      "\n",
      "abnormal synaptonemal complex\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 1\n",
      "0.6026655635990128\n",
      "(-1.4708351133117534, 2.5347202077255795)\n",
      "0.531942547206913\n",
      "\n",
      "abnormal neuron differentiation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 328\n",
      "number overlapping genes = 26\n",
      "6.186102617109004e-10\n",
      "(0.8907096733641977, 1.7169771967687086)\n",
      "1.3038434350664532\n",
      "\n",
      "abnormal hippocampus neuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 114\n",
      "number overlapping genes = 13\n",
      "1.747217780971037e-08\n",
      "(1.0989079257457695, 2.2709375920468275)\n",
      "1.6849227588962985\n",
      "\n",
      "abnormal neuron proliferation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 126\n",
      "number overlapping genes = 13\n",
      "1.232358281222643e-07\n",
      "(0.989376877199215, 2.1544940655884397)\n",
      "1.5719354713938274\n",
      "\n",
      "abnormal sensory neuron physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 65\n",
      "number overlapping genes = 1\n",
      "0.6546996648512512\n",
      "(-2.4289229173548277, 1.5263676744176573)\n",
      "-0.4512776214685852\n",
      "\n",
      "abnormal synaptic dopamine release\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.7549294929290006\n",
      "(-2.3761830029985522, 3.276423574217636)\n",
      "0.45012028560954187\n",
      "\n",
      "increased nervous system tumor incidence\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 68\n",
      "number overlapping genes = 4\n",
      "0.068782309717835\n",
      "(-0.0725559638667892, 1.957277391680919)\n",
      "0.9423607139070649\n",
      "\n",
      "reduced AMPA-mediated synaptic currents\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 14\n",
      "number overlapping genes = 1\n",
      "0.27013619222211616\n",
      "(-0.8906054573375162, 3.1820124992722514)\n",
      "1.1457035209673676\n",
      "\n",
      "abnormal excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 36\n",
      "number overlapping genes = 1\n",
      "0.8794602708365866\n",
      "(-1.8361879271758437, 2.144170695005549)\n",
      "0.1539913839148528\n",
      "\n",
      "increased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 1\n",
      "0.2394509045294989\n",
      "(-0.8165405499709597, 3.268152328315956)\n",
      "1.225805889172498\n",
      "\n",
      "decreased excitatory postsynaptic current amplitude\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 24\n",
      "number overlapping genes = 0\n",
      "0.908877295922573\n",
      "(-2.9658816452323102, 2.638600268930485)\n",
      "-0.1636406881509127\n",
      "\n",
      "decreased dopaminergic neuron number\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 25\n",
      "number overlapping genes = 0\n",
      "0.8862057530609384\n",
      "(-3.005620735499072, 2.596575963607253)\n",
      "-0.20452238594590932\n",
      "\n",
      "increased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 13\n",
      "number overlapping genes = 0\n",
      "0.7549294929290006\n",
      "(-2.3761830029985522, 3.276423574217636)\n",
      "0.45012028560954187\n",
      "\n",
      "decreased susceptibility to dopaminergic neuron neurotoxicity\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 0\n",
      "0.7135224554861573\n",
      "(-2.3004336307867996, 3.3608789384161444)\n",
      "0.5302226538146724\n",
      "\n",
      "ectopic neuron\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 64\n",
      "number overlapping genes = 5\n",
      "0.007644560531087309\n",
      "(0.3314082702962303, 2.167809945265465)\n",
      "1.2496091077808478\n",
      "\n",
      "impaired neuron differentiation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 2\n",
      "0.015566260396467281\n",
      "(0.3493651758185117, 3.333220234547218)\n",
      "1.8412927051828647\n",
      "\n",
      "abnormal neuronal stem cell physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 3\n",
      "9.051201983871072e-05\n",
      "(1.3068411806748137, 3.927497210981303)\n",
      "2.6171691958280583\n",
      "\n",
      "nervous system inclusion bodies\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 63\n",
      "number overlapping genes = 0\n",
      "0.4259576721333993\n",
      "(-3.915557939950595, 1.6534526254619375)\n",
      "-1.1310526572443287\n",
      "\n",
      "abnormal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 15\n",
      "number overlapping genes = 0\n",
      "0.8310511337317399\n",
      "(-2.512424406195861, 3.1262246266488294)\n",
      "0.3069001102264841\n",
      "\n",
      "abnormal hippocampal pyramidal neuron dendrite morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.6696387209853685\n",
      "(-2.218498432362735, 3.4530858079161892)\n",
      "0.6172936877767272\n",
      "\n",
      "increased neuronal precursor proliferation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 12\n",
      "number overlapping genes = 1\n",
      "0.20928029732403952\n",
      "(-0.7365817674891977, 3.3623356137583036)\n",
      "1.312876923134553\n",
      "\n",
      "decreased neuronal precursor proliferation\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 3\n",
      "0.0009019351779285804\n",
      "(0.8625114194974768, 3.349459819595747)\n",
      "2.105985619546612\n",
      "\n",
      "abnormal postsynaptic density morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 19\n",
      "number overlapping genes = 0\n",
      "0.9609039130397335\n",
      "(-2.7394739056984654, 2.8800191575960485)\n",
      "0.07027262594879158\n",
      "\n",
      "abnormal synaptic physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 962\n",
      "number overlapping genes = 56\n",
      "6.133723962028307e-12\n",
      "(0.7265983759075835, 1.3059103777943672)\n",
      "1.0162543768509753\n",
      "\n",
      "impaired synaptic physiology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 43\n",
      "number overlapping genes = 6\n",
      "1.750942822155272e-05\n",
      "(1.034126013622632, 2.7705371983592273)\n",
      "1.9023316059909297\n",
      "\n",
      "abnormal ventral interneuron morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 18\n",
      "number overlapping genes = 0\n",
      "0.9308985730304116\n",
      "(-2.687345106188171, 2.9361441644173345)\n",
      "0.12439952911458185\n",
      "\n",
      "abnormal ventral interneuron 3 morphology\n",
      "number of genes in root node = 411\n",
      "number of genes in focal MPO term = 11\n",
      "number overlapping genes = 0\n",
      "0.6696387209853685\n",
      "(-2.218498432362735, 3.4530858079161892)\n",
      "0.6172936877767272\n"
     ]
    }
   ],
   "source": [
    "# add a negative control phenotype: abnormal innate immunity: MP:0002419\n",
    "# negative controls are tough here because we're dealing with development, which impacts almost everything.\n",
    "MP_focal_list = ['MP:0002419']+MP_focal_brain_list\n",
    "root_KO_brain_df=validation.MPO_enrichment_root(hier_df,MPO,mgi_df,MP_focal_list,G_int,verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:30:18.391599Z",
     "start_time": "2022-06-24T21:30:18.382481Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0002419</th>\n",
       "      <td>1.411070e-01</td>\n",
       "      <td>-0.499799</td>\n",
       "      <td>-1.165424</td>\n",
       "      <td>0.165826</td>\n",
       "      <td>606</td>\n",
       "      <td>abnormal innate immunity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000778</th>\n",
       "      <td>8.846118e-11</td>\n",
       "      <td>1.526470</td>\n",
       "      <td>1.065157</td>\n",
       "      <td>1.987783</td>\n",
       "      <td>215</td>\n",
       "      <td>abnormal nervous system tract morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000811</th>\n",
       "      <td>2.267526e-04</td>\n",
       "      <td>2.012712</td>\n",
       "      <td>0.942831</td>\n",
       "      <td>3.082592</td>\n",
       "      <td>26</td>\n",
       "      <td>hippocampal neuron degeneration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000937</th>\n",
       "      <td>4.609268e-02</td>\n",
       "      <td>0.729662</td>\n",
       "      <td>0.012650</td>\n",
       "      <td>1.446674</td>\n",
       "      <td>167</td>\n",
       "      <td>abnormal motor neuron morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000938</th>\n",
       "      <td>6.711207e-02</td>\n",
       "      <td>1.098911</td>\n",
       "      <td>-0.077450</td>\n",
       "      <td>2.275272</td>\n",
       "      <td>44</td>\n",
       "      <td>motor neuron degeneration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0002419  1.411070e-01 -0.499799        -1.165424         0.165826   \n",
       "MP:0000778  8.846118e-11  1.526470         1.065157         1.987783   \n",
       "MP:0000811  2.267526e-04  2.012712         0.942831         3.082592   \n",
       "MP:0000937  4.609268e-02  0.729662         0.012650         1.446674   \n",
       "MP:0000938  6.711207e-02  1.098911        -0.077450         2.275272   \n",
       "\n",
       "            num_genes_in_term                            MP_description  \n",
       "MP:0002419                606                  abnormal innate immunity  \n",
       "MP:0000778                215  abnormal nervous system tract morphology  \n",
       "MP:0000811                 26           hippocampal neuron degeneration  \n",
       "MP:0000937                167          abnormal motor neuron morphology  \n",
       "MP:0000938                 44                 motor neuron degeneration  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_brain_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:46:17.391341Z",
     "start_time": "2022-06-24T21:46:17.381335Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR_p</th>\n",
       "      <th>log_OR</th>\n",
       "      <th>log_OR_CI_lower</th>\n",
       "      <th>log_OR_CI_upper</th>\n",
       "      <th>num_genes_in_term</th>\n",
       "      <th>MP_description</th>\n",
       "      <th>MPO_term_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MP:0003861</th>\n",
       "      <td>7.653100e-13</td>\n",
       "      <td>0.975982</td>\n",
       "      <td>0.709088</td>\n",
       "      <td>1.242876</td>\n",
       "      <td>1234</td>\n",
       "      <td>abnormal nervous system development</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0021009</th>\n",
       "      <td>6.133724e-12</td>\n",
       "      <td>1.016254</td>\n",
       "      <td>0.726598</td>\n",
       "      <td>1.305910</td>\n",
       "      <td>962</td>\n",
       "      <td>abnormal synaptic physiology</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0003635</th>\n",
       "      <td>1.748691e-11</td>\n",
       "      <td>1.001272</td>\n",
       "      <td>0.709483</td>\n",
       "      <td>1.293061</td>\n",
       "      <td>956</td>\n",
       "      <td>abnormal synaptic transmission</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0002206</th>\n",
       "      <td>3.332023e-11</td>\n",
       "      <td>1.020312</td>\n",
       "      <td>0.718737</td>\n",
       "      <td>1.321887</td>\n",
       "      <td>866</td>\n",
       "      <td>abnormal CNS synaptic transmission</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MP:0000778</th>\n",
       "      <td>8.846118e-11</td>\n",
       "      <td>1.526470</td>\n",
       "      <td>1.065157</td>\n",
       "      <td>1.987783</td>\n",
       "      <td>215</td>\n",
       "      <td>abnormal nervous system tract morphology</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    OR_p    log_OR  log_OR_CI_lower  log_OR_CI_upper  \\\n",
       "MP:0003861  7.653100e-13  0.975982         0.709088         1.242876   \n",
       "MP:0021009  6.133724e-12  1.016254         0.726598         1.305910   \n",
       "MP:0003635  1.748691e-11  1.001272         0.709483         1.293061   \n",
       "MP:0002206  3.332023e-11  1.020312         0.718737         1.321887   \n",
       "MP:0000778  8.846118e-11  1.526470         1.065157         1.987783   \n",
       "\n",
       "            num_genes_in_term                            MP_description  \\\n",
       "MP:0003861               1234       abnormal nervous system development   \n",
       "MP:0021009                962              abnormal synaptic physiology   \n",
       "MP:0003635                956            abnormal synaptic transmission   \n",
       "MP:0002206                866        abnormal CNS synaptic transmission   \n",
       "MP:0000778                215  abnormal nervous system tract morphology   \n",
       "\n",
       "           MPO_term_type  \n",
       "MP:0003861         brain  \n",
       "MP:0021009         brain  \n",
       "MP:0003635         brain  \n",
       "MP:0002206         brain  \n",
       "MP:0000778         brain  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join brain and heart results together\n",
    "root_KO_brain_df['MPO_term_type']='brain'\n",
    "root_KO_df = root_KO_brain_df\n",
    "root_KO_df.loc['MP:0002419','MPO_term_type'] = 'neg_ctrl'\n",
    "root_KO_df = root_KO_df.sort_values('OR_p')\n",
    "root_KO_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T21:46:29.694026Z",
     "start_time": "2022-06-24T21:46:29.689194Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "brain       106\n",
       "neg_ctrl      1\n",
       "Name: MPO_term_type, dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_KO_df['MPO_term_type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot top performing brain and heart terms + negative control term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:47:52.999640Z",
     "start_time": "2022-06-14T22:47:52.995453Z"
    }
   },
   "outputs": [],
   "source": [
    "# plot top performing brain and heart terms + negative control term (MP:0002419) (for terms which have at least 150 genes)\n",
    "\n",
    "brain_terms_plot = root_KO_brain_df[root_KO_brain_df['num_genes_in_term']>150]\n",
    "brain_terms_plot = brain_terms_plot.sort_values('OR_p',ascending=True).head(5).index.tolist() #+brain_terms_plot.sort_values('OR_p',ascending=True).tail(5).index.tolist()\n",
    "\n",
    "neg_ctrl_terms_plot=['MP:0002419']\n",
    "\n",
    "terms_plot = brain_terms_plot +neg_ctrl_terms_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:55:00.432835Z",
     "start_time": "2022-06-14T22:54:59.449097Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeYAAAF/CAYAAACc+mpEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nOzdeZxVdf348RfDsA+b7JvK+gZBRXMyd8slK1s0w5RSsqxvZppmftGyXFrUNvqamX3zB5pomEsl3yIVC0gtUQHF4A3DosCw7wPDMsvvj/fnyuFyZ+bOzJ177r28n4/HfcDcOffzeZ9zz5z3+SznnFa1tbU455xzLjcUxR2Ac8455w7wxOycc87lEE/MzjnnXA7xxOycc87lEE/MzjnnXA4pjjsA55xrCSLSDigF1gLVMYfjXLLWQD9grqrujf7CE7NzrlCVAnPiDsK5BpwB/DP6hidm51yhWgswdepU+vbtG3csWffkk08CcMkll8QciUtl3bp1jB8/HsJ+GuWJ2TlXqKoB+vbty8CBA+OOJeu6desGcFiue545ZJjFJ38555xzOcQTs3POOZdDvCvbOecK0OjRo+MOwTWRJ2bnnCtApaWlcYfgmsi7sp1zrgDt37+f/fv3xx2GawJPzM45V4CmTp3K1KlT4w7DNYEnZueccy6HeGJ2zjnncognZueccy6HeGJ2zjnncohfLuWccwVo7NixcYfgmsgTs3POFSBPzPnLu7Kdc64A7d69m927d8cdhmsCT8zOOVeAnnjiCZ544om4w3BN4InZOeecyyGemJ1zzrkc4pO/nHPOxaa6Gnbtgp077bVnD9TUQG0tFBVB69ZQUmKvzp2hQ4e4I255npidc85lVU0NbNwI5eWwZQvs2HEgMVdWHpyYi4sPJOWSEujWDXr3hgEDoFOnuNekZXhids65AnTSSSfFHcIh9u61ZFxeDmvW2L/btjX8ue3bD/y/uBj69LHE3L+//duzJ7Rq1XJxZ5snZuecK0BjxoyJO4SDrFsHqrBihSXkffuaVk5VlSX1NWuge3cYNAiGDAGRwunm9sTsnHMFaHtoZnbt2jXWOPbsgSVLYNkyS8x79mSu7K1b7bVhg/07YoS1ovO99eyJ2TnnCtAzzzwDwIQJE2KLYds2WLDAEnJ5ecvVs26dJeYtW2D4cBg92iaN5StPzM455zJu82aYNw/mz4eKipavb+9eeOstm+FdXQ3HHmvj0fkoT8N2zjmXq7ZutaT8xhuQ7buCLl9uiRnguOPys+Xsidk551zG7NplreR587KflBPeecday23aQI7NgUuLJ2bnnHMZUVMDixfD229bgo7TsmV27XPPntC3b7yxNJYnZuecK0CnnHJK1utcvdoS4qZNWa86JVU44gi7rKpdu7ijSZ8nZuecK0AiktX6du2yy6KWLs1qtfWqrLTk3L27TQbLF/4QC+ecK0CbNm1iUxabritXWlLevz9rVaZlzRq7qUk6dxjLFZ6YnXOuAE2fPp3p06dnpa49e+w65Q0bslJdoyVu/5kvPDE755xrlrVrLfnV1sYdSWqJB2bs3Rt3JOnxxOycc67Jamos6a1dG3ckdUvEuG5d3JGkxxOzc865Jtu507qwc701um6dtZzzQV4lZhG5XURWxx1HpojIBBGpFRGfHd/CRGSMiHwsB+IYFr7zs7NY5xQR+We26mtJInKUiHw27jjcATt32vOUc93u3RZnTU3ckTQsrxKzc83wV6A07iBcsz0KXBB3EPngzDPP5Mwzz2zxenbutFc+2LEj/hufpMNbau5wkecPgnOBf49pGjJkSFbqyafEnIi1c+e4I6lfTiVmERkD3A2cBnQC1gAPqOq9ScvdCtwAtAOeAq5X1R3hd7XAl4FxwOmhjIdV9a7I5z8K3AYcC+wG/gTcrKpbw+9XhnI/DPQHLgcmAq8APYDLgL3A/wDTgAeBk4F3QywzQjmDgHuBc4DuwAbgMWCiqlansT2mAK2BdcCV4f9zgGtUtTws0y3UcVHYHvOBW1X1n+H3twPnAu8AHw/rdTHwHVW9L1LX9WGb9FfVfSJyBXAjMALYAjwO3KaqeyLb+WpV/W2kjNXAb1X1dhFpDfwwbLvewEpgkqo+UMe61rm8iBwf1uscVX0x8pnfAkNU9UMicjLwE+AEoAr4O/ANVX0nfJ8DgO+JyARVPVpEioBvAV8B+gFlwC8S6xO6mv8OfBD4FTAEWAB8HrgE+DrQAfg/4IuJ7ZJivcZg+8nJQDlwT4plPgF8DxgdlnkK+J6q7g7reIKqvi+yfA9gLXCZqj4lIgOBn2L7azXwKvAtVV1YR0yDgB8A5wHdgLmk3mf+iv2dFWN/I9ep6vawTG3YFlcBp4R4vgHUYvvjUcDrwARVXRY+k86+ejbwF+A6oHOI7euqukhE/oEdG04TkStV1ZN0PdaFmU59W/B+lLW1lugyOb5cVvY0c+feTUVFOSUl/SktnciwYRdnpOxdu+K7f3dj5ExXtoh0BJ4HtgGnYgepacA9InJSZNEBwPnhdWFYdlpScT8GHgaOB/4A3CkiZ4V6LgKmAy8A7wPGA2cAz4WDdcK12IH7XGB2eO9GYFUo95fAHdhB5BehrBXAIyKSOGA8iyXy8wHBDko3YQemdI0LZZyFHQjPxA6qhHr+CgzHkm5pWK+/i0i0D+s0YDswFjvxeQL4XFI9nwMeC0n5G8BvgclhXb+e+H0j4r4W+AxwKZbc7wN+Vc/Yap3Lq+oC7CB/ZWJhEekQlp8ckvp0YBZ2svUhYCC2D4Btl7VY8kp0Z/8IuAa4HtvX7gHuDSd9UT/Hks8HsBOGV4DjQh3/hX0/X0q1QiLSFZgJVGCJ+WvAd5OW+Sj2ffw2xHE11lX7x7DIZOBEERkV+dhngR3AsyLSKax3EZbUzgSWA/8SkREpYuoCvAQMxU7QSoElwMykv7NS4CPYvvsp7CT3yaTiJgH3Y9v8P1g38/eAL2AnNEcRTkQasa+egu3rH8P+vodiJ0aEeF8N26tf8rq5g82YMYMZM2a0aB3V1VBVlbnyysqeZs6cm6moWAPUUlGxhjlzbqas7OmMlF9dfeDJU7ksl1rMnbA/9Acird/vAv+N/eG/FpbbB1yqquvDMtdiSXWUqi4Kyzyiqo+G39+KHRBPxQ5gE4HpqnpbWFZFZHwo/wIs0QI8p6p/TQQXbm+3KNHyFpGfYQehaar6THjvfuDPQF8R2Qb8DnhSVd8JxfxCRG4O65N8kKtLBfAVVd0PLBKRqdjBEqwl/gGgt6om5hveGVqPN3LghAKstbs5xDkZeElEhqvqUrGVOwn4SjiATgR+raq/CJ9dGk5anhSRY1T1P2nEPRjYBbyjqmuAX4rIYmBRE5d/CPixiFyjqruwgzRY67ILdvKyLnx+hYhcCvQBUNWNIlIDVIT/l2AJ+XOq+mwoZ7mI9ANuEZG7I3Hdoar/CtvtGewk5YshhkUi8i3s+0zls0AJcGXojVkoIjdiJ4sJt2K9DImehGUichXwqoiMVdWXRGQJdmL07bBM9CTqCuAI4PKwjwB8LZyIXoO1YqPGYycYJ6vq2rBeV2OJ8mbsRAOsy/jS8F0gItcAM0RktKq+HZZ5VFWfCr//DZZMb41srz9gyRXS31fbAldE9tVfYj0pqOoWEdkPVKrqQRe+hNZ4t6R1HYhrUbW16U2mmj79krTKW7/+dWpq9h30XlVVJbNmfZPFixtuF1x4Yf2H1Zqa/Jj8lTOJORwwHwA+KyInAMOw1hpYF27CskRSDv4d/h3NgYO4RsqtFZEd2B88WGvnNiJU9XURqQi/SyTmVHd81cj/Ex0iZZH3KsO/7VS1MiTqS0TkJuzM/zisa7wxTwhdHjnggrV8E+tyQvh3RdJ9cdtirfeErYkDHYCqviwiih2kb8da4m+q6hsi0gdLaNGkDnZSQ1iHdBLz/VjPwCoRmQc8Bzye9N01ZvnHsBbvp4FHgAnYSdFuYLeI3It1Gd8hIjOx1tnjddQ1CutKfSQMFyS0BtoTEnqQ/J2vDUk5oTKUlcqx2P66NfLey0nLnACUisiEOuKcD0wBviwi38H2ow9gSTfx+S7A1qR9oD02jJPsOKAskZThvb+ROVjvUEJZIikHr0TWKZGY0/l7SGybdPfVTdF9lYP39/p8AztRdlnUqhUUZbDfNTkpN/R+YxUVZTbelpIziVlE+mJJdgM2nvU3bHwp+fKo5I6IRJKLjnKkGvFoaDyqKOlzqUYiUt0FNuX5V+hinAN0xLrepmDr19jLVupbl6IQ59gGYk21LlOAL4rIHdi47n0plolK7M71jSa1SfwntMSHYd2r5wIfBW4SkS8kejOiGlpeVbeLyFPA50XkBawr+bTI5yeKyK/C587BhheuE5FTVbUyqbrEulwOpBqH3YgNPcCh33ljz7eT97vk8oqAn2E9AskSJyWPAHdh63sudhI1L/L5Mqy1mix5veuTvP8nx5n4O6uuZxmoe/uku682dbRyErZPRw3E/gZdCykqgtZpNDMaaskmPP74+0M39sFKSgakXUZ98iUx51KIlwM9gdNU9fuq+kesiw4OPrgNFZHonLrTsQknb5OeN7FxuPeISCmWQNMtIx0fxloJ56jqd1X1CWAn1hrL1KSVt7C4O6pqWeKFdbc2dK3nI1j38X9hB7CpAKGFup6kbYSN+8GBbbQPa6kBICLdse7kxM83Apeo6vOq+t+qejzwInBFqmDSXP4hLHFfZaG+12U6SkQeBDao6q9V9TPY9h/LgZZa9GaBi7FkcGTSdjsL603JVGfXPGCYiPSKvJd8ydZbgCTF0Q3rHeiJregabP7Fp7GJh5OTPn8ksCPy+WXY/IfzUsT0ZojpvTHaMExxBgfv/8PCd5pwavj3jYZXO6Xm7KtRKW/6qKrbVHVl9MWhJ/Uuw4qKbIZz23T6NNJQWjqR4uIOB71XXNyB0tKJGSm/Uyfo2DEjRbWonGkxY5OqOgCXishsrMXy8/C7aFdhO+APInILlgjuw8bblqdZz93A0yJyFzZZZRA2kes1LBFkSuKgMF5EpoV6foS1KjP1ZNC/YQfKx0XkOmwm8xewg90n6/ugqpaLyHPY5Jy/qGr09vN3YxOhlmGzjo/FWqB/VNUlYZmXgatFZBaWpH/IwS2fnsCtIrILm808EjgR67JOJZ3lZ2Gzy2/l4G7LjdjBvX0YH67Gkvc2DnS77wRGiEj/sO4PYGOc27BejFOwrvAHVLVGMvPIvN8D3wEeC8MZ3bDWcdSPgKdCz8VUbFLTb7Au3JWR5SZjE8Tah+USpmJzAp4K8xe2YGPFn05RV2L5W7D5Ajdhk8huwP7eroos1xHr6r8FO5n8JfBUYoZ1EzR5X02yExgsIkdF5m64FM4555ys1FNSYsl58+aGl21IYvZ1S83K7tzZ4s11udRifhKbtfxjbKx4EtZCmo3NaE2Yx4Ek+iT2B/+VdCsJE7Uuwbr+3sQmaL0InKeqGZtfqKqvYpNarsVaaFOwxPI4B69Pc+qoxiaC/SuUuxBbr3GRSU31mYxdkjIlqdxJ2AHzq1himxSWvSyy2Fexmc4vYePyz3NgvB8scf4WO3FagiWbB4Hv1xFLg8uram2ItQ32vSXe34RN3DsK2xbzsLHY81Q18bC3n2CzjN8Su9Pajdi+dju2v92FnZDcXEd8jRbGoj+Inbi8jPVS3J20zDPY7PILsVblE9g2/WhY34Q/YSc+f4lMniJcvnQmNvHtL9js9ZHAR1T19RQx7cBax2uBGdh3NgQ4W1XnRhYtx4aS5mBXPTyLzUVokgzsqwn3Yeu3SET6NzWew8GgQYMYNGhQi9fTuXNmrwseNuxiLrvsVa6+ejWXXfZqxpIy5E9iblWbq48DcS4FEfk10EdVG3PJmWsEseuJv6SqeT2rWUSOBlbMnDmTgQPzelWaZNWqVQAtnpy3boUXX4S33mrRapqtQwf40IegNEfu/7d69epEr8bgMPTynlzqynauTiJyLjZL+Ups/Ng5V4+ZM2cCMGHChBatp0sX6NUL2rSB/ammA+aI3r2hR4+Gl8sFudSV7Vx9rsLGsX+kqsmXcjnnYtK6NfTvDy14g7Fma9XKYuyfJ4Mf3mJ2eUFVL487hsOFqt6Ojb07l5YBA+wVes9zTo8elpTbt487kvR4i9k551yzdOgA/frlblfxgAH501oGT8zOOecyYPBgGDEivRuOZFPv3nD00XDEEQ0umjO8K9s55wrQBRdk97HVXbrAsGGwaRMsTXVD4xi0bQsjR4KIjTPnC0/MzjlXgFrycY91Ofpou9HIxo2wbVuDi7e4ESNg+PD8uNtXlHdlO+dcAVq+fDnLl6d7Q8TMaN3aWqjHHBP/RKuBA2HoUPs333iL2TnnCtDs2XZV4ZAhQ7Jab5cucPzx9tzjefNgX2YeDNUo/fpZDKNH51cXdoInZueccxnVpw+MDc8RW7AA9uzJXt0DBlhSPu64zD1cI9s8MTvnnMu4fv2sa7u4GBYvzsxDLurTurV1XYvAscfmb1IGT8zOOedaSO/ecMop0L07lJXZqypjjwo6oHt3G9sePhyOOir3LtlqLE/MzjnnWkxJCZx4IvTsaTcgWb4c1q+Hmgw89bxjRxg0yFrKI0fa+HYh8MTsnHMF6MILL4w7hPcUFdmlVL162Szp8vIDr8rKxpXVqpUl+MTdvBKvfG8lR3lids65AtSzZ8+4QzhEp04wapS1cNets8S8eTPs3HngVVlprenaWkvoxcXW6k48S7lrV5tcNmCAdWHn46zrhnhids65AqSqAIhIzJEcqm1bOPJIe1VWQkXFgcS8Z8/Bibl1a0vI0eRcXOCZq8BXzznnDk+vvPIKkJuJOapDB3v16hV3JLnD7/zlnHPO5RBPzM4551wO8cTsnHPO5RBPzM4551wO8clfzjlXgC666KK4Q3BN5InZOecKUNeuXeMOwTWRd2U751wBWrhwIQsXLow7DNcE3mJ2zrkC9NprrwEwZsyYmCNxjeWJ2TnnXM7Ztw+qqw/cAaxNm8K6H3Z9PDE755yLVU0NbNkC27cffN/saGJu397uk524NWfPnvZeIfLE7JxzLhZ79sDatfYqL7fkvHMn7NqV+rGQ7dtbUu7cGfr2hX79CvNhFp6YnXPOZVV1Naxcac9mXrPGEvPevQ1/bs8ee23cCCtWHHj846BBIGIJuxB4YnbOuQI0bty4uENIaccOWLwYli61xFxd3bRyamth0yZ7rVplre3hw+Goo6zrO595YnbOuQLUsWPHuEM4xOrV8PbbsGgRbNuWuXK3bIFXX7VnO2/ZAscea4+WzFeemJ1zrgDNnz8fgLFjx8YciVm5Et58E956C/bvz3z5VVWgas92rq6G44+Hdu0yX082eGJ2zrkClEuJedUqS8oLFjS96zpda9bYxLFWrWDsWLvMKt/keU+8c865XLZtGyxcaIm5pZNywtq11jJXzU59meaJ2TnnXIuorraJXosXW1dzNr37LixbBuvXZ7feTPCubOeccy1i+XKbfb19ezz1L15s1zh365Zf483eYnbOOZdxu3dbi3XFivhiqKy0E4N33okvhqbwFrNzzhWg8ePHx1p/efmBiVhxx1FeDkOGQHGeZDxvMTvnXAFq06YNbWKaklxdbclw3bpYqj8kljVrciOWdHlids65AjR37lzmzp0bS90bN1pibonrlZuivNxmaucLT8x5SkRuF5HVcceRKSIyQURqRSRPOpsOJSJtROTGyM8tuk4tvQ80tvxC2yfz3dtvv83bb78dS93bttlduHLFrl0WT66cKDTEE7NzmXMF8NPIz9OAfqqa5QtFMuYnwAlxB+HyT+KxjbkkF2OqS962TpzLQQc9eE5VK4HKmGJpNlWtACrijsPll+pquzxqz57mlVNW9jRz595NRUU5JSX9KS2dyLBhFze5vIoKex1xRPPiygZPzDlKRMYAdwOnAZ2ANcADqnpv0nK3AjcA7YCngOtVdUf4XS3wZWAccHoo42FVvSvy+Y8CtwHHAruBPwE3q+rW8PuVodwPA/2By4GJwCtAD+AyYC/wP1gL8UHgZODdEMuMUM4g4F7gHKA7sAF4DJioqg3eD0hEegP3Ax8M22MecKuq/kNErgd+CPRV1Z2RzywFfgc8AqwI2+Em4Ljw8yRV/U1Yth3wfeASYACwC5gJXKOqG0Tk6PCZz4X1Hwa8BXxTVeeIyATgfyPb/QshjMlAG1WtEpFOoY5xQDdgQdjW/6xjnWuBa4DPAycCZcB3VPWPScvdBFwP9AReBb6sqioiPwcuBo5W1dqwbAdgPXCtqj4SPvtVYCCwNsR7l6rWiMjtwJdUdWD47CDgB8B5If654TuoK/4jgDuATwK9gbeBO1X1T5Flzsf282PC+v0U+H/A4PC5Or9XVb0zVb0uXpWVlgCbo6zsaebMuZmqKjuvrahYw5w5NwM0OTlXVFiXdj7wxJyDRKQj8DyWGE4FqoAvAveIyIuq+lpYdABwfnh1xhLDNOAjkeJ+DFwLfA1LFneKyGxVnSUiF2FJ9wfABOBI4D7gORE5WVUTFzpcC3wKO6AvxhLTjViSOR7rwr0DuBJLfAr8DHhERPqEpPAslozPB3YAHwcmAf8GnkxjszwItAHOwk4Evg38WUQGAI8C92BJdXLYhqcDQ4GHOdCS/VnYDkuBm4EHROQFVV2OnTR8KqzDCix5Twa+G9Y/YRJwHTAf+O+wrY4N2/0ILLH0A7YDlyatwzQsAX0pxHADMENERqpqXWOz92Lb+wvYd/S0iJylqnPC7weEbfIx7OTsdyHuU4GHgG8AZwN/D8t/Ovz7pIh8Erg1xKnAKdhJzDvAlGgQItIFeAlYhSX7naHsmSJyWmSfTCzfGtuHOwJXhc9dATwjIp9S1T+LyFjg/4BfYCd4Y4EHIsU09L1G6+uGnSxEDcRlXXV16rt8TZ9+SdplrF//OjU1+w56r6qqklmzvsnixY+lVcaFFx58WKmuzt4tQZvLx5hzUycsAVyjqotUdSmWIMBatgn7gEtVdZ6qzsYSyAUiMiqyzCOq+qiqLsEOwjuxgzbYAX+6qt6m5nlgPHAScEGkjOdU9a+q+oaq7g7vLVLVu0JS+1l4b5qqPqOq/8Fat72AvqGV9jvgalWdr6rLVfUXQHnS+tRnMLANWKGqZVgL8WKgSlU3Yy39KyLLXwnMVNXorQV+rqp/VtVF2AlEEda6B2v9XaGq/1DVd1T1WeBvKeK7V1UfD2VcDWwEvhK6rXcAqOq68PN7RESw5Hlt2JZlWIL/NdaDUJdHVPX+8P3cgp3IXBf5fTXweVV9U1XnYicw7wtxLMRa0MnbZVr4HgdjJznvquq7qjoN69H4O4caj7V6L1HVV0LZV2MnajenWP58rJU/XlVfCPF/G0vE3wnL3ADMV9Wbwu+nAd9LFNCI7xXsJGFF0msOh7EJEyYwYcKErNdbW9v8MpKTckPvp1dm/NdUp8tbzDlIVTeKyAPAZ0XkBKzb9Pjw69aRRZepavROsP8O/44GFiWKi5RbKyI7gMSTSo/DurGjdb8uIhXhd38Jby9NFWbk/4lkXRZ5L5GY2qlqpYjcD1wSuk6HhvL7J61Pfe7AWlCfFpF/As9h3ZmJeh7CWp9HYS3zcVgXbV0xJ24SmNgWU4FzReQeYDgwEhCsyz5q1nuFqe4XkddI7+Qiscy/Ip+vxk4Q6jMr6edXgI9Gft6oqtEn224F2opI61D+Q8BPReRr2NDDh4AzwrKPYq3ZxSLyH6yF+4cUSQ/s+ypT1fcuOgn70xzg3DqW36Wqb6RYn8RQyokcehKQnEzT+V7BTmSnJL03MEV5roW1amWvZMkt2Po8/vj7qahYc8j7JSUDGlVOVFGRvfJBnoR5eBGRvtj45dXYuPD9pJ4dm9wxk0hyeyPv7eVQKf5sDlKU9LndKZZJdeFByvPRMLb6MtZS2oodQE8D0r60RlWfwRL5BKw1dD2wMIzFA7yAdZd+DhubrAWeTiqmvm3xINbV3Bb4M9a1+niK5ZPXuzWHfg+pNPVUv6H66qo7sV6/x77Pi7BW71JVfRlAVTdh3cenY8MJJwNzROSWRsSXvK9EpWo7FXFgW1TR8L6YzveKqm5T1ZXRF43YvwrRyy+/zMsvv5z1etu0gbZtG16uPqWlEyku7nDQe8XFHSgtndjkMtu2zZ9HQHpizk2XYxN5TlPV74fJPom5hNED2VAR6Rz5+XTswJXuxYtvAmdG3xCRUmxcMJMXQH4YO7E4R1W/q6pPYF3qfWj4wIyIdBCRScAQVZ2mql/GehHaAJ8ACOPhk7Ex1IuBx1U1rXmhItIDOwn6uqreoKpTVHUBMCpFfKWRz7XFuo0TrcL6OvESPRjRzxeJyFIR+WI9nytN+vnUSH0NUpsI+AcObJfJkfqvxIZLXlLV76nqKdg8hS+kKOpNYJiI9IvGj7W+U+0rbwIlInJi0vtnRZZfwIGhhISDfm7O93q4W7JkCUuWLMl6vR06QNeuzWudDht2MWeccS8lJQOAVpSUDOCMM+5t1qzszp3tlQ+8Kzs3rQI6AJeKyGysS/Xn4XfRZ6S0A/4QWjg9sIlbj4Vx33TcjU0mugvr1hwE/BJ4DXix2WtxQKLlMl5EpoV6foQl1gaf+RK6wt8PnCYi12Gzhz+CTXiLdjVPxrrmR2IJIF3bw+sTIvIq0B74OtbV+nrSsneKSDmwDJuAVoKNE4OdbCRObhZFP6SqS0XkKeA+EfkqNmv9Bux7+1s9sV0nIm9jY+Bfxlq4X27EuoF1Bz+HtbY/GXm/HfDjMLwxG+v6/SDWu5FsKnALNmnsJmw8/QZs37wqxfLPYScQj4rI17F94AqsG/6isMxPgPkicjc2E/sYDnRzR09ymvq9uhi0agVdukCnTs27bnjYsIublYiT5VNi9hZzbnoSm437Y+wAPwk7uM7m4BbFPA4k0SexA/xX0q0kdA9fgk1KehOboPUicJ5m8KYYqvoqNov7Wmyy0BRsrPFxDm0x1eUSbAz7j9hY8VexSU/vjVGGsdG/A8vDRKh046sCPoO1kN8EZmC9BrcCo5J6JX6FfTevYzOiz1bVxGDYc9jM5ZeA/0pR1VUhvt9j390xwPn1zMgGSxYAuEsAACAASURBVPrXYa3LM4EPq+q8dNctrN8cLDH+LWmM+DfYpMLbsO/lD9g487UpytiBtY7XYtvn38AQbP0P2dZhfPt87MTp99gs9nOAT4QeoMTktIuAC7GhmzuxYRuIdP039Xt18Skpya0kmDhZ6Ngx7kjS06o2E1PonMsRIrII+HWY9Z3Jco/GxrbPU9UXMll2PXXWYjPZf9vMctpjCXWCRq4hjlvoBanRyKVWInI51nouiZ4cNuV7TXxnM2fOZODAw+/KqSlTpgDEMjN72zaYNQvmNeoUsuX07AlnnQXHpnsNSBasXr2ac845B2BwmBPxHu/Kdnkv3Iv6U9gYe18OnZ17WArX9p6Djc1uBKbHG9Ehjgd+IiKfx1rUw7DZ979XuyGLf6/NENeTpQC6dYP+/WHJkty4qUf//vbKF56YXd4LB/FJ2NDM51V1e0OfOUwUY0Mgm4HLNI07rGXZb7GbsUzChgU2YMMb3wP/Xpsr7ucx9+8PAwZYco5Thw4WSz7cijPBE7MrCBpuG9mC5a8kjRnkGa6zWfWFy6GS74aVM9TuCHdneNW1zOHXB10g+vSBQYNg9WrYneqCyywZPBgGDkx9bXWu8slfzjlXgGbNmsWsWcn3p8me4mIQsVdcSbFXLxg2LL+6scETs3POFaQVK1awYsWKWGPo3RuGDrUu7Wxr0wZGjrRXvtzxK8G7sp1zzrWYESNslnZFhf2bDUVFcMwxVne+XCIVlWfnEc455/JJ27YwdiyccIJdS9zSEkl5zBgb485H3mJ2zjnXokpKLDG3agULFsDmzS1TT3ExjB5tSXnYsJapIxs8MTvnXAHqmGN9uF26wPveZ5cvLV0Ky5Zl9vnIPXrYeLIIHHlk5sqNgydm55wrQOPGjYs7hEN06gQnnmhJtEcPu8a5ua3ntm1hyBBrIY8caa3zfOeJ2TnnXNa0bm2JtGdPu9a5vNxe69ZBVSPu0N+tm8327t/frlMeNCj/Zl/XxROzc84VoBdesFu6n3vuuTFHklqXLjYpbPhwS8xr18KWLTZ7e+dOu5VndTXU1lrCbdfuwBOiOne2pD5ggCX4QknICZ6YnXOuAK1eXd9Dy3JHp06WnIcOtWScSMy7dx+cmNu2tYSceHJV27ZxR95yPDE755yLXVHRgdZwv35xRxOvAusAcM455/KbJ2bnnHMuh3hXtnPOFaAu2bjNlmsRnpidc64AXXzxxXGH4JrIu7Kdc865HOKJ2TnnCtCMGTOYMWNG3GG4JvCubOecK0Dr1q2LOwTXRN5ids4553KIJ2bnnHMuh3hids4553KIjzE751wB6tGjR9whuCbyxOyccwXo4x//eNwhNElVlT3MoqoKamrsHtrFxfawi+LDJGMdJqvpnHMuF1VXw8aNsHWrPVVq+3b7d/9+e7JUq1bQpo093KJrV3u61BFHQK9e9mznQuSJ2TnnCtCzzz4L5G7LubLSnsOceG3caAl57966P5N4JnOvXtC//4FXhw7ZizsbPDE751wB2rx5c9whpFRTA+++C0uXwqpVsHattY7TsXevvTZtgrIyezzkoEEwbBgcdZR1excCT8zOOeeyoqICFi+2pFpWZuPITbV/vyX48nLYvNleI0daV3e+88TsnHOuxW3cCPPnW2LOZGO+qsrK3LDBxqmPPx56985c+XHwxOycc65FrV8P8+ZZYt6zp2Xq2LIFXn/dJpONHQt9+7ZMPdngidk55wpQ3xzJTJs2WUKeN6/+iV2ZsGeP1dOqlV1a1bNny9bXUjwxO+dcAbrgggviDoG9e+E//4EFC1o+KUfrnD8f2reHD3zAZnLnmwKZw+accy7XLFkCqrB7d3brray0upcsyW69meKJ2TnnCtDTTz/N008/HVv969fDsmWwZk089a9ZY/Xn49MvvSvbOecK0I4dO2Kru7YWli+Pv8WqCn362CztfLrGOY9Cdc45lw82b7bri7PdhZ0scXexLVvijaOxPDE755zLqMRtNnPBmjXxdac3lSdm55xzGbN3r91mM1daqVu2WDzZmhWeCZ6YYyYit4vI6rjjyBQRmSAitSKSt/MXRKSNiNwY+blZ6yQiR4nIZzMXYfaE9f5SBso5OpR1bibicg0bOHAgAwcOzHq9O3bYHbhqa7NedZ22bbO48kXeHjyda0FXAD8FfhZ+ngbMUNWm3tn3UWAZ8PsMxJZt/YDtGShnVSgrR9pRhe/cc+M5B9q50+6JnUt27rRXr15xR5IeT8zOHapV9AdVrQQqM1VePlHVjFxsoqrVQB5euOIaK5EEm6us7Gnmzr2biopySkr6U1o6kWHDLo41pmzxxNzCRGQMcDdwGtAJWAM8oKr3Ji13K3AD0A54CrheVXeE39UCXwbGAaeHMh5W1bsin/8ocBtwLLAb+BNws6puDb9fGcr9MNAfuByYCLwC9AAuA/YC/4O1EB8ETgbeDbHMCOUMAu4FzgG6AxuAx4CJ4eDb0PboDdwPfDBsj3nArar6DxG5Hvgh0FdVd0Y+sxT4HfAIsCJsh5uA48LPk1T1N2HZdsD3gUuAAcAuYCZwjapuEJGjw2c+F9Z/GPAW8E1VnSMiE4D/jWz3L4QwJgNtVLVKRDqFOsYB3YAFYVv/M8X6/gP77k8TkStVtVV4rwwYDRwDfDOsXzpx17fudW7b8PuVwCTgDOAjwFbg9rD+94dYFgNXq+prkW1wtar+No3yTwZ+ApwAVAF/B76hqu9E4j9PVV8QkdbA14GvAIOBcuA3wL2qWpPO+rr6PfHEEwCMGzcuq/VWVjZ/NnZZ2dPMmXMzVVV2PlxRsYY5c24GaFJy3r3b4soXnphbkIh0BJ7HDrCnYgerLwL3iMiLiYMfdiA+P7w6Y4lhGnbwTPgxcC3wNSxZ3Ckis1V1lohchCXdHwATgCOB+4DnRORkVa0JZVwLfApYjx2AJwI3YgnheKwL9w7gSuxgqFh37iMi0kdVa4FnsWR8PrAD+Dh2sP838GQam+VBoA1wFnYi8G3gzyIyAOvyvQdLTpPDNjwdGAo8zIGW58/CdlgK3Aw8ICIvqOpy7KThU2EdVmAH9MnAd8P6J0wCrgPmA/8dttWxYbsfgXVlJ7pxL01ah2lYEvtSiOEGYIaIjFTV5PkCFwN/BVYC10fevyq8XgU2NSLu+ta9zm0bOdH5AfAN4FvALcCvsO/5G8DGUOevgZM4VH3f3W5geljmCuyk7UHsezs7RVk/xfbj64CXsJOFXwC9sX0ynfV9j4h0w06SorI/wJpDdsd0rVJNjT1IImr69EsaVcb69a9TU7PvoPeqqiqZNeubLF78WFplXHjhgcNRTY298oVP/mpZnbAEcI2qLlLVpdiBFqxlm7APuFRV56nqbOxAfIGIjIos84iqPqqqS4BbgZ1YsgdLsNNV9TY1zwPjsYNr9Ia5z6nqX1X1DVVN/NUuUtW7woHuvTFVVX1GVf+DtZB6AX1FpAPWsrtaVeer6nJV/QXW2omuT30GA9uAFapahiWri4EqVd2MtfSviCx/JTBTVd+JvPdzVf2zqi7CTiCKsNY9wFzgClX9h6q+o6rPAn9LEd+9qvp4KONqLCl9JXRb7wDrxg0/v0dEBPgYcG3YlmVYcvk1lowOoqpbgP1AZVK38NuqOkVV/6OqGxoRd33rXue2jXz+eVX93/B934ednP+Pqr6gqguwxFzXd1lf+V2wnpd1wDuq+gZ2QnNLciEi0gW4BrhDVR9W1TJVnYydFF4rIl3TXN+ob2AnNNHXnDrWw+W45KTc0PsNqa3NrcloDfEWcwtS1Y0i8gDwWRE5Aes2PT78unVk0WWquj7y87/Dv6OBRYniIuXWisgOoG146zisGzta9+siUhF+95fw9tJUYUb+n0jWZZH3EompnapWisj9wCUichPWkj0O6xqPrk997sBaxp8WkX8CzwG/iyTAh7DW51FYy3wc8NV6Yk5MTEpsi6nAuSJyDzAcGAkI1mUfNeu9wlT3i8hrpHdykVjmX5HPV2NJozGSv4t0465v3Rvatsmfr+v7bktq9ZVfKSL3YkMhd4jITKyn4PEU5YzCWt6zk96fFd4fxYHx6PrWN2oSMCXpvYF4cs66oiJ7RVuo0dZrOh5//P1UVBx68XFJyYBGlxWNKV/kUaj5R0T6YuN3V2Pjwvdj42/JksdmE0kueuVdqqvwGppUVJT0uVR9W/tTvJey0yeMrb4MfAcbn5yCjZ+mfbmXqj6DJfIJWKvmemBhGIsHeAGbwfs54JNALZB8w9/6tsWDWFdzW+DP2Nh5quSQvN6tOfR7SKVpp+yHSv4u0o27znVPY9tCI77vZA2Vr6oTgaOxLu5WWNf0K6GnJR2J41Gj93tV3aaqK6MvGrFfusxp186e7NQcpaUTKS4+eLcpLu5AaenEJpXXoUPzY8ombzG3rMuBnsBwVd0HEMYx4eCDy1AR6RwZBzwdS0hvp1nPm8CZ2MQbQj2lQMdGlJGOD2MnFgNVdU2o5wigD2nMPA4H6B9hE9emAdNEpD025v0JYGGY+DMZ+DTWkntcVdN6tLqI9MBOgj6nqlMj748CkssoxSYvISJtgfcRxrWxbV+XRA9GKdbVjIgUYS27u1X1oRSfqbcTrZFx11VGg9s2nXKaWr6IVGPdyder6q+BX4vIKdiJ3AnYcEfCIuwE4Uzgtcj7ibHrZdg4v2uGwYMHx1Jv5872as4lU4kJXpmald25M5SUND2ebPPE3LJWAR2AS0VkNtY1+fPwu+hTQtsBfxCRW7BxuvuAx5InuNTjbuBpEbkL62ocBPwSO+i92Oy1OCDRAhkvItNCPT/Cuh8bfOpp6Ap/PzZD+TpgLTbBrTMHd9lOxrrmR2IH63RtD69PiMirQHts5u+JwOtJy94pIuVYEvg2UIKNE4ON3ydObhZFP6SqS0XkKeA+EfkqNmv9Bux7+1sdce0EBovIUUlj5U2JO6VGbNsmSaP8jcBngfYicjfW+3AVNib9HyITs1R1h4j8CvieiGzCkvfp2Hf+QPi9J+ZmOuusxvzpZE4iMa9d27xyhg27uMmJOFlJicWUL7wru2U9ic22/TF2gJ+EjaHO5uAJLPM4kESfxA7wX0m3ktDFeAk2KelNbILWi9ilKU29KUaqel7FZsxei83qnoKNCz5O6gk5qVyCtYT/iLUyvwp8XlX/HqnnHexSm+WqOrcR8VUBn8HGKN8EZmC9BrcCo0Qk+qf5K+y7eR2bFX92ohcAGzt9Kbz+K0VVV4X4fo99d8cA56eYkZ1wH3aSsUhE+jcz7vo0uG2bqc7yVXUTNtHwKGz8fR42B+E8Vd2WoqxvYtv/Tixx3xr+/80Mxepi0rkzdOkSdxQH69o1vxJzq9p8mqrmDhsisgj4dZj1nclyjyZyPW0my3a5JfFdz5w5M5ZbU8Zt6lQbFRk/fnzW637tNZg1Kzdu6tG5M5x1FpyU6gLAGK1evZpzzjkHYHCYE/Ee78p2OSPci/pTWLdmXw6dZeucS9P+/anm+WXHgAH2Wrw4thDe07+/vfKJd2W7nBG6dCdhl0h9XlUzcY9m51yW9e5tybBNm3jjaNPGThD69Ik3jsbyFrPLKaraon2Oocsob+9d7Vw+aN0ajjwShg6Nt9U8dKjF0TrduyzkCG8xO+ecy7gjj4Rhw6BHj3jqP+IIq//II+Opvzm8xeyccwVoxIgRsdZfVAQjR9qzmf/9b6jK2PUhDSsutrpHjsyvO34leGJ2zrkCdOqppza8UAsrKbHkuHs3LFiQnQdJFBXBscdavfl0U5EoT8zOOedazKBBsG+fJeWFCw998lQmtW4NY8bYKx+7sBM8MTvnXAGaMmUKABMmTIg1DrBJWEVFljgXLWqZZyO3bw/HHGNJeciQzJefTZ6YnXPOtbjBg6FjR+jWDVRhzaEPj2qyAQNgxAgQgb59M1duXDwxO+ecy4o+fSwxd+8Oy5bBO+/Y5LCm6t4djjrKWuQjRtiTrQqBJ2bnnHNZ066dTc7q1w+OPtoedrFmDWzcCHtTPeQzxed79TpwR6+BA6FnzxYPO6s8MTvnnMu6nj3tNWIElJfDli2wfbvdX3vnTru8qqbGxqaLi22GdZcu9kCK7t2t+7pjx7jXomV4YnbOuQI0evTouENIS8eOdiMQsGRcUQG7dh2cmFu3tsRcUmJJutAdBqvonHOHn9LS0rhDaLTiYhuD7tat4WULWR7eE8U551xD9u/fH+sTplzTeWJ2zrkCNHXq1Peeyezyiydm55xzLod4YnbOOedyiCdm55xzLod4YnbOOedyiF8u5ZxzBWjs2LFxh+CayBOzc84VIE/M+cu7sp1zrgDt3r2b3bt3xx2GawJPzM45V4CeeOIJnnjiibjDcE3gidk555zLIZ6YnXPOuRzik7+cc87ljOpqe7pURcXBT5hq0wY6d7anURUVeJPSE7NzzrlY7d0La9fCxo2wY4e9Kipg//4DibltW0vMiWcy9+4NffpYwi40npidc64AnXTSSXGH0KCdO+Hdd6G83F7r1kFlZf2fadUKOnWCfv2gf397HXkkdOiQnZizwROzc84VoDFjxsQdQp1qamDlSli6FJYts5ZybW16n62ttdb00qWwfDn07QvDhsGIEZakW7Vq0dCzwhOzc84VoO3btwPQtWvXmCM5WEUFLFp0IClXVze9rOpqWLMGNm2CLVtg6FAYNQratctcvHHwxOyccwXomWeeAWDChAnxBhKxfTvMnw8LFlgizZS9e+Gtt2DzZtizB44/Pr+7tj0xO+eca3E7d8K8efDGGza5qyWUlx+YMHbiidC+fcvU09IKfNK5c865uO3bZy3lefNaLiknbNxoLfK33rIEnY88MTvnnGtRS5fCkiXWlZ0N69dDWRm880526ss0T8zOOedazIYNliRXr85uvcuW2QnBzp3ZrTcTfIzZOecK0CmnnBJ3CNTUWHJUTf9yqEypqoLFi6FnTxtvzieemJ1zrgCJSNwhsGmTXc4U19Mnt2yxCWEjRkBJSTwxNIV3ZTvnXAHatGkTmzZtijWGNWssMcapvNziyCeHXYtZRG4HvqSqA+OOJRNEZAIwGWijqlX1LHcacCNwKtAVeAd4DPiZqu5KKusGVZ2U9PmjgRXAear6QnivH/A94GNAH2Ar8A/g+6r6VqbWMQ4ichRwiqr+Pvz8D2C1qn6uieV9HFimqv/JXJQtL939K82ybqeA/vZy3fTp04H4rmPevdvuf71tWyzVv2f9eotj2DBo3TreWNLlLebDgIh8Hfg7llgvBMYAdwBfAWaISPLVfj8UkeENlNkOmAUMBS4HhgOfBNoCL0su9KM1z6PABZGfLwa+1pSCRGQo8GegdwbiyrZpQL/mJuXgJ8AJGSjH5YEdO2Dr1rijsLuDbd1qT6zKF4ddi/lwIyLHAT8DblTV+yK/Wi4iCrwOfBG4P/K7tcBkETlTVeu6EvA8LBmfrqobwnvviMg44F3gy8A3M7gq2XbQHXdVtTn3Kcrbu/eqaiXQwGMF0i6rAqjIRFku9+3cmTszohOxdOkSdyTpKbjELCJjgLuB04BOwBrgAVW9N2m5W4EbgHbAU8D1qroj/K4WSyzjgNNDGQ+r6l2Rz38UuA04FtgN/Am4WVW3ht+vDOV+GOiPtSonAq8APYDLgL3A/2CtkgeBk7Gkdr2qzgjlDALuBc4BugMbsC7oiaqazl1mrwa2AQ8k/0JV54nIh4DkbuergBeB64Gf11Fuou4Lgf8XKXN/6Dav8wActt2dwDHALuAvWPf5FhGZB7ylqldElj8dmA0MASYAZ4fPXAd0BuYCX1fVRWH5eveB0KV6LvBXbB8oxr6/61R1e+i2Pg04TUSuVNVWyV3ZInICcA9wCpa4pmMnPwd13IUhgKXhx7+LyMPA7VjvxbeBrwP7gfdhwwENxd3Qute3bc/Gek4+CPwqbM8FwOeBS0IsHYD/A76oqnuSu7LrKz/UfxPwVWAg4QQPuEtVa5K7ssO+/QPsJK9bWJdbVfWf6a6vy13NTcxlZU8zd+7dVFSUU1LSn9LSiQwbdnEssWRbQXVli0hH4HksEZ0KjMaS3j0iEn0G2gDg/PC6MCw7Lam4HwMPA8cDfwDuFJGzQj0XYQfiF7AD6njgDOA5EYlu02uBb2FJYHZ470ZgVSj3l1iX8l+AX4SyVgCPiEiilfUslsjPBwRL0jcBF6W5WU4CXq2rK1JV/6Gqm5PefinE8wMRGVFHuS8A/wYeEpGlInK/iHxWRHqp6vJIK/ogItILeAZL5qOAT2Hb7mdhkYeAi0UkOodyAvB3VV0Zfj4FOAsb274Q607/VSg/3X2gFPgItl0/hZ2APRl+dzHwKvAE0C/FOgzGvs9NoY5PYCdVU1Ks8qoQL8CnsZOdhKuwpPRpLMmlE3d9697Qtk34eaj7A1j3+ivAccCHgP/CTki/lGK96y1fRD4J3BrKGA78d/j5ihRldcH2s6HY9i4FlgAz011fl9v27Gn4EY51KSt7mjlzbqaiYg1QS0XFGubMuZmysqebVF5lpcWTLwqtxdwJmIS1MhKt3+9iB4hjgdfCcvuAS1V1fVjmWiypjoqciT+iqo+G39+KjS+eio2rTgSmq+ptYVkVkfGh/AuwRAvwnKr+NRFcGHZdlGh5i8jPsMlT01T1mfDe/dh4ZF8R2Qb8DnhSVRP3sPmFiNwc1ieRSOpzBLAsjeWSfZvQGhaRM5N/GVrGZ2MnH5diB+NrgGoReRS4NnRdJhuAjUOvDuv0TpgYldgXp2InRZ8GHhaRDsBnQtkJbYErEicUIvJL4Ifhd+nuA62wfWBNWOYabLx9tKq+LSL7gUpVXZdiHb4MbAcmqOq+8PkvAh8TkaJo97+qVotIYmrsltAi7x5+fkBVF4bP90oz7vrWvaFtm3CHqv4rfP4ZrKX8xTAJcJGIfCvUmayh8gdjvUDvquq7wLsisgY7OUk2HjspOFlV14ZYrsYS9M3YyUFD6/seEemGtbqjDutJZmeeecifbVbV1Bx8S8zp0y9J+7Pr179OTc2+g96rqqpk1qxvsnjxY2mXc+GFT6aMJdcVVItZVTdiXbafFZEHROR5IHG/meh8vGWJpBz8O/w7OlpcpNxaYAd2kABrXcwmurDq61j37XGRt5dyKI38P3F1X1nkvcQ5Zrswvnc/cIaI3CcifxGR1VjXeLrzCzdiLe5GCXV/AWuxXF/HMntU9SeqWhrq+CTWyrsS+E0dn5kPPA78SUTKReQRbLu/FX6/FWuVXRk+cjFQC0RPlTcltfK3E76bRuwDZYmkHLwS/k2VkJIdC7yRSMqh3n+p6m31jMmn8t7+0Yi461v3erdtRPI+uDYxMz+oxIZ4Dv5Qw+U/iu1vi0XkbRGZBNRGTiqjjsO+g7WR8muBORz8N1Tn+ib5BtbbFH3NSbHcYWPIkCEMGTIk1hia+mzk5KTc0PuFpqBazCLSF0uyG7Axw79hY1LJN4NLHptNHPj2Rt7by6Ea2s2Kkj6X6rL6/SneS3kwF5FO2MGlI9atOgVbv382EEfUy8AXRaR1qjFpEXkQWK6q9yT/TlVfEpFfYOOAbyd97ktAe1X9ZVh2G9bS/3NoIV5VV0CqermI3IF1JZ8LPAJ8QUQ+Gg7ODwHPhzHIzwO/DycKCam+m0Rc6e4Dyd9DYh9IZ9x+H5n523lv/2hE3HWuO9S/bSOLJa972icTDXx3m0RkLHYylxgquk5Evq2qP0qziuS/oXrXN2IShw4lDOQwTs7r1llnT9++fWOpv3VrKCo68LzlROs1HY8//v7QjX2wkpIBjSonobg4fy6VggJLzNgEq57A8EgXY6IFFE2qQ0Wks6ompgOcjrXKDko+9XgTOBO7/INQTymWQNMtIx0fxi4vGRjpcj0CmySU7rnoZGyC0zVAdFY2InIiNiO7vtnT38bG95LH9UYBE0Tk0eQJT8AWLMEcIkwM+4yqfgNruU0SkcuwCW39gHJs4tlKbGz5Q9g4ZrrS3QeGiUj3xGQ9bJgC4I3wb303EFwEXCEixYmxexE5H0tSI1Nsj3RuRphu3HVKY9s2S0Pli8h5QOdwsvYS8L1w4vcFIDkxv4kl9H6Rruwi7Ltu9DXwYZsnT7xrbDEFZcaMGUB81zF37AidOjXtaVKlpROZM+dmqqoOnI8XF3egtHRik2Lp1MniyReFlphXYbNKLxWR2dhkqcSs4mjXXDvgDyJyC9YFex/wmKouT7Oeu4GnReQurPtuEDaR6zUsqWRKorU0XkSmhXp+BLQhRVdjKqq6KKznz0ML9DFsotEZ2FjdbODX9Xy+UkSuIqnrHvgp8FlgTmhBvY6N8Z2BTXi7htS2AdeIyF7gt1i35GXAcmB9qLNWRCYDtwBLVfXfdZSVSrr7QEdskt0t2InOL4GnVDUxHr8TGCwiR6Xoiv0lNkv4NyLyE2y2/CRgToqknCgL4DgRqSvppBt3fRratnVN5EtXQ+W3A34sIjuw/WUgNgP85RRlTcW+3yfFZnLvwE4ghXp6W1z+6NzZXk1JzInZ15malV1SYrHki4IaY8YmQ92LTR5ahB0sH8IOEidHlpvHgST6JNZt+JV0KwkTtS7BWpJvYhO0XsTuipWJGzEk6nkVm8V9LbAY66qbhY3znVz3Jw8p5yfYDNoTgOewS2RuxGZeX6iqDXWPvoRty+h75cD7sW71e7HtPRsbE75MVR+po6y3wzIfxL6Hl7EW5YeTutqnYInq/yWX0YB094FyrKt4DjYu/izWbZ5wHzASmwzVP2kd1mLdtEOxE5KnsMuQUiaUMEP9NyGuyc2Mu06N2LZN0lD5qvob4LvYZYSLsasZnsf23+SydmAncWuBGVg3/hDgbFWd29xYXfwSibmphg27mMsue5Wrr17NZZe92uSknIlYsq1VbbYf+eFcGkTkA9hJyMAwMSqTZd+O3xqy4Em4jezMmTMZOPDw+6qnTJkCxNeVvW8fzJ4NL78cmyRsFAAAIABJREFU74zodu3grLPgAx+wMe9csXr1as455xyAwZFLQYHC68p2eU5sYPA47DKhqZlOys657GjbFvr3h1697H7Vcenf3165lJQb4onZ5ZrhWDf2a9hYtXOuCUJrLFaJpBhXYm7Vyurv1+ypj9nlidnlFFWdjt0kpCXruB27LaZzBWvQoEFxh0C3bjBwILz7LmxOvr9gFgwYYPW3S3f6ZI7Io8a9c865dK1atYpVq1LddC27RGDkSLuWOJs6drS6hw7Nbr2Z4InZOecK0MyZM5k5c2bcYVBSAsOH2/OQs6VVK0vKIjbWnW+8K9s551yLOvJI2LIFKipgdfK97FrA8OH26p2PT0DHE7NzzrkWVlQEY8bY7Tlra2HNoXfbzJhhw2D0aOs+z1eemJ1zzrW4tm3h+OOtm7l9e1ixIrPXNxcXw4gRMGoUHHNMfl0elcwTs3POuaxo2xbGjoUuXeCII2DxYti5s+HPNaRHD2shDx9u3eb5nJTBE7NzzhWkCy64IO4QUkq0bHv1suS8cqV1bVekenp7A7p1s0uiBg+2xFxSkvFwY+GJ2TnnClBcj3tMV/fuUFpqLdzycnutX28Pvdi1y8aikxUVWfLt2hX69rWbhwwYAD175n8rOcoTs3POFaDly+1heUOGDIk5kroVF1tiHTAAtm2DTZusa3vHDnvt328JuqjIusG7dj3wQIo+fexxjoXIE7NzzhWg2bPtSa25nJijunWzF1gyrqyEqqoDiblNG5s0djjwxOyccy6ntGpld+46XBVQr7xzzjmX/zwxO+eccznEE7NzzjmXQ3yM2TnnCtCFF14YdwiuiTwxO+dcAerZs2fcIbgm8q5s55wrQKqKqsYdhmsCbzE751wBeuWVVwAQkZgjcY3lLWbnnHMuh3hids4553KIJ2bnnHMuh3hids4553KIT/5yzrkCdNFFF8UdgmsiT8zOOVeAunbtGncIrom8K9s55wrQwoULWbhwYdxhuCbwFrNzzhWg1157DYAxY8bEHIlrLG8xO+eccznEE7NzzjmXQzwxO+eccznEE7NzzjmXQ3zyl3POFaBx48bFHYJrIk/MzjlXgDp27Bh3CK6JvCvbOecK0Pz585k/f37cYbgm8MTsnHMFyBNz/vLE7JxzzuUQH2N2zjl32KmpgS1bYN8+aNUKOnWCLl3ijsp4YnbOOXfY2L8fVq2C8nJYuxYqKy0xd+0K/fvDgAHQt6+9F5dmJ2YRuR34kqoObH448RORCcBkoI2qVsUcTkETkTHAUar6f4VWr4i0Aq4A/qqqG1qqnkwRkZXAo6r6nWaU8Q9gtap+LkNhOZdRe/bAggWweDGsWWOt5aglS2DwYBg1CkaOhKKYBnu9xezi9FfgISCriTlL9X4ImAIMbsE6nKvT+PHj4w4hp1RVwZtvwhtvwPr1qZepqICFC23ZNm1g+PDsxpjgk79cnOLqLMpGvTF2hDkHbdq0oU2bNnGHkTM2bIBly+pOygm1taAKK1ZYCzsODbaYQ7ff3cBpQCdgDfCAqt6btNytwA1AO+Ap4HpV3RF+Vwt8GRgHnB7KeFhV74p8/v+3d+dhUlRXA4d/LLI57CjDgAoKHNEooI47ouIWlyyE4K6oIUYlxhXRxH2NW9yJiUZE0agEkuCnqKAOoDEibhHhwMCAwACCbDOAwsB8f5zbUNP0MN2zdTOc93nmaai6devW0nXqLlV9KnAzcACwDvgXMFRVV4b580K+JwM5wDnAMOA/QFvgbOAH4DHgFeBp4DDgm1CW8SGfPYD7gX5Aa+Bb4CVgmKpuSmJ/jAAaAEuAC8O/JwOXq2phSNMqrOPnYX98DtykqlPC/NuAE4D5wBlhu/oDf1DVxyPr+l3YJzmqukFELgCuAboDK4CXgZtV9fvIfh6sqs9E8lgIPKOqt4lIA+CesO92B+YBj6jq8HK2tdz0ItIzbFc/VX03sswzwN6qeryIHAY8CPQGSoD3gKtUdX44nh2BW0VkkKp2FpH6wPXApUAHIB94NLY9InJsyOM44Clgb+AL4HxgAPBboClWE74ktl/itinReuex7bn1EXYMTwvbvhIYB1ypqmtDXl3C9vUDNoey/Q7oCrwTVlkgIrer6m0JylIayn4xcASwGLgKKA3r3guYBgxS1TlhmT2Au4ETgVbAVLZ/br2GnZ93AbcCtwMtgXeBK1R1YaRI7UVkNHAK9l16Ebg21qUjIkcAdwKHhDK+DVynqgvity2k3w87f47Cjssk7Dv9VSTNVWGfZQNTQlkvDsdlLNBWVY+JpN8DOw+PU9VJidbrzNSpUwHIzc1Nc0kyw6JF9heVnz+GqVPvo7i4kKysHHJzh9G1a382b7a0ixdb03Zt226NWUSaYReYVcCRwP5Y0PujiBwSSdoROCn8nR7SvhKX3QPA80BP7GJxh4j0Dev5OfA6MAE4GDgX6AO8HS7WMUOwC/cJ2JccLFAtCPk+gV143gAeDXkVACNDnx/YxbVtKKtgF8DrsCCarIEhj77YhfUY7GIZ61t8E+iGXRhzw3a9JyLHRPI4ClgN9MJufF4F4vvmzgNeCkH5KuAZrP+7JxaEzsNuKpI1BPglcCYW3B8HngoBL6X0qvoFFjQujCUWkaYh/XMhqL8O5GE3W8cDnbBzAGy/LAYeCv8GuBe4HLtQ7w/8Ebg/3PRF/QkLZodjQfM/wIFhHb/Bjs+vytmmROuNbWv03HoeC0D9sWN5Nba/Lwvb2hILIi3CMscCu2Hn14ehDACHYsG7PI8AT2L76GssGN4KXITdgOwV9gMi0gL4ANgnlCsXmAVMjPs+Rs+t2A307th35eyQbw72/YpWqS4K23RA2BdXYvsZETkUeB+YG/I/DdgTmCwizeM3SkT2CvuhHvZd64PdvEwWkT1DmiuwY34Xdk6/H7Y95lngaBHZOzLtQuw7PTnBvnQR06dPZ/r06ekuRkZYv95qzMXFW6fl549h8uShFBcvAkopLl7E5MlDyc8fA1jNetmy9JS3ohrzrtiFY3ik9nsLcAP25f0kpNsAnKmqS0OaIdiXvoeqzghpRqrqi2H+TcAVWADPw2q+r6vqzSGtisi5If9TsEAL8LaqvhkrnIgAzIjVvEXkYeyL/Yqqjg3TngT+DWSLyCrgBWC0qs4P2TwqIkPD9oxOYp8BFAOXqupGYIaIjMIuPmC1p8OB3VU1dljvCLXHa9h6QwFW2/0ulPM54AMR6aaqs8U27hDg0hDshwF/VtVHw7Kzw03LaBHZT1W/TqLcXYC1wHxVXQQ8ISIzgRmVTP8s8ICIXB5qkf3D9H9gAast1rIwX1ULRORMoD2Aqi4Tkc1Acfh3FhaQz1PVcSGfuSLSAbhRRO6LlOt2Vf0o7Lex2E3KJaEMM0Tkeux4biN+vZFZ8efWO8DkcAMCME9ELo/kexbQBjgrcgwvAQYDTbAaNsAyVY1cDrbxoqr+Iyz/Fyzg3RTZvtewm12wG9bdgcNUdXGYPxgL0EPZejMAZc+tI4BdsJr31DDtPGAmdt7G+tr/FTm/CkKLzSHAX4BrseN+qaqWhjx+gQXq87EWjKjLge/D/lkf0g8M6YeE8l4PPKGqz4Zl7haRg7AbarAb3MUh/9vDtAuA52JliAmtVK3iylAnBqS6rQYMGFCp5UpKYOVKWLt267SlS6exefOGuHTrycu7lpkzrb4zcaKN1q6s0aOTDSllbbfGHC5cw4GzRGR4uFjFmr4aRJLOiQXl4L/hc/9odpF8S4E1QKMw6UDKBixUdRoWAA+MTJ6dqJiRf68Ln/mRaevDZ+NwgXgS6CMij4vIG6GpNydueyoyNwTlmNWRbekdPgtEpDj2hzU99ogsszJ24QRQ1Q/DtsRGbJwPfKmqn2IX4/bE7SPspgbK7qPteRILmAtEZJqI3At8G3fsUkn/EnYO/SL8fxB2U7QudEHcj3UtLBORV7EWhs/KWVcPrNl/ZNx+uwPrcmgfSRt/zBfHmpeD9SGvVMSfW8OBvUXkYRH5t4jMwZqbY+fJAUB+3DGcrapDYzexSUrm/I1ty4FhnYsj6yzFao/Rc6DMuRXLOxaUw3KK3TxEb2BmxS2zEmuCjq17SjQghq6bfBKffwcCn8SCcki/DrvZPlBE2mKtAf+JW25yJP0mrOXiAgAROQprLRiZYH1XYTXp6J/Xqh1gjz7FP/4UH5TjpydaprZst8YsItlYkP0W6/N9C+vTWhiXNL5vNnbx+iEy7Qe2VdFm149bbl2CNBsTTNucKDMR2RX7sjbDmo5HYNs3pYJyxNvettQP5exVQVkTbcsI4BIRuR3r53w8QZqo2I1VovLEbGmqDDXxrliz6wnAqcB1InJRrDUjqqL0qrpaRP4BnC8iE7Cm5KMiyw8TkafCcv2w7oUrReTI6AU7blvOAb5iW8uwrgfY9pgnPN4p2nI8QkvE/wE/wpqWX8ECyjOR9Im/1alL+vzdjmS+J4ke/WtA2e9uojEWqX5HKxJLHytPRfn/DRgWgvL5wDvl9Gk/gn1/ojrhwblOqWwNdPNm+PBDmDRp6yNSL798aGjGLisrqyOnnz6a7Gzo29cenaptFY3KPgdoBxylqnep6j+x5jso+4XaJ66f6WhscEiyHRxfYv20W4hILhZAq7OT5GSsRttPVW9R1VeBIqw2Vl33Rv/Dyt1MVfNjf1hz61kVLDsSaz7+DXZRGQUQaqhLidtHWA0Utu6jDVgNFwARaY01J8f+fw0wQFXfUdUbVLUnNgjogkSFSTL9s1jgvtiKuqUJtoeIPI3VsP+sqr/E9n8vtrYqRJsjZ2JBas+4/dYXGwBXHcE3prSC+b1CWc9U1WGqOgprgu3K1vNkBtA17GMARKS7iKwMg54qWkdlfBnW2SGyzvpY/21F35MWItI9stz+2LnyaQrr7hMZq4GIdMRqsInW/SVwSBinEku/K9Y0Pl1VV2MD1A6LW67M/8M5MAkbu/AzbIzFNlR1larOi/6xbQXC7aTq17eXh3TosHVabu4wGjZsWiZdw4ZNyc0dBlj6nJzaLGWkHBXMX4A1ZZ0pIpOwGsufwrxoU2Fj4DURuRELBI9jg5bmJlmO+4AxInInVkPZAxvI9QkWCKpL7It6roi8EtZzL1arTLXpszxvYRe7l0XkSmwE6UVYYP7p9hZU1UIReRsb7POGln0xxX3YQKg5WG3uAKwG+k9VjTVBfggMFpE8LEjfQ9kaWTvgJhFZi41m3hc4CGuyTiSZ9HnYBfYmyg7cWYbdiDQJ/cObsOC9ChvkBHZT1F1EcsK2D8f641dhrRhHYE3hw1V1cxhTUB3KrDfB/CVYjW6giBRi5/TvsZHDsfNkFHbD8KKI/B4L2I9hTeIzgKyQrreIrAyBqKpGATdi4wquw7qDrsa+lxcnsfzIMP6jAdZU/zE2kjwZD2HH5GkReRQb2f0Q1gccP9ATrM/5Uux7cCu2f+7AunyeDmnuAx4SkenYoLYzsH7yb+LyejYssx74Z5Ll3ekNGjQo3UXIKDk50LmzDehatw66drUhMYlGZXfoAJ06Va1/uSoqqjGPxvoJH8AuNo9gX5JJlL2z/YytQXQ0FpwuTbYQYaDWAGzgy5fYAK13gRO1Gt++paofYwOwhmA1tBFYYHmZbe/cK7uOTdiAmo9Cvl9h2zUwMqhpe54DmhPXLKeqj2DB/TIssD0S0p4dSXYZdqH8ABsw9w5b+/vBAucz2I3TLGxQz9PYqNhEKkwf+hxHYDc3L0SmL8cG7u2F7YvPsNrViaq6KiR7EPgx8D8RaYgdmweA27Dz7U7s4j20nPJVVvx6ywjB+kKsCX4Gdk4vwm5Kc0WkfugvjQ34m4KNvP8G+EnYJ59jgw7/HrajykLfdR/sGI/Hju3ewLHR/uPtGImNlH8Lq+WeqqpJtUSE785JwH7Yd30cMAdrTdvmpiMMrjwGu5GZgo24LgWOiA28VNU/Y08z3I19T07GzqX4boLR2I3SS6qaSrO5c1s0aQIHHggHHbQ14Hbt2p+zz/6YwYMXcvbZH9OtW386dYKePWH//befX02qV1paEy1ubmciIn8G2qtqKo+cuVoiGfqaWbF3F3wdmp1j0/4C7KOq/SLTOmKtMgdHRsknk39noGDixIl06rTzDdD+8MMPATjyyCPTXJLMsmoV5OeX/67snBzo3h0aNao4r6pYuHAh/fr1A+gS/Q6Av5LTVYGInICNpr4Qq+04l4pzgf1F5DdYK0Bf7FnxywFEpBPWkvVrYFIqQdnBrFnWw+WBuaxWreCQQ2D1antWOfbrUs2aWR90kybpLqEHZlc1F2P9gveqv4XJpe63wMNYv3ErrGn8KlUdEea3w5q257L1GXnnqkXLlunrQ66IB2ZXaap6TrrL4CoWAt2INBdjG6q6Anv2vbz5n2PjLZzbqfiPWDjnnHMZxGvMzjlXB/kvS+24PDA751wd5L/HvOPypmznnHMug3hgds65OigvL4+8vLyKE7qM44HZOefqoIKCAgoKCtJdDFcJHpidc865DOKB2TnnnMsgHpidc865DOKPSznnXB3UrFmzihO5jOSB2Tnn6qCBAwemuwiukrwp2znnnMsgHpidc64OmjBhAhMmTEh3MVwleFO2c87VQQsXLkx3EVwleY3ZOeecyyAemJ1zzrkM4oHZOeecyyDex+ycc3VQixYt0l0EV0kemJ1zrg7q379/uovgKsmbsp1zzrkM4oHZOefqoPHjxzN+/Ph0F8NVgjdlO+dcHbRkyZJ0F8FVkteYnXPOuQzigdk555zLIB6YnXPOuQzifczOOVcHtW3bNt1FcJXkgdk55+qgM844I91FcJXkgdk552pJSUkJxcXFFBcXU1JSAkCDBg3IysoiKyuLXXbZJc0ldJnAA7NzztWgTZs28e2331JYWMiKFStYs2YNxcXFbNy4EYCGDRuSlZVF8+bNadOmDR06dKB9+/ZVDtLjxo0DvOa8I/LA7JxzNaC0tJRFixYxd+5cCgsLKSwsZM2aNdtdZtdddyUnJ4eOHTvSuXNn9txzT+rXr9wY3e+++65Sy7n088DsnHPVbP369agqc+bMYdasWfzwww9JLbd27Vpmz55NQUEBS5YsYfny5ey7775kZWXVcIldJvHA7Jxz1ai4uJjPPvuM6dOnV/rtWyUlJcycOZNly5axZs0aevfuTevWrau5pC5TeWB2zrlqsnbtWj799FM+/fRTVq1aVeX8vvvuO6ZNm0ZpaSkHH3wwrVq1qoZSukzngdk556rBpk2bmD59Ol988UW1BOWYtWvX8vnnn9O4cWMOPfRQGjVqlNRy2dnZ1VYGV7s8MDvnXDWYN28es2fPrpFBV0VFRcyaNYs2bdqw3377JbXMKaecUu3lcLXDX8npnHNVFBu0NXfu3Bpbx4IFC8jPz2fFihU1tg6XGTwwO+dcFS1atIgFCxawadOmGltHaWkp8+fPZ+HChUmlHzNmDGPGjKmx8ria44HZbZeI3CYiyV0JdgAiMkhESkUkYTeOiIwQkSm1XKYsERlSxTxqvdzVIf54iMheInJWusuVipKSEgoLC2vl949XrFjB4sWLWb9+fYVp16xZU+Fz0y4zeWB2rqzfAT+p5XXeAFxXxTzSUe7q8ArQQVVLwv9fBHaoztGVK1eydOnSGq0tx5SWlm55vtnVXT74y7kIVV2dhtXWq2oGaSp3lanqeiBa/avyvqhtRUVFKddM8/PzmTp1KsXFxWRlZZGbm0vXrl2TXl9RUVFliup2EB6Yd3Ii8iPgPuAoYFdgETBcVe+PS3cTcDXQGPgH8DtVXRPmlQK/BgYCR4c8nlfVOyPLnwrcDBwArAP+BQxV1ZVh/ryQ78lADnAOMAz4D9AWOBv4AXgMq2U9DRwGfBPKMj7kswdwP9APaA18C7wEDFPVCqs0IjIC6KqqR4vIscB7wKnAA8DegAK3qeq/Qvr3gU/Cun4BfA+8A1yuqkUhzU+Bm4AfAQ2Ar4E/qOobInIb8PvIfjxOVd8XkaOAe4FcYDkwPmxDwiG/5ZT7OOCpUO4vgPOBAcBvgabA/wGXqOr3IjIIuA0YCvwRyAamAIOAa8NnKVajvUpVS0PZf6WqnSLl+BXwV1WtF/4/DxgOHIod2xXAP4FrVLUkrPc5YBdgAnYeHiUiFwI/C2m7quqcyDomAN+o6sWJ9kVtSzVQ5ufnM3ny5C0/YlFcXMzkyZMBkgrORUVFFBcXV66wbofggXknJiLNsCAyETgSKAEuAf4oIu+q6ichaUfgpPDXHPgrFhx/HMnuAWAIcAVwEXCHiExS1TwR+TkWdO/GLvB7Ao8Db4vIYaq6OeQxBLsYLwVmYoH5GuAuoCdwAXA7cCHW9KvAw8BIEWmvqqXAOCwYnwSsAc4AHgH+C4yu5K56ALgSWIbdxLwgIh1jgRcLdA9hwScX+BuQD9wuIgcDY7CAdxbQErgnlLkj8GCYNiAsu0JEemFB6h7gV0AbLFjmiUhvVd2YZLn/BFyM3SyMxW5y3gGOx/bnC2HaEyF9Dnb8fga0wPbll8BI4HCsifnRkMfrSZYB7JgNxY7nqaFc04Dn49L1B94E5mFN88uBJdhxvxW23HgdBxwbXVBEWgHxb9/oRAUGDBiQwmYktmbNGlauXElpaWlS6ZcuXcrmzZvLTCspKSEvL4+ZM2cmlcfEiRMrfNlI+/btAXj99fIP1ejRlf1KuJrkfcw7t12xoHW5qs5Q1dnALWHeAZF0G4AzVfUzVZ2EBdBTRKRHJM1IVX1RVWdhtcMiLNiDXZBfV9Wb1bwDnAscQtn+xLdV9U1V/VRV14VpM1T1TlWdiwVhgFdUdayqfg08CewGZItIUyzYDFbVz1V1rqo+ChTGbU+qblHVd1X1f2H/NMdqvzGqqjep6ixVHQW8Hdn2TcCVqvqQqhao6udYYGoL5KhqMbAW2KSqS1R1A3bTMTFs9yxV/Qj4JbAfcFoK5b5dVT8K6xyLBdtLwrH+OxZ0o/tlF6z14QtVnQy8izUzXxvK8RhW4011X76tqo+p6uxwPKZH9s8WqroC2AisD/uiBLspOF9EYk3cFwBzQ/mirgIK4v7i02SE+KBc0fTKWrp0KUuXLq3WPF3t8BrzTkxVl4nIcOAsEekNdMVqUmBNrjFzVDX6Df9v+NwfmBHLLpJvqYisAWKvKDoQa8aOrnuaiBSHeW+EybMTFTPy71iwzo9Mi/VPNlbV9SLyJDBARK4D9gn558RtT6qiZYj15TYqZ34sTUcAVf1cRFaKyDBgX2wf9wrpyitTb6Bb2D9R9YAeWPNuquVeByxW1bWRaeuxrontLTM30qJR3jKplANs/yT3+ip4Fqtt9wEmYYH5uQTpHgFGxE3rRAXBuTpqjFOnTmXChAlJ/1DFyy+/nLApOisri9NPPz2pPPr160efPn1SKqfbcXhg3omJSDYWZL/F+nzfAqYC8Y9HxffNxgJK9EqU6KpU0UCe+nHLrUuQJlGzbcKqhYjsil2ImwGvYhfq/2J9pVVR0baVOz/0976F9edOxvpod2X7wbU+8HfgjgTzUnm7RPy+S6ZKlsoyidpuE/2IcGXODQBUdVZ4DOx8EdmA3djEN4GjqquAMu/BFJFkVlFlTZo0oWnTpkkH5tzc3DJ9zGC/yZybm5vU8k2bNqVp06YVpnv11VcBGDhwYFL5uszhgXnndg7QDugWmlARkVgzZfTCuY+INI/0qR6NXZSnJ7meL4FjsP5UwnpysQCabB7JOBmrbXZS1UVhPW2A9qRvtO81wCRV7R+bICK/Df+MlSk+wP0Pa42YE/rNEZFO2IC324GPa7TEydsANBeRerFyYoGzKhIF+2exfv41WLP4oiquo1o1b96c5s2bJ/1+7NgAr8qOyo6tryLr1iW6z3U7Ag/MO7cF2OjcM0VkEiBY/yeUba5sDLwmIjdifaOPAy+Fft9k3AeMEZE7sRrjHtiAo0+wfszqEqvpnysir4T13IvV4lJtfq0uC4D+ItIn/Ps4IDZaPVamIqC1WBVvPnYDMwUYLiKPAVnYoKtsbER3pvgQ67e+UURewkZUX1DFPIuALiKyl6rOD9New0bjX4YN/MsozZs3p2XLlixYsCDpZbp27Zp0II7XsmXLpAKz23H54K+d22js0aIHsL7iR7DaySTsUaSYz9gaREdjTbOXJrsSVR2LjTo+Das9vxDyOjHyYokqU9WPsRrqEGxU9wggD3iZsttTm27BAtg44HNgMDZSen2kTK9gQftL4CdhO07Cas3TsJHKhcDxYbBYRlDVPOxRryHY+XM+cH0Vs30c64ufISI5YT1rsa6J9ViXS0Zp3bo12dnZSf/qU1XUr1+f7Oxs2rVrV+PrculTL9kh/s45ly4iMh4boX91Cst0BgomTpxIp04VPjlVJXPnziUvL4/58+dXnLgKsrOz6du3Lz169Kgw7YgRIwAYNGhQjZbJVc7ChQvp168fQBdVnRed503ZzrmMJSI/wUbWH4c9Y52RcnJy2GuvvVi8eDEbNmyokXU0aNCAzp07J32T0aVLlxoph6t5Hpidc5nsBuwRsSuib//KNE2aNKF79+6sWLGCr776qkbWsffee9OtW7ek+5f79u1bI+VwNc8Ds3MuY6nqUekuQ7I6duxI165dWblyJYsWVe/A8Xbt2tGtWzc6d+5crfm6zOSB2TnnqkG9evXYb7/92LBhA5s3b2bx4sXVkm/btm3p1asXP/rRj2jQIPn35IwaNQqAc889t1rK4WqPB2bnnKsmjRo1omfPntSrV4+ZM2dSUFBQ6Vdt1qtXj06dOtGjRw969uxJs2bNUlp+48ZkX6nuMo0HZuecq0aNGzemd+/etGzZkrZt26KqrF6d2q9yZmVl0b17d/bZZx+6detWK49iuczhgdk556pZw4YN6d69O7vtthvt27ensLCQwsJCli9fXm5NtmHDhrRp04acnBxycnLo0qULu+22Wy2X3GUCD8zOOVdDWrcPGzv1AAAL00lEQVRuzcEHH0y3bt0oLCxk1apVrF69mqKiIkpKSigtLaVhw4ZkZWXRsmVLWrVqRYcOHWjdunW6i+7SyAOzc87VsBYtWtCiRQsANmzYwNq1a7f8iEWDBg3Iysqq9ubq7t27V2t+rvZ4YHbOuVrUqFGjWukzPvLIbX7y2u0g/F3ZzjnnXAbxwOycc3XQiBEjtrwv2+1YPDA755xzGcQDs3POOZdBPDA755xzGcQDs3POOZdB/HEp55yrg/bff/90F8FVkgdm55yrg3Jzc9NdBFdJHpidc3VVA4AlS5akuxxpEXuzWMOGfpnPRJHzcpvf8vQj5pyrqzqA/x6xy3gdgDnRCR6YnXN11VSgD7AY2JTmsqRDJ2Aytg8WprkslbGjlx+2vw0NsKA8NX6heqWlpTVfNOecc7VKRDoDBUAXVZ2X3tKkbkcvP1R+G/xxKeeccy6DeGB2zjnnMogHZueccy6DeGB2zrm6aRVwe/jcEe3o5YdKboMP/nLOOecyiNeYnXPOuQzigdk555zLIP6CEeecq2NEpD5wK/AroDUwBbhcVfPTWrBKEJEbgdNU9eh0lyVZItIcuAP4OdAOmAncoar/TmZ5rzE751zdcwtwGTAYOBwoAd4SkSZpLVWKRORy4O50l6MSRgCnYzdGvYAxwFgROT6Zhb3G7JxzdYiINAauBW5Q1TfCtLOwV5P+EnghjcVLiojkAE8DxwGa5uKkRESygf7A6ao6IUy+R0T6AZcA71aUh9eYnXOubukFZAHvxSao6hrgU+CYdBUqRQcDRcCBwH/TXJZUrQV+DEyKm14KtEkmA68xO+dc3dIxfMb/aEIhsEctl6VSVHUcMA5ARNJcmtSoahEwPjpNRA4HjgeuTCYPrzE751zd0ix8/hA3/Qdgh+pjrgtEpAcwFqv5P53MMh6YnXOublkfPhvHTW8MFNdyWXZqInIMNiJ+MTayfGMyy3lgds65umVB+MyJm57Djvu7xjscETkXeAeYBvRV1RXJLuuB2Tnn6pYvgDXAsbEJItICOAjIS1OZdioicg42+v1VrKZclMryPvjLOefqEFX9QUSewB7RWQIUAPcBi4B/pLVwOwER6QT8FRsVPxRoGxnAtiGZmrMHZuecq3tuARoAfwF2BSYDp6jqhrSWaufQHxuAdzw2Ej7qA6DCN5j5r0s555xzGcT7mJ1zzrkM4oHZOeecyyAemJ1zzrkM4oHZOeecyyAemJ1zzrkM4oHZOeecyyD+HLNzzpVDROYBU1T1vBrKf0/gQyBXVRdHph8OXIM989oWe9fyZOBRVf0kLo8RwIUJsl8LzAdeAe5W1U0h/WXAj1X1J9W+QRUQkYOAe4FcoAR4HbhRVZdG0hwCTE2wuKrqviFNNvActn++Aa5W1bcjeXTAfsf5SFX9Ksmy1cf244XA/tgPfszF3uD1lKqui6QdAZygqp2S2/LUeI3ZOefSQETqAX8DnogLytdiwXp34EbgVOB2oAvwkYgk+unAZUCfuL/zgC/DsndF0j4N7CUil1T3Nm1PuNn4COgAXAH8GtgH+I+ItIwk7QVsBo6j7PacE0nzANAO+CUW3EeH147G3Ab8I4Wg3AT7qcansH32a2Ag8AZwBzBeRHZNYXOrxGvMzjmXHj8DDgG21FxF5FTgQeBeVb0pmjjU0p4EHhGR6ao6MTJ7g6pOSbCOf4pIZ+ASLMijqptF5C7gKRF5OVoTLI+IDAKeU9V6KWxfvJuwd3gfp6rfhXzfwmq2Q4Hfh3S9gFmq+v528joIu6EZLyJ5YfnuwCcisi8WxPdLoWwPAn2B41X1g8j0N0VkChb8b8DeqFbjvMbsnHNJEpEmIvIHEZkpIt+LSL6I3CgiDeLSXScic0VkvYhMFpFTRaRURI6NJLsJGBMXGG8FZrE1SG2hqqXA1dgvRN2aQrFXJ5g2DmiE1Qxry77AB7GgDKCq67HfKT4tkq4X9kMc21MAHB9q2mcAm9j6q1r3AcNVdUF5C0eJSFtsPzwbF5RjZfw/rPk9P5n8qoPXmJ1zLgmh6XkccCRwN/AZ9gtOdwLdgItDuluwptSHgXeBk4HRcXnti9WWb4lMa4f1vT4cgvA2wg9UjAV+KyJto0FORKLX8/pYU/h5wEmhLNF8vheRccC5wCPlbG99tlbe6idYx6byylmO5cBeCabvA+wd8q8HHAisF5H/YkF6JdaffEvk94z/gNViVwEbgetUdamIHAUcA1yUQrlOBHbBjm1C8a0XNc0Ds3POJecU4ATgQlUdGaa9KSLrgNtE5BFgDjAMq7FdF9K8ISJNgcGRvI4Pnx9HpnUG6mEDjrYnP6TbE4gF5o5YgIo3H7gZuD/BvKnA2SLSSlVXJZj/N7YdVBZdx0XAiArKGp/fX0XkcaxWuwlrAdgPC4xgQbo5Vrsehv0i1glYU/We2I0Eqvq5iOyNBfTFqhprFbgfuAeoJyKvYUH+HeD6UDtPZI/wWZDCttQoD8zOOZecY7FBSX+Pm/48VkM+FqulNgVei0szirKBuQtQHK3xYsEWEgfYqNj8aH/vt2xtDm6JBbJc4IrQFJvIPKwmvCdW84x3G/BE+PfpWPN5bmR+QahVN4pbboOqbo7PTFWfCQO07gCGAKXAWOwXsGI13CXYDdB0VV0Ypr0vIt8Dd4rIPao6PeS3AZgZy19E+mNB9gngRSzw/zTkfxvWR5xISfhsUM78Wud9zM45l5w2wMoEP524JHy2AnYL/14Wl2Zp3P9bYo8zRc0Ln10qKMc+4fObyLSNqvpJ+JuIBdKvgTGheTeR2PpbJpqpqvNiecbKFlnHJ+Gm4hhgfdzfMeUVXFUfBloDPYBsVf1FWP+KML9YVd+KBOWY18Nnr0T5hib2e7DWgVhA/pOqzsQGzA0sr0xYqwJYi0VCItIhjNyuFR6YnXMuOSuA1iISX0PMCZ/LsYFZAO3j0uwe9//lWCDfQlWXYY9J/SLURLcRBpn1xwZRLS+voKEvdhBWwx9ZTlBpEylLZU3DatHRv2mJEorIISIyQFU3qupMVf02zDoY+DSkOUBEhohIs7jFm1ZQ1sHABuyZ47ZYa3CsNWIlkL2dbXgXa4U4bTtpRgFz4vrYa4wHZuecS04eds08K2567OUjU4DPsVHQv4hLMyDu//OAxiKyW9z0W7CBZIn6hAEewmrUd5UzfwtVzcf6cvfG+mvj7YEF7vjaadJUtSiuFv2JqhaVk7wvMCqMggZARE7EXubxzzCpM/A49ihZ1NlAMfBJ3HTC88W3AsNCE/pyrNbcISTJYdsWi+g2rAKeAS4RkcMS5P8zrJtipKqWxM+vCd7H7JxzyXkTeA8YLiIdsVHZfYHrgZdU9UsAEbkf6w9dG9L3BX4T8oj1vb4VPo/G+lkBUNWJInI18JCI9MZGIy8COmGjvvtgA5nGJ1nmPwIXADeIyPOqGh3gdDRW8y4vkG6hqiNIbaBXIi9iz1KPFpEHsG16APggzAN7ycd/gCfDKPXZWLP8ZcDQuD75mOuAGar6RihriYiMB24XkYfC/LEJlou6ERsl/56I/BmrRTfABp79GpiE9VPXCq8xO+dcEsKjQacDw4HLgX9hzcq/x4JfzL1YDe5srCZ4BBa8wWp9qOp8bET2Ns2nqvoo9kjWEuyxrDfD5wLgCFV9KIUyfw9chb1e8k+x6WGU+LFsO0itxoTXbp6IDVp7BWsdeA57PeimkCbWpPwi9krSsVif9WBVfTA+TxHZHbiWbQd2XRY+RwHTqeC57zCq+9iQri/WJP4CdhyGAier6g8pbXAV1CstTeUxNOecc+UJfZAXABNU9ZvI9CHY88JtY4/2iMhPscCRo6prarmcgwjN3Mm8+cvVLg/MzjlXjUTkK+wRnHuw/s4DsEeEXlXVwXFpJwNvqWqFfcbVWL4G2PugH1PVp2trvS553pTtnHPV6wzs+drHsWboIVhf6mUJ0l4AXCYiOQnm1ZTfAPM9KGcurzE755xzGcRrzM4551wG8cDsnHPOZRAPzM4551wG8cDsnHPOZRAPzM4551wG8cDsnHPOZZD/B4xF2e4A8ZYZAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 216x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(3,6))\n",
    "\n",
    "plt.errorbar(root_KO_df.loc[terms_plot]['log_OR'],np.arange(len(terms_plot)),\n",
    "            xerr=[np.subtract(root_KO_df.loc[terms_plot]['log_OR'],root_KO_df.loc[terms_plot]['log_OR_CI_lower']),\n",
    "                   np.subtract(root_KO_df.loc[terms_plot]['log_OR_CI_upper'],root_KO_df.loc[terms_plot]['log_OR'])],color='k',fmt='o')\n",
    "\n",
    "color_temp = root_KO_df.loc[terms_plot]['MPO_term_type'].map({'brain':'blue','heart':'red','neg_ctrl':'black'})\n",
    "\n",
    "sns.scatterplot(x=root_KO_df.loc[terms_plot]['log_OR'],\n",
    "                y=np.arange(len(terms_plot)),size=root_KO_df.loc[terms_plot]['num_genes_in_term'],sizes=(200, 2000),\n",
    "                alpha=.5,\n",
    "               hue=color_temp.tolist(),palette={'blue':'blue','red':'red','black':'black'},legend=False)\n",
    "\n",
    "plt.yticks(np.arange(len(terms_plot)),root_KO_df.loc[terms_plot]['MP_description'])\n",
    "plt.xticks([0,1,2])\n",
    "plt.xlabel('log(OR) +- 95% CI')\n",
    "\n",
    "plt.plot([0,0],[-.8,len(terms_plot)-.5],'--',color='gray')\n",
    "plt.ylim([-0.8,len(terms_plot)-.5])\n",
    "\n",
    "plt.gca().invert_yaxis()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 18. Compute the enrichment of phenotype(s) in NetColoc subsystems.\n",
    "\n",
    "Some phenotypes may have stronger associations with NetColoc subsystems than with the root node. In this step we calculate the enrichment of selected phenotypes in each NetColoc subsystem.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:49:27.671993Z",
     "start_time": "2022-06-14T22:49:04.578910Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abnormal nervous system development\n",
      "Genes and Terms to keep: 191\n",
      "abnormal synaptic physiology\n",
      "Genes and Terms to keep: 78\n",
      "abnormal synaptic transmission\n",
      "Genes and Terms to keep: 73\n",
      "abnormal CNS synaptic transmission\n",
      "Genes and Terms to keep: 46\n",
      "abnormal nervous system tract morphology\n",
      "Genes and Terms to keep: 62\n",
      "abnormal nervous system electrophysiology\n",
      "Genes and Terms to keep: 34\n",
      "abnormal neuron differentiation\n",
      "Genes and Terms to keep: 16\n",
      "abnormal neuron morphology\n",
      "Genes and Terms to keep: 267\n",
      "abnormal excitatory postsynaptic currents\n",
      "Genes and Terms to keep: 9\n",
      "abnormal hippocampus neuron morphology\n",
      "Genes and Terms to keep: 12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abnormal nervous system development:-log(OR_p)</th>\n",
       "      <th>abnormal nervous system development:log_OR</th>\n",
       "      <th>abnormal nervous system development:num_genes</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal synaptic physiology:-log(OR_p)</th>\n",
       "      <th>abnormal synaptic physiology:log_OR</th>\n",
       "      <th>abnormal synaptic physiology:num_genes</th>\n",
       "      <th>abnormal synaptic physiology:gene_ids</th>\n",
       "      <th>abnormal synaptic transmission:-log(OR_p)</th>\n",
       "      <th>abnormal synaptic transmission:log_OR</th>\n",
       "      <th>...</th>\n",
       "      <th>abnormal neuron morphology:num_genes</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal excitatory postsynaptic currents:-log(OR_p)</th>\n",
       "      <th>abnormal excitatory postsynaptic currents:log_OR</th>\n",
       "      <th>abnormal excitatory postsynaptic currents:num_genes</th>\n",
       "      <th>abnormal excitatory postsynaptic currents:gene_ids</th>\n",
       "      <th>abnormal hippocampus neuron morphology:-log(OR_p)</th>\n",
       "      <th>abnormal hippocampus neuron morphology:log_OR</th>\n",
       "      <th>abnormal hippocampus neuron morphology:num_genes</th>\n",
       "      <th>abnormal hippocampus neuron morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C416</th>\n",
       "      <td>12.116163</td>\n",
       "      <td>0.975982</td>\n",
       "      <td>68</td>\n",
       "      <td>ADNP AR ARID1B AUTS2 BCL11A CADPS2 CHD8 COL1A1...</td>\n",
       "      <td>11.212276</td>\n",
       "      <td>1.016254</td>\n",
       "      <td>56</td>\n",
       "      <td>ANK2 ANKRD11 ARID1B BCL11A BEX2 BICD1 CACNA1E ...</td>\n",
       "      <td>10.757287</td>\n",
       "      <td>1.001272</td>\n",
       "      <td>...</td>\n",
       "      <td>72</td>\n",
       "      <td>ALCAM ANK2 ANK3 AR ARID1B AUTS2 BCL11A CADPS2 ...</td>\n",
       "      <td>8.307267</td>\n",
       "      <td>1.626520</td>\n",
       "      <td>15</td>\n",
       "      <td>BCL11A CCKBR EMX1 GABRB3 GRIA2 GRIN2B NRXN1 PT...</td>\n",
       "      <td>7.757653</td>\n",
       "      <td>1.684923</td>\n",
       "      <td>13</td>\n",
       "      <td>ANK3 CTNNB1 EMX1 FOXG1 GABRB3 GFAP GRIA2 HIVEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C514</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0.870249</td>\n",
       "      <td>1.727414</td>\n",
       "      <td>1</td>\n",
       "      <td>ANKRD11</td>\n",
       "      <td>0.875077</td>\n",
       "      <td>1.734047</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>ALCAM</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C515</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C525</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C417</th>\n",
       "      <td>6.171345</td>\n",
       "      <td>1.046484</td>\n",
       "      <td>28</td>\n",
       "      <td>AR ARID1B BCL11A CREBBP CTNNB1 DPY30 GATAD2A H...</td>\n",
       "      <td>1.141963</td>\n",
       "      <td>0.507190</td>\n",
       "      <td>14</td>\n",
       "      <td>ARID1B BCL11A BEX2 CREBBP CTNNB1 DBN1 HIVEP2 N...</td>\n",
       "      <td>0.858088</td>\n",
       "      <td>0.431712</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>AR ARID1B BCL11A CTNNB1 DBN1 DNMT3A HBP1 HIVEP...</td>\n",
       "      <td>0.286361</td>\n",
       "      <td>0.464277</td>\n",
       "      <td>2</td>\n",
       "      <td>BCL11A PTEN</td>\n",
       "      <td>1.184225</td>\n",
       "      <td>1.088204</td>\n",
       "      <td>3</td>\n",
       "      <td>CTNNB1 HIVEP2 PTEN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      abnormal nervous system development:-log(OR_p)  \\\n",
       "C416                                       12.116163   \n",
       "C514                                       -0.000000   \n",
       "C515                                       -0.000000   \n",
       "C525                                       -0.000000   \n",
       "C417                                        6.171345   \n",
       "\n",
       "      abnormal nervous system development:log_OR  \\\n",
       "C416                                    0.975982   \n",
       "C514                                    0.000000   \n",
       "C515                                    0.000000   \n",
       "C525                                    0.000000   \n",
       "C417                                    1.046484   \n",
       "\n",
       "      abnormal nervous system development:num_genes  \\\n",
       "C416                                             68   \n",
       "C514                                              0   \n",
       "C515                                              0   \n",
       "C525                                              0   \n",
       "C417                                             28   \n",
       "\n",
       "           abnormal nervous system development:gene_ids  \\\n",
       "C416  ADNP AR ARID1B AUTS2 BCL11A CADPS2 CHD8 COL1A1...   \n",
       "C514                                                      \n",
       "C515                                                      \n",
       "C525                                                      \n",
       "C417  AR ARID1B BCL11A CREBBP CTNNB1 DPY30 GATAD2A H...   \n",
       "\n",
       "      abnormal synaptic physiology:-log(OR_p)  \\\n",
       "C416                                11.212276   \n",
       "C514                                 0.870249   \n",
       "C515                                -0.000000   \n",
       "C525                                -0.000000   \n",
       "C417                                 1.141963   \n",
       "\n",
       "      abnormal synaptic physiology:log_OR  \\\n",
       "C416                             1.016254   \n",
       "C514                             1.727414   \n",
       "C515                             0.000000   \n",
       "C525                             0.000000   \n",
       "C417                             0.507190   \n",
       "\n",
       "      abnormal synaptic physiology:num_genes  \\\n",
       "C416                                      56   \n",
       "C514                                       1   \n",
       "C515                                       0   \n",
       "C525                                       0   \n",
       "C417                                      14   \n",
       "\n",
       "                  abnormal synaptic physiology:gene_ids  \\\n",
       "C416  ANK2 ANKRD11 ARID1B BCL11A BEX2 BICD1 CACNA1E ...   \n",
       "C514                                            ANKRD11   \n",
       "C515                                                      \n",
       "C525                                                      \n",
       "C417  ARID1B BCL11A BEX2 CREBBP CTNNB1 DBN1 HIVEP2 N...   \n",
       "\n",
       "      abnormal synaptic transmission:-log(OR_p)  \\\n",
       "C416                                  10.757287   \n",
       "C514                                   0.875077   \n",
       "C515                                  -0.000000   \n",
       "C525                                  -0.000000   \n",
       "C417                                   0.858088   \n",
       "\n",
       "      abnormal synaptic transmission:log_OR  ...  \\\n",
       "C416                               1.001272  ...   \n",
       "C514                               1.734047  ...   \n",
       "C515                               0.000000  ...   \n",
       "C525                               0.000000  ...   \n",
       "C417                               0.431712  ...   \n",
       "\n",
       "      abnormal neuron morphology:num_genes  \\\n",
       "C416                                    72   \n",
       "C514                                     1   \n",
       "C515                                     0   \n",
       "C525                                     0   \n",
       "C417                                    25   \n",
       "\n",
       "                    abnormal neuron morphology:gene_ids  \\\n",
       "C416  ALCAM ANK2 ANK3 AR ARID1B AUTS2 BCL11A CADPS2 ...   \n",
       "C514                                              ALCAM   \n",
       "C515                                                      \n",
       "C525                                                      \n",
       "C417  AR ARID1B BCL11A CTNNB1 DBN1 DNMT3A HBP1 HIVEP...   \n",
       "\n",
       "      abnormal excitatory postsynaptic currents:-log(OR_p)  \\\n",
       "C416                                           8.307267      \n",
       "C514                                          -0.000000      \n",
       "C515                                          -0.000000      \n",
       "C525                                          -0.000000      \n",
       "C417                                           0.286361      \n",
       "\n",
       "      abnormal excitatory postsynaptic currents:log_OR  \\\n",
       "C416                                          1.626520   \n",
       "C514                                          0.000000   \n",
       "C515                                          0.000000   \n",
       "C525                                          0.000000   \n",
       "C417                                          0.464277   \n",
       "\n",
       "      abnormal excitatory postsynaptic currents:num_genes  \\\n",
       "C416                                                 15     \n",
       "C514                                                  0     \n",
       "C515                                                  0     \n",
       "C525                                                  0     \n",
       "C417                                                  2     \n",
       "\n",
       "     abnormal excitatory postsynaptic currents:gene_ids  \\\n",
       "C416  BCL11A CCKBR EMX1 GABRB3 GRIA2 GRIN2B NRXN1 PT...   \n",
       "C514                                                      \n",
       "C515                                                      \n",
       "C525                                                      \n",
       "C417                                        BCL11A PTEN   \n",
       "\n",
       "      abnormal hippocampus neuron morphology:-log(OR_p)  \\\n",
       "C416                                           7.757653   \n",
       "C514                                          -0.000000   \n",
       "C515                                          -0.000000   \n",
       "C525                                          -0.000000   \n",
       "C417                                           1.184225   \n",
       "\n",
       "      abnormal hippocampus neuron morphology:log_OR  \\\n",
       "C416                                       1.684923   \n",
       "C514                                       0.000000   \n",
       "C515                                       0.000000   \n",
       "C525                                       0.000000   \n",
       "C417                                       1.088204   \n",
       "\n",
       "      abnormal hippocampus neuron morphology:num_genes  \\\n",
       "C416                                                13   \n",
       "C514                                                 0   \n",
       "C515                                                 0   \n",
       "C525                                                 0   \n",
       "C417                                                 3   \n",
       "\n",
       "        abnormal hippocampus neuron morphology:gene_ids  \n",
       "C416  ANK3 CTNNB1 EMX1 FOXG1 GABRB3 GFAP GRIA2 HIVEP...  \n",
       "C514                                                     \n",
       "C515                                                     \n",
       "C525                                                     \n",
       "C417                                 CTNNB1 HIVEP2 PTEN  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MP_focal_top = root_KO_df.head(10).index.tolist() # record the top 10 overall\n",
    "MP_full_results_df = validation.MPO_enrichment_full(hier_df,MPO,mgi_df,MP_focal_top,G_int)\n",
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:50:15.905976Z",
     "start_time": "2022-06-14T22:50:15.903648Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top ten terms:\n",
      "MP:0003861 - abnormal nervous system development\n",
      "MP:0021009 - abnormal synaptic physiology\n",
      "MP:0003635 - abnormal synaptic transmission\n",
      "MP:0002206 - abnormal CNS synaptic transmission\n",
      "MP:0000778 - abnormal nervous system tract morphology\n",
      "MP:0002272 - abnormal nervous system electrophysiology\n",
      "MP:0009937 - abnormal neuron differentiation\n",
      "MP:0002882 - abnormal neuron morphology\n",
      "MP:0002910 - abnormal excitatory postsynaptic currents\n",
      "MP:0009939 - abnormal hippocampus neuron morphology\n"
     ]
    }
   ],
   "source": [
    "print(\"Top ten terms:\")\n",
    "print(\"\\n\".join([mp+\" - \"+get_MP_description(mp) for mp in MP_focal_top]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 19. Annotate the NetColoc systems map with mouse variant data, input genes, and enriched GO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:50:56.830837Z",
     "start_time": "2022-06-14T22:50:56.814509Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'd2_name' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-0e9d38b295fa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# don't annotate the root node\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mroot_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhier_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'CD_MemberList_Size'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mMP_full_results_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mroot_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gprofiler_name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0md1_name\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'-'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0md2_name\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m' systems map'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# also add the frac_seeds/num_seeds data here\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'd2_name' is not defined"
     ]
    }
   ],
   "source": [
    "# add the best gprofiler annotation\n",
    "MP_full_results_df['gprofiler_name']=pd.Series(system_name_list,index=hier_df.index.tolist())\n",
    "# don't annotate the root node\n",
    "root_node = hier_df['CD_MemberList_Size'].sort_values(ascending=False).head(1).index.tolist()[0]\n",
    "MP_full_results_df.loc[root_node, 'gprofiler_name']=d1_name+'-'+d2_name+' systems map'\n",
    "\n",
    "# also add the frac_seeds/num_seeds data here\n",
    "MP_full_results_df=MP_full_results_df.join(hier_df[['num_d1_seeds','frac_d1_seeds','frac_no_seeds']],\n",
    "                                          how='left')\n",
    "\n",
    "MP_full_results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 20. Export the NetColoc systems map to NDEx, with default style. \n",
    "\n",
    "Default style maps the fraction of seed genes from input set 1 (red) and input set 2 (blue) to node pie charts. The fraction of genes in each system that are in neither input set, but that are implicated by the network propagation are indicated in white."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:51:03.539863Z",
     "start_time": "2022-06-14T22:51:02.582840Z"
    }
   },
   "outputs": [],
   "source": [
    "# Convert G_hier to nice cx network\n",
    "node_id_to_node_name = nx.get_node_attributes(G_hier, 'name')\n",
    "for node_id in list(G_hier.nodes):\n",
    "    del G_hier.nodes[node_id]['name']\n",
    "\n",
    "G_hier_cx = ndex2.create_nice_cx_from_networkx(G_hier)\n",
    "\n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    node['n'] = node_id_to_node_name[node_id]\n",
    "\n",
    "G_hier_cx.set_name(d1_name+'_systems_map') \n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    data = MP_full_results_df.loc[node['r']]\n",
    "    for row, value in data.items():\n",
    "        if (row.find('gene_ids')>-1) or (row=='gprofiler_name'):\n",
    "            data_type = \"string\"\n",
    "            value=str(value)\n",
    "        else:\n",
    "            data_type = \"double\"\n",
    "            value = str(value) # nice cx can only accept strings as values...\n",
    "            if value=='inf': # check if inf, set to -1 if so\n",
    "                value='-1'\n",
    "            \n",
    "        G_hier_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "# Restore some hierarchy properties to their state before networkx conversion.\n",
    "for node_id, node in G_hier_cx.get_nodes():\n",
    "    for i in np.arange(len(G_hier_cx.nodeAttributes[node_id])):\n",
    "        dict_temp = G_hier_cx.nodeAttributes[node_id][i]\n",
    "        if dict_temp['n'] in ['CD_MemberList_Size','CD_MemberList_LogSize','HiDeF_persistence']:\n",
    "            G_hier_cx.set_node_attribute(node_id, dict_temp['n'], dict_temp['v'], type='double',overwrite=True)\n",
    "            \n",
    "# this is required so we can easily make subgraphs from systems\n",
    "G_hier_cx.set_network_attribute('__CD_OriginalNetwork',\n",
    "                                       values='0', type='long')\n",
    "            \n",
    "          \n",
    "# use apply_style_from_network-- this should overwrite the existing style\n",
    "netcoloc_template = ndex2.create_nice_cx_from_server('ndexbio.org',\n",
    "            uuid='f338dea0-117c-11ec-9e8e-0ac135e8bacf')\n",
    "G_hier_cx.apply_style_from_network(netcoloc_template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:51:25.145208Z",
     "start_time": "2022-06-14T22:51:16.029241Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDEx server (probably ndexbio.org): ndexbio.org\n",
      "NDEx user name: sbrosenthal@health.ucsd.edu\n",
      "NDEx password: ········\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "#Upload to NDEx. Enter your ndex username and password. \n",
    "# If you are new to ndex, make a new account on the website (ndexbio.org)\n",
    "G_hier_cx.set_name(d1_name+'_systems_map') \n",
    "SERVER = input('NDEx server (probably ndexbio.org): ')\n",
    "USERNAME = input('NDEx user name: ')\n",
    "PASSWORD = getpass.getpass('NDEx password: ')\n",
    "network_uuid_hier = G_hier_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 21. Apply another template style to NetColoc systems map for mouse knockout view, and export to NDEx. \n",
    "\n",
    "Select the property to be mapped to system node colors (should be one of the mouse knockout phenotypes we identified above). In this style, the log odds ratio is mapped to the system node color. Systems which are not significantly enriched for the phenotype are white (p<0.05).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:51:27.372741Z",
     "start_time": "2022-06-14T22:51:27.371098Z"
    }
   },
   "outputs": [],
   "source": [
    "# ------ modify this based on your project. Should be a system identified above -----\n",
    "# set the property we should map to system node colors\n",
    "mouse_KO_mapping_property = 'abnormal neuron morphology'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:51:30.311081Z",
     "start_time": "2022-06-14T22:51:29.349989Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating CX\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "# apply a template style \n",
    "G_hier_cx.set_name(d1_name+'_systems_map_mouse_KO:'+mouse_KO_mapping_property) \n",
    "# use apply_style_from_network-- this should overwrite existing style\n",
    "netcoloc_template = ndex2.create_nice_cx_from_server('ndexbio.org',\n",
    "            uuid='4958993c-df46-11eb-b666-0ac135e8bacf')\n",
    "raw_cx_st = json.dumps(netcoloc_template.to_cx())\n",
    "\n",
    "# replace the default template values with mouse_KO_mapping_property\n",
    "updated_raw_cx = re.sub('COL=abnormal heart development:log_OR', 'COL='+mouse_KO_mapping_property+':log_OR', raw_cx_st)\n",
    "updated_raw_cx = re.sub('COL=abnormal heart development:-log', \n",
    "                        'COL='+mouse_KO_mapping_property+':-log', updated_raw_cx)\n",
    "updated_raw_cx=json.loads(updated_raw_cx)\n",
    "netcoloc_template_updated = ndex2.create_nice_cx_from_raw_cx(updated_raw_cx)\n",
    "G_hier_cx.apply_style_from_network(netcoloc_template_updated)\n",
    "\n",
    "network_uuid_hier_mouse_KO = G_hier_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 22. Add genes associated with mouse variant phenotypes to NetColoc subnetwork, export to NDEx. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:51:53.725611Z",
     "start_time": "2022-06-14T22:51:53.709627Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d1_seeds</th>\n",
       "      <th>d1_name</th>\n",
       "      <th>z_d1</th>\n",
       "      <th>sum_seeds</th>\n",
       "      <th>abnormal nervous system development:gene_ids</th>\n",
       "      <th>abnormal synaptic physiology:gene_ids</th>\n",
       "      <th>abnormal synaptic transmission:gene_ids</th>\n",
       "      <th>abnormal CNS synaptic transmission:gene_ids</th>\n",
       "      <th>abnormal nervous system tract morphology:gene_ids</th>\n",
       "      <th>abnormal nervous system electrophysiology:gene_ids</th>\n",
       "      <th>abnormal neuron differentiation:gene_ids</th>\n",
       "      <th>abnormal neuron morphology:gene_ids</th>\n",
       "      <th>abnormal excitatory postsynaptic currents:gene_ids</th>\n",
       "      <th>abnormal hippocampus neuron morphology:gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KDM6B</th>\n",
       "      <td>1</td>\n",
       "      <td>ASD</td>\n",
       "      <td>10.991122</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASXL3</th>\n",
       "      <td>1</td>\n",
       "      <td>ASD</td>\n",
       "      <td>10.905475</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRAF7</th>\n",
       "      <td>1</td>\n",
       "      <td>ASD</td>\n",
       "      <td>10.895410</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2E</th>\n",
       "      <td>1</td>\n",
       "      <td>ASD</td>\n",
       "      <td>10.783263</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DIP2A</th>\n",
       "      <td>1</td>\n",
       "      <td>ASD</td>\n",
       "      <td>10.639522</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       d1_seeds d1_name       z_d1  sum_seeds  \\\n",
       "KDM6B         1     ASD  10.991122          1   \n",
       "ASXL3         1     ASD  10.905475          1   \n",
       "TRAF7         1     ASD  10.895410          1   \n",
       "KMT2E         1     ASD  10.783263          1   \n",
       "DIP2A         1     ASD  10.639522          1   \n",
       "\n",
       "       abnormal nervous system development:gene_ids  \\\n",
       "KDM6B                                             0   \n",
       "ASXL3                                             0   \n",
       "TRAF7                                             0   \n",
       "KMT2E                                             0   \n",
       "DIP2A                                             0   \n",
       "\n",
       "       abnormal synaptic physiology:gene_ids  \\\n",
       "KDM6B                                      0   \n",
       "ASXL3                                      0   \n",
       "TRAF7                                      0   \n",
       "KMT2E                                      0   \n",
       "DIP2A                                      0   \n",
       "\n",
       "       abnormal synaptic transmission:gene_ids  \\\n",
       "KDM6B                                        0   \n",
       "ASXL3                                        0   \n",
       "TRAF7                                        0   \n",
       "KMT2E                                        0   \n",
       "DIP2A                                        0   \n",
       "\n",
       "       abnormal CNS synaptic transmission:gene_ids  \\\n",
       "KDM6B                                            0   \n",
       "ASXL3                                            0   \n",
       "TRAF7                                            0   \n",
       "KMT2E                                            0   \n",
       "DIP2A                                            0   \n",
       "\n",
       "       abnormal nervous system tract morphology:gene_ids  \\\n",
       "KDM6B                                                  0   \n",
       "ASXL3                                                  0   \n",
       "TRAF7                                                  0   \n",
       "KMT2E                                                  0   \n",
       "DIP2A                                                  0   \n",
       "\n",
       "       abnormal nervous system electrophysiology:gene_ids  \\\n",
       "KDM6B                                                  0    \n",
       "ASXL3                                                  0    \n",
       "TRAF7                                                  0    \n",
       "KMT2E                                                  0    \n",
       "DIP2A                                                  0    \n",
       "\n",
       "       abnormal neuron differentiation:gene_ids  \\\n",
       "KDM6B                                         0   \n",
       "ASXL3                                         0   \n",
       "TRAF7                                         0   \n",
       "KMT2E                                         0   \n",
       "DIP2A                                         0   \n",
       "\n",
       "       abnormal neuron morphology:gene_ids  \\\n",
       "KDM6B                                    0   \n",
       "ASXL3                                    0   \n",
       "TRAF7                                    0   \n",
       "KMT2E                                    0   \n",
       "DIP2A                                    0   \n",
       "\n",
       "       abnormal excitatory postsynaptic currents:gene_ids  \\\n",
       "KDM6B                                                  0    \n",
       "ASXL3                                                  0    \n",
       "TRAF7                                                  0    \n",
       "KMT2E                                                  0    \n",
       "DIP2A                                                  0    \n",
       "\n",
       "       abnormal hippocampus neuron morphology:gene_ids  \n",
       "KDM6B                                                0  \n",
       "ASXL3                                                0  \n",
       "TRAF7                                                0  \n",
       "KMT2E                                                0  \n",
       "DIP2A                                                0  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add fields to node_df for genes in each mouse_KO phenotype of interest\n",
    "MP_genes_columns = [c for c in MP_full_results_df.columns.tolist() if c.find(':gene_ids')>-1]\n",
    "\n",
    "# look up overlapping genes in the root node, add them to node_df\n",
    "for MP in MP_genes_columns:\n",
    "    focal_genes = MP_full_results_df.loc[root_node, MP].split(' ')\n",
    "    node_df[MP]=0\n",
    "    node_df.loc[focal_genes, MP]=1\n",
    "node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T22:52:09.300898Z",
     "start_time": "2022-06-14T22:52:08.559539Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "\n",
    "G_prox_cx = ndex2.create_nice_cx_from_networkx(G_prox)\n",
    "G_prox_cx.set_name(d1_name+'_NetColoc_subgraph') \n",
    "for node_id, node in G_prox_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_prox_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "\n",
    "# apply a template style (834b6ad4-d2ea-11eb-b666-0ac135e8bacf)\n",
    "G_prox_cx.apply_template('ndexbio.org','834b6ad4-d2ea-11eb-b666-0ac135e8bacf')\n",
    "\n",
    "network_uuid_NetColoc = G_prox_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 23. Upload cosine-similarity transformed NetColoc subnetwork to NDEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T23:00:27.691424Z",
     "start_time": "2022-06-14T23:00:26.793299Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes in cosine similarity network: 416\n",
      "Number of edges in cosine similarity network: 1864\n",
      "Generating CX\n"
     ]
    }
   ],
   "source": [
    "#Annotate network\n",
    "print(\"Number of nodes in cosine similarity network:\", len(G_cosSim.nodes()))\n",
    "print(\"Number of edges in cosine similarity network:\", len(G_cosSim.edges()))\n",
    "G_cosSim_cx = ndex2.create_nice_cx_from_networkx(G_cosSim)\n",
    "G_cosSim_cx.set_name(d1_name+'_NetColoc_subgraph_CosSim95') \n",
    "for node_id, node in G_cosSim_cx.get_nodes():\n",
    "    data = node_df.loc[node['n']]\n",
    "    for row, value in data.items():\n",
    "        if row == 'd1_seeds' or row == 'd2_seeds' or row=='sum_seeds':\n",
    "            data_type = 'double'\n",
    "        elif row=='d1_name' or row=='d2_name':\n",
    "            data_type='string'\n",
    "        else:\n",
    "            data_type = 'double'\n",
    "        G_cosSim_cx.set_node_attribute(node_id, row, value, type=data_type)\n",
    "        \n",
    "        \n",
    "\n",
    "# apply a template style (834b6ad4-d2ea-11eb-b666-0ac135e8bacf)\n",
    "G_cosSim_cx.apply_template('ndexbio.org','2cbed84b-e5c3-11eb-b666-0ac135e8bacf')\n",
    "\n",
    "network_uuid_NetColoc_CosSim = G_cosSim_cx.upload_to(SERVER, USERNAME, PASSWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 24. Add 4 networks from above to network set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T19:50:59.999796Z",
     "start_time": "2022-06-14T19:50:59.863530Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ndexbio.org/v2/networkset/a860fa0f-5198-11ed-ae36-0ac135e8bacf'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# append the datestring to the network set to guarantee uniqueness\n",
    "datestr = str(datetime.now())\n",
    "networkSetURL=ndex2.client.Ndex2(host=SERVER,username=USERNAME,password=PASSWORD).create_networkset(d1_name+' network set: '+datestr,'network set for '+d1_name+' NetColoc subgraph and systems map')\n",
    "networkSetURL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-14T19:51:02.492806Z",
     "start_time": "2022-06-14T19:51:02.358891Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ndexbio.org/v2/networkset/a860fa0f-5198-11ed-ae36-0ac135e8bacf/members'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse out UUID from URL strings\n",
    "networkSetUUID = networkSetURL.split('/')[-1]\n",
    "networkSetUUID\n",
    "\n",
    "networkURLs = [network_uuid_NetColoc,network_uuid_NetColoc_CosSim,network_uuid_hier,network_uuid_hier_mouse_KO]\n",
    "networkUUIDs = [n.split('/')[-1] for n in networkURLs]\n",
    "\n",
    "ndex2.client.Ndex2(host=SERVER,username=USERNAME,password=PASSWORD).add_networks_to_networkset(networkSetUUID,\n",
    "                                                                                              networkUUIDs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
